---
output: 
  html_document:
      toc: false
      toc_float: false
      toc_depth: 4
      number_sections: false
      code_folding: hide
      df_print: paged
editor_options: 
  chunk_output_type: console
---

<!-- CSS style for tables to look ok: -->
<style>

table, td, th {
  border: none;
  padding-left: 1em;
  padding-right: 1em;
  min-width: 50%;
  <!-- max-width: 1600px; -->
  margin-left: 1em;
  margin-right: 1em;
  margin-top: 1em;
  margin-bottom: 1em;
}

</style>

<!-- Markdown setup -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,comment = '',warning=F,error = F)
options(width = 600)
```

# {.tabset .tabset-fade .tabset-pills}

 <!-- This script reads from a folder containing the CAW output, plus copy number data generated by controlfreec.-->
```{r,include=FALSE}
library(data.table)
library(knitr)
library(kableExtra)
library(ggplot2)
library(ggrepel)
library(VariantAnnotation)
library(gridExtra)
chrsz=structure(list(chr = c("1", "2", "3", "4", "5", "6", "7", "8", 
"9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", 
"20", "21", "22", "X", "Y"), starts = c(0, 253902921, 500669562, 
703689723, 898025604, 1084280925, 1259870526, 1424144247, 1574191848, 
1717288929, 1855896810, 1995944331, 2134165212, 2253485013, 2363996694, 
2470839615, 2565902256, 2654091777, 2739309138, 2802869979, 2871941700, 
2923598421, 2979326382, 3140008743), length = c(248902921, 241766641, 
198020161, 189335881, 181255321, 170589601, 159273721, 145047601, 
138097081, 133607881, 135047521, 133220881, 114319801, 105511681, 
101842921, 90062641, 83189521, 80217361, 58560841, 64071721, 
46656721, 50727961, 155682361, 56827081)), .Names = c("chr", 
"starts", "length"), row.names = c(NA, -24L), class = "data.frame")
#sampleDir='~/BTBdata/' # current dev sample
#sampleDir='.' # we assume we are in a sample dir
```

## Then {.tabset}

<!-- First, make a list of files to read and visualize -->
```{r,include=FALSE}

sampleData <- data.table(
  name=basename(getwd()),
  directory=getwd(),
  date=date()
)
knitr::kable(chrsz, format = "latex", caption = "Sample data, name and folder")

files=dir(recursive = T,full.names = T)
files=files[-grep(pattern = '.png',x = files)]
# control freec files
freec_Tratio_file=files[grep(pattern = 'T.hg38.pileup.gz_ratio.txt',files)]
freec_Nratio_file=files[grep(pattern = 'T.hg38.pileup.gz_normal_ratio.txt',files)]
freec_Tbaf_file=files[grep(pattern = 'T.hg38.pileup.gz_BAF.txt',files)]
freec_Nbaf_file=files[grep(pattern = 'B.hg38.pileup.gz_BAF.txt',files)]
freec_info_file=files[grep(pattern = 'hg38.pileup.gz_info.txt',files)]
# VEP structural variant files
vep_Tstruct_file=grep(pattern = glob2rx("*VEP/Manta*somaticSV.ann.vcf"),files,value = T)
vep_Nstruct_file=grep(pattern = glob2rx("*VEP/Manta*diploidSV.ann.vcf"),files,value = T)
# VEP small variant files
vep_Tsmall_file=grep(pattern = glob2rx("*VEP/haplotypecaller*T.vcf.ann.vcf"),files,value = T)
vep_Nsmall_file=grep(pattern = glob2rx("*VEP/haplotypecaller*N.vcf.ann.vcf"),files,value = T)
# snpEff Murect2 file
snpEff_mutect2_file=grep(pattern = glob2rx("*SnpEff/mutect2_*vcf.ann.vcf"),files,value = T)
```

### Mid1 

```{r}
kable(sampleData, "html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left")

kable(chrsz, "html") %>%
  kable_styling(bootstrap_options = "striped",position = "left") %>%
  scroll_box(width = "1000px", height = "600px")
```

### Mid2 {.active .tabset}

Tralala

#### tjosan

mämä

#### hoppsan

mömö

## Last


