---
title: "CAW output overview"
#author: "Markus Mayrhofer"
#date: "25 oktober 2017"
output: html_document
---

<!-- Markdown setup -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

 <!-- This script reads from a folder containing the CAW output, plus copy number data generated by controlfreec.-->
```{r,include=FALSE}
library(ggplot2)
library(VariantAnnotation)
library(gridExtra)
#sampleDir='~/BTBdata/' # current dev sample
sampleDir='.' # we assume we are in a sample dir
```

<!-- First, make a list of files to read and visualize -->
```{r,include=FALSE,echo=FALSE}
files=dir(path = sampleDir, recursive = T,full.names = T)
files=files[-grep(pattern = '.png',x = files)]
# control freec files
freec_Tratio_file=files[grep(pattern = 'T.hg38.pileup.gz_ratio.txt',files)]
freec_Nratio_file=files[grep(pattern = 'T.hg38.pileup.gz_normal_ratio.txt',files)]
freec_Tbaf_file=files[grep(pattern = 'T.hg38.pileup.gz_BAF.txt',files)]
freec_Nbaf_file=files[grep(pattern = 'B.hg38.pileup.gz_BAF.txt',files)]
freec_info_file=files[grep(pattern = 'hg38.pileup.gz_info.txt',files)]
# VEP structural variant files
vep_Tstruct_file=grep(pattern = glob2rx("*VEP/Manta*somaticSV.ann.vcf"),files,value = T)
vep_Nstruct_file=grep(pattern = glob2rx("*VEP/Manta*diploidSV.ann.vcf"),files,value = T)
# VEP small variant files
vep_Tsmall_file=grep(pattern = glob2rx("*VEP/haplotypecaller*T.vcf.ann.vcf"),files,value = T)
vep_Nsmall_file=grep(pattern = glob2rx("*VEP/haplotypecaller*N.vcf.ann.vcf"),files,value = T)
```


<!-- At top of document, list the files found and used -->
```{r,echo=F,include=T}
{
  cat(date(),'\n')
  cat('Current folder:', getwd(),'\n')
}
{
  cat('Folder Content:\n')
  print(files)
}

```



<!-- Read copy number data, use eval=F to skip -->
```{r,echo=F,warning=F}
# tumor log ratio
tratio=read.delim(file = freec_Tratio_file,stringsAsFactors = F)
tratio=tratio[order(as.numeric(tratio$Chromosome),tratio$Start),]
# normal log ratio
nratio=read.delim(file = freec_Nratio_file,stringsAsFactors = F)
nratio=nratio[order(as.numeric(nratio$Chromosome),nratio$Start),]
# tumor BAF
tbaf=read.delim(freec_Tbaf_file,stringsAsFactors = F)
# normal BAF
nbaf=read.delim(freec_Nbaf_file,stringsAsFactors = F)
# merged
mbaf=merge(tbaf[,1:3],nbaf[,1:3],by=1:2,all=F)

# manipulate for plot:
chrsz=data.frame(chr=unique(tratio$Chromosome),starts=0,length=0,stringsAsFactors = F)
for (i in 1:nrow(chrsz)) {
  chrsz$length[i]=max(tratio$Start[tratio$Chromosome==chrsz$chr[i]])
  if (i>1) chrsz$starts[i]=max(tratio$Start[tratio$Chromosome==chrsz$chr[i-1]])+chrsz$starts[i-1]+5e6
}
tratio$cumstart=tratio$Start
nratio$cumstart=nratio$Start
tbaf$cumstart=tbaf$Position
nbaf$cumstart=nbaf$Position
for (i in 2:nrow(chrsz)) {
  ix=tratio$Chromosome==chrsz$chr[i]
  tratio$cumstart[ix]=tratio$Start[ix]+chrsz$starts[i]
  ix=nratio$Chromosome==chrsz$chr[i]
  nratio$cumstart[ix]=nratio$Start[ix]+chrsz$starts[i]
  ix=tbaf$Chromosome==chrsz$chr[i]
  tbaf$cumstart[ix]=tbaf$Position[ix]+chrsz$starts[i]
  ix=nbaf$Chromosome==chrsz$chr[i]
  nbaf$cumstart[ix]=nbaf$Position[ix]+chrsz$starts[i]
}

# copy number hack
tratio$CopyNumber[tratio$CopyNumber>4]=4
nratio$CopyNumber[nratio$CopyNumber>4]=4
tratio$CopyNumber[1:2]=c(0,4)
nratio$CopyNumber[1:2]=c(0,4)
tratio$BAF[tratio$BAF<0.5 | tratio$BAF>1]=NA
nratio$BAF[nratio$BAF<0.5 | nratio$BAF>1]=NA

# smooth data for view
binned=NULL
for (i in 1:nrow(chrsz)) {
  temp=data.frame(
    chr=chrsz$chr[i],
    pos=seq(5e5,chrsz$length[i],5e5),
    cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
    tratio=NA,
    nratio=NA,
    tmaf=NA,
    nmaf=NA)
  ctratio=tratio[tratio$Chromosome==chrsz$chr[i],]
  cnratio=nratio[nratio$Chromosome==chrsz$chr[i],]
  #ctbaf=tbaf[tbaf$Chromosome==chrsz$chr[i],]
  #cnbaf=tbaf[tbaf$Chromosome==chrsz$chr[i],]
  for (j in 1:nrow(temp)) {
    ix = ctratio$Start>temp$pos[j]-5e6 & ctratio$Start<temp$pos[j]+5e6
    temp$tratio[j]=median(ctratio$Ratio[ix],na.rm = T)
    if (sum(ix)>20) {
      d=density(ctratio$BAF[ix],na.rm=T)
      temp$tmaf[j]=d$x[which.max(d$y)]
    }
    ix = cnratio$Start>temp$pos[j]-5e6 & cnratio$Start<temp$pos[j]+5e6
    temp$nratio[j]=median(cnratio$Ratio[ix],na.rm = T)
    if (sum(ix)>20) {
      d=density(cnratio$BAF[ix],na.rm=T)
      temp$nmaf[j]=d$x[which.max(d$y)]
    }
  }
  binned=rbind(binned,temp)
}
```

#### Variant allele ratio concordance
<!-- Compare the normal and tumor SNP allele ratios -->
```{r, echo=FALSE,,fig.width=3,fig.height=3}
g=ggplot()+theme_light()+
  geom_point(aes(x = BAF.y,y = BAF.x),data=mbaf,alpha=1/50,shape='.')+
  xlab('Normal allele ratio')+ylab('Tumor allele ratio')
g  

```


<!-- Scatterplots of binned data -->
#### Tumor copy number and major allele ratio 
```{r, echo=FALSE,,fig.width=8,fig.height=4,warning=F}
g=ggplot()+theme_light()+expand_limits(x=c(0.4,1.4))+expand_limits(y=c(.5,1))+
  geom_point(aes(x = tratio,y = tmaf,col=chr),data=binned)+
  xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')
g
```



#### Genome copy number alteration

<!-- Plot tumor copy number log ratio -->
```{r, echo=FALSE,fig.width=15,fig.height=3,warning=F}
{
  par(mai=c(0,4,0,2))
  g=ggplot(data = tratio,aes(x=cumstart,y=Ratio,col=CopyNumber))+
    #scale_color_gradientn(colours = c('black','red','orange'))+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    geom_point(alpha=1/5,shape='.')+
    ylab('Tumor DNA ratio')+
    expand_limits(y=c(0.4,1.6))+
    expand_limits(x=c(0,3200e6))
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=binned)
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0.5,yend=1.5),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.6,label=chr),data = chrsz,inherit.aes = F)
  
  
  
  g
}
```

<!-- Plot normal copy number log ratio -->
```{r, echo=FALSE,fig.width=15,fig.height=3,warning=F}
{
  par(mai=c(0,4,0,2))
  g=ggplot(data = nratio,aes(x=cumstart,y=Ratio,col=CopyNumber))+
    geom_point(alpha=1/5,shape='.')+
    ylab('Normal DNA ratio')+ylim(c(0.4,1.7))+
    expand_limits(x=c(0,3200e6))+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))
    #scale_color_gradientn(midpoint=2, low="lightblue", mid="darkblue",high="red", space ="Lab" )
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=nratio),stat="identity",colour="green",data=binned)
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0.5,yend=1.5),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.6,label=chr),data = chrsz,inherit.aes = F)
  
  
  
  g
}
```


#### Variant allele ratio

<!-- Plot tumor copy number BAF -->
```{r, echo=FALSE,fig.width=15,fig.height=3}
{
  par(mai=c(0,4,0,2))
  #plot(tbaf$cumstart,tbaf$BAF,pch=16,col='#00000010',cex=0.2,xlab=NA,ylab='Tumor allele ratio',xaxt='n')
  #text(x = chrsz$starts+chrsz$length/2,y = .2,labels = chrsz$chr,col='blue')
  g=ggplot(data = tbaf,aes(x=cumstart,y=BAF))+
    geom_point(alpha=1/50,shape='.')+
    ylab('Tumor allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))
  # add smoothed
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  
  
  g
}
```



<!-- Plot normal copy number BAF -->
```{r, echo=FALSE,fig.width=15,fig.height=3}
{
  par(mai=c(0,4,0,2))
  g=ggplot(data = nbaf,aes(x=cumstart,y=BAF))+
    geom_point(alpha=1/50,shape='.')+
    ylab('Normal allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))
  # add smoothed
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  
  
  g
}
```



#### Structural variants

<!-- Read the tumor structural variant data -->
```{r, echo=FALSE}
tstruc_vcf=readVcf(file = vep_Tstruct_file,genome = 'GRCh38')
g=geno(tstruc_vcf)
inf=info(tstruc_vcf)
rr=rowRanges(tstruc_vcf)

# manipulate into a data frame with relevant data
tstruc=data.frame(rr,stringsAsFactors = F)
tstruc$chr <- substr(x = tstruc$seqnames,start = 4,stop = 6)
tstruc$cumstart=tstruc$start
tstruc$cumend=tstruc$end
tstruc$altchr=tstruc$chr
tstruc$altpos=tstruc$end
# loop through all and extract endpoint chr and pos
for (i in 1:nrow(tstruc)) {
  t=strsplit(x = tstruc$ALT[[i]],split = ':')[[1]]
  if (length(t)>1 & t[1]!="<DUP") {
    tstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
    tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
    tstruc$altpos[i]=as.numeric(tt)
  }
}
# # for each chr get cumulative pos
tstruc$altcumpos=tstruc$altpos
for (i in 1:nrow(chrsz)) {
  ix=tstruc$chr==chrsz$chr[i]
  tstruc$cumstart[ix]=tstruc$start[ix]+chrsz$starts[i]
  tstruc$cumend[ix]=tstruc$end[ix]+chrsz$starts[i]
  ix=tstruc$altchr==chrsz$chr[i]
  tstruc$altcumpos[ix]=tstruc$altpos[ix]+chrsz$starts[i]
}
# "info"" annotations also need to be added
tstruc=cbind(tstruc,data.frame(info(tstruc_vcf),stringsAsFactors = F))
# much annotation is in the CSQ column. get Symbol and Consequence for now.
tstruc$symbol=''
tstruc$consequence=''
h=strsplit(info(header(tstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
h=strsplit(h,'\\|')[[1]]
for (i in 1:length(tstruc$CSQ)) { # for each variant
  symbol=NULL
  consequence=NULL
  for (j in 1:length(tstruc$CSQ[[i]])) { # for each annotated transcript
    symbol=c(symbol,strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]][h=='SYMBOL'])
    consequence=c(consequence,strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]][h=='Consequence'])
  }
  tstruc$symbol[i]=paste(unique(symbol),collapse=',')
  tstruc$consequence[i]=paste(unique(consequence),collapse=',')
}
```

<!-- Read the normal structural variant data -->
```{r, echo=FALSE}
nstruc_vcf=readVcf(file = vep_Nstruct_file,genome = 'GRCh38')
g=geno(nstruc_vcf)
inf=info(nstruc_vcf)
rr=rowRanges(nstruc_vcf)

# manipulate into a data frame with relevant data
nstruc=data.frame(rr,stringsAsFactors = F)
nstruc$chr <- substr(x = nstruc$seqnames,start = 4,stop = 6)
nstruc$cumstart=nstruc$start
nstruc$cumend=nstruc$end
nstruc$altchr=nstruc$chr
nstruc$altpos=nstruc$end
# loop through all and extract endpoint chr and pos
for (i in 1:nrow(nstruc)) {
  t=strsplit(x = nstruc$ALT[[i]],split = ':')[[1]]
  if (length(t)>1 & t[1]!="<DUP") {
    nstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
    tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
    nstruc$altpos[i]=as.numeric(tt)
  }
}
# # for each chr get cumulative pos
nstruc$altcumpos=nstruc$altpos
for (i in 1:nrow(chrsz)) {
  ix=nstruc$chr==chrsz$chr[i]
  nstruc$cumstart[ix]=nstruc$start[ix]+chrsz$starts[i]
  nstruc$cumend[ix]=nstruc$end[ix]+chrsz$starts[i]
  ix=nstruc$altchr==chrsz$chr[i]
  nstruc$altcumpos[ix]=nstruc$altpos[ix]+chrsz$starts[i]
}
# "info"" annotations also need to be added
nstruc=cbind(nstruc,data.frame(info(nstruc_vcf),stringsAsFactors = F))
# much annotation is in the CSQ column. get Symbol and Consequence for now.
nstruc$symbol=''
nstruc$consequence=''
h=strsplit(info(header(nstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
h=strsplit(h,'\\|')[[1]]
for (i in 1:length(nstruc$CSQ)) { # for each variant
  symbol=NULL
  consequence=NULL
  for (j in 1:length(nstruc$CSQ[[i]])) { # for each annotated transcript
    symbol=c(symbol,strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]][h=='SYMBOL'])
    consequence=c(consequence,strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]][h=='Consequence'])
  }
  nstruc$symbol[i]=paste(unique(symbol),collapse=',')
  nstruc$consequence[i]=paste(unique(consequence),collapse=',')
}
```



<!-- Plot the tumor structural variant data -->
```{r, echo=FALSE,fig.width=15,fig.height=3}
{
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Somatic structural variants')+expand_limits(x=c(0,3210e6))
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=2),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=1,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=1,yend=1.001,col=SOMATICSCORE),
                 data = tstruc[tstruc$FILTER=='PASS',],alpha=1/10,
                 stat = "identity",
                 position = "identity", curvature = 0.2, angle = 90, ncp = 50,
                 arrow = NULL, lineend = "butt", na.rm = FALSE, show.legend = NA,
                 inherit.aes = FALSE)
  g
}
```



<!-- Plot the normal structural variant data -->
```{r, echo=FALSE,fig.width=15,fig.height=3}
{
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Diploid structural variants')+expand_limits(x=c(0,3210e6))
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=2),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=1,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=1,yend=1.001,col=QUAL),
                 data = nstruc[tstruc$FILTER=='PASS',],alpha=1/10,
                 stat = "identity",
                 position = "identity", curvature = 0.2, angle = 90, ncp = 50,
                 arrow = NULL, lineend = "butt", na.rm = FALSE, show.legend = NA,
                 inherit.aes = FALSE)
  g
}
```


<!-- Read annotated small variant data (Mutect 2) -->
```{r, echo=FALSE}
vcf=readVcf(file = '~/BTBdata/SnpEff/mutect2_P2233_103T_vs_P2233_122N.vcf.ann.vcf',genome ='GRCh38')
g=geno(vcf)
inf=(info(vcf))
rr=rowRanges(vcf)
# manipulate into a data frame with relevant data
tmuts=data.frame(ranges(rowRanges(vcf)),stringsAsFactors = F)
tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
tmuts$cumstart=tmuts$start
tmuts$cumend=tmuts$end
tmuts$ref=as.character(ref(vcf))
tmuts$alt=as.character(unlist(alt(vcf)))
tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
tmuts$type[nchar(tmuts$type)>3]='other'
tmuts$type[tmuts$type=='T>G']='A>C'
tmuts$type[tmuts$type=='T>C']='A>G'
tmuts$type[tmuts$type=='T>A']='A>T'
tmuts$type[tmuts$type=='G>T']='C>A'
tmuts$type[tmuts$type=='G>C']='C>G'
tmuts$type[tmuts$type=='G>A']='C>T'
tmuts$filter=rr$FILTER
# # for each chr get cumulative pos
for (i in 1:nrow(chrsz)) {
  ix=tmuts$chr==chrsz$chr[i]
  tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
  tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  ix=tmuts$altchr==chrsz$chr[i]
  tmuts$altcumpos[ix]=tmuts$altpos[ix]+chrsz$starts[i]
}
tad=t(as.data.frame((g$AD[,1])))
tsum=tad[,1]+tad[,2]
nad=t(as.data.frame((g$AD[,2])))
nsum=nad[,1]+nad[,2]
tmuts=cbind(tmuts,g$AF)
tmuts$AD=tad[,2]
tmuts$DP=tsum
tmuts$ADn=nad[,2]
tmuts$DPn=nsum
# "info"" annotations also need to be added
tmuts=cbind(tmuts,data.frame(info(vcf),stringsAsFactors = F))
# much annotation is in the CSQ column. get Symbol, Consequence, aa, ppos for now.
tmuts$symbol=''
tmuts$consequence=''
#tmuts$protein_pos=''
#tmuts$amino_acid=''
h=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
h=strsplit(h,' \\| ')[[1]]
for (i in 1:length(tmuts$ANN)) { # for each variant
  symbol=NULL
  consequence=NULL
  for (j in 1:length(tmuts$ANN[[i]])) { # for each annotated transcript
    symbol=c(symbol,strsplit(tmuts$ANN[[i]][j],'\\|')[[1]][h=='Gene_Name'])
    consequence=c(consequence,strsplit(tmuts$ANN[[i]][j],'\\|')[[1]][h=='Annotation'])
    protein_pos=c(consequence,strsplit(tmuts$ANN[[i]][j],'\\|')[[1]][h=='HGVS.p'])
  }
  tmuts$symbol[i]=paste(unique(symbol),collapse=',')
  tmuts$consequence[i]=paste(unique(consequence),collapse=',')
  #tmuts$protein_pos[i]=paste(unique(protein_pos),collapse=',')
}
```

#### Mutect 2 somatic mutations

<!-- Plot the mutations -->
```{r, echo=FALSE,fig.width=15,fig.height=3}
{
  par(mai=c(0,4,0,2))
  g=ggplot(data = tmuts[tmuts$DPn>=10 & tmuts$ADn<1 & tmuts$AD>2 &
                          tmuts$TUMOR>0.02 & !tmuts$DB & tmuts$filter=='PASS',],
           aes(x=cumstart,y=TUMOR,col=type))+
    geom_point(alpha=1)+
    ylab('Somatic mutation allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))#+
    #scale_color_gradient2(low = 'lightgrey',mid = 'black',high = 'red',midpoint = 8)
  # add smoothed
  # fix theme
  g=g+theme_light()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  
  
  g
}
```







<!-- Plot the same/similar data but chromosome wise -->
<!-- todo -->











