---
output: 
  html_document:
      toc: false
      toc_float: false
      toc_depth: 4
      code_folding: hide
      df_print: paged
editor_options: 
  chunk_output_type: console
---

 <!-- This script reads from a folder containing CAW (aka Sarek) output, including copy number data from ControlFreec and Ascat, structural variation from Manta, somatic point mutations from Mutact 2 and Strelka, and germline mutations from HaplotypeCaller.-->


# {.tabset .tabset-pills}

## Setup {.tabset}
<!-- Markdown setup -->
```{r setup}
knitr::opts_chunk$set(echo = TRUE,comment = '',warning=F,error = F)
```


### List files {.active}
<!-- First, make a list of files to read and visualize -->
```{r}
files <- dir(recursive = T,full.names = T)
files <- files[-grep(pattern = '.png',x = files)]
# control freec files
freec_Tratio_file <- files[grep(pattern = 'T.hg38.pileup.gz_ratio.txt',files)]
freec_Nratio_file <- files[grep(pattern = 'T.hg38.pileup.gz_normal_ratio.txt',files)]
freec_Tbaf_file <- files[grep(pattern = 'T.hg38.pileup.gz_BAF.txt',files)]
freec_Nbaf_file <- files[grep(pattern = 'B.hg38.pileup.gz_BAF.txt',files)]
freec_info_file <- files[grep(pattern = 'hg38.pileup.gz_info.txt',files)]
freec_cnv_file <- files[grep(pattern = glob2rx('*T.hg38.pileup.gz_CNVs'),files)]
# Ascat files
ascat_Tratio_file <- files[grep(pattern = glob2rx("*Ascat*T.LogR"),files)]
ascat_Nratio_file <- files[grep(pattern = glob2rx("*Ascat*N.LogR"),files)]
ascat_Tbaf_file <- files[grep(pattern = glob2rx("*Ascat*T.BAF"),files)]
ascat_Nbaf_file <- files[grep(pattern = glob2rx("*Ascat*N.BAF"),files)]
ascat_segment_file <- files[grep(pattern = glob2rx("*Ascat*T.cnvs.txt"),files)]
# Manta structural variant files (VEP here)
vep_Tstruct_file <- grep(pattern = glob2rx("*VEP/Manta*somaticSV.*ann.vcf"),files,value = T)
vep_Nstruct_file <- grep(pattern = glob2rx("*VEP/Manta*diploidSV.*ann.vcf"),files,value = T)
# Haplotype caller variant files
haplotypecaller_T_file <- grep(pattern = glob2rx("*VEP/haplotypecaller*T.vcf.vep.ann.vcf"),files,value = T) # local
if (length(haplotypecaller_T_file)==0)
  haplotypecaller_T_file <- grep(pattern = glob2rx("*/Annotation/VEP/haplotypecaller_*T.vcf.vep.ann.vcf"),files,value = T) # bianca
haplotypecaller_N_file <- grep(pattern = glob2rx("*VEP/haplotypecaller*N.vcf.vep.ann.vcf"),files,value = T) # local
if (length(haplotypecaller_N_file)==0)
  haplotypecaller_N_file <- grep(pattern = glob2rx("*/Annotation/VEP/haplotypecaller_*N.vcf.vep.ann.vcf"),files,value = T) # bianca
# snpEff+VEP Mutect2 file
mutect2_file <- grep(pattern = glob2rx("*/Annotation/mutect2_*canon.vep.ann.pick.PASS.recode.vcf"),files,value = T) # <- local
if (length(mutect2_file)==0)
  mutect2_file <- grep(pattern = glob2rx("*Annotation/VEP/mutect2_*.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca
# snpEff + VEP Strelka snv file
strelka_snv_file <- grep(pattern = glob2rx("*/Annotation/Strelka_*somatic_snvs.snpeff.ann.canon.vep.ann.pick.vcf"),files,value = T) # <- local
if (length(strelka_snv_file)==0)
  strelka_snv_file <- grep(pattern = glob2rx("*Annotation/VEP/Strelka_*_somatic_snvs.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca
# snpEff + VEP Strelka indel file
strelka_indel_file <- grep(pattern = glob2rx("*/Annotation/Strelka_*somatic_indels.snpeff.ann.canon.vep.ann.pick.vcf"),files,value = T) # <- local
if (length(strelka_indel_file)==0)
  strelka_indel_file <- grep(pattern = glob2rx("*Annotation/VEP/Strelka_*_somatic_indels.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca

rm(files)
cat('Selected files in',getwd(),'\n\n')
for ( obj in ls() ) { 
  cat('---',obj,'---\n') 
  print(get(obj)) 
}


```


```{r}
{
  cat('All files in',getwd(),'\n\n')
  print(dir(recursive = T,full.names = T))
}
```

### Load dependencies
<!-- Dependencies and reference data -->
```{r}
library(data.table)
library(ggplot2)
theme_set(theme_light())
library(ggrepel)
library(VariantAnnotation)
library(VennDiagram)

load('~/reports/swegen_hg38_allele_counts.Rdata') #  swegen_hg38_allele_counts
load('~/reports/mutations_genes_ref.Rdata') #  cosmic,msk,oncokb
load('~/reports/BTB_hg38_manta_counts.Rdata') #  btb_hg38_manta


listfix <- function(list) {
  l=lapply(list,length)
  for (i in which(l!=1)) list[[i]]=0
  unlist(list)
}

sampleData <- data.table(
  date=date,
  directory=getwd(),
  name=basename(getwd())
)

```

## Copy number {.tabset}

### Parse ControlFreec
<!-- Read Control Freec copy number data -->
```{r}
if (length(freec_Tratio_file)>0 &
    length(freec_Nratio_file)>0 &
    length(freec_Tbaf_file)>0 &
    length(freec_Nbaf_file)>0) {
  # tumor log ratio
  tratio <- fread(file = freec_Tratio_file)
  tratio <- tratio[order(Chromosome,Start)]
  # normal log ratio
  nratio <- fread(file = freec_Nratio_file)
  nratio <- nratio[order(Chromosome,Start)]
  # tumor BAF
  tbaf <- fread(freec_Tbaf_file)[,1:3]
  # normal BAF
  nbaf <- fread(freec_Nbaf_file)[,1:3]
  # merged
  mbaf <- merge(tbaf,nbaf,by = c('Chromosome','Position'))

  # manipulate for plot:
  tratio$cumstart <- tratio$Start
  nratio$cumstart <- nratio$Start
  tbaf$cumstart <- tbaf$Position
  nbaf$cumstart <- nbaf$Position
  for (i in 2:nrow(chrsz)) {
    ix <- tratio$Chromosome==chrsz$chr[i]
    tratio$cumstart[ix] <- tratio$Start[ix]+chrsz$starts[i]
    ix <- nratio$Chromosome==chrsz$chr[i]
    nratio$cumstart[ix] <- nratio$Start[ix]+chrsz$starts[i]
    ix <- tbaf$Chromosome==chrsz$chr[i]
    tbaf$cumstart[ix] <- tbaf$Position[ix]+chrsz$starts[i]
    ix <- nbaf$Chromosome==chrsz$chr[i]
    nbaf$cumstart[ix] <- nbaf$Position[ix]+chrsz$starts[i]
  }
  
  # modify for plot
  tratio$CopyNumber[tratio$CopyNumber>4] <- 4
  nratio$CopyNumber[nratio$CopyNumber>4] <- 4
  tratio$CopyNumber[1:2] <- c(0,4)
  nratio$CopyNumber[1:2] <- c(0,4)
  tratio$BAF[tratio$BAF<0.5 | tratio$BAF>1] <- NA
  nratio$BAF[nratio$BAF<0.5 | nratio$BAF>1] <- NA
  
  # smooth data for plot
  binned <- NULL
  for (i in 1:nrow(chrsz)) {
    temp <- data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      nratio=NA,
      tmaf=NA,
      nmaf=NA)
    ctratio <- tratio[Chromosome==chrsz$chr[i]]
    cnratio <- nratio[Chromosome==chrsz$chr[i]]
    for (j in 1:nrow(temp)) {
      ix <- ctratio$Start>temp$pos[j]-5e6 & ctratio$Start<temp$pos[j]+5e6
      temp$tratio[j] <- median(ctratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) {
        d <- density(ctratio$BAF[ix],na.rm=T)
        temp$tmaf[j] <- d$x[which.max(d$y)]
      }
      ix <- cnratio$Start>temp$pos[j]-5e6 & cnratio$Start<temp$pos[j]+5e6
      temp$nratio[j] <- median(cnratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) try({
        d <- density(cnratio$BAF[ix],na.rm=T)
        temp$nmaf[j] <- d$x[which.max(d$y)]
      },silent=T)
    }
    binned <- rbind(binned,temp)
  }
  
  # make CNA burden estimate from binned:
  sampleData[['Control_Freec_CNA_burden']] <- round(binned[chr %in% 1:22,sum(abs(tratio-median(tratio))>0.1,na.rm = T)/.N]*100)
  
  # check which regions are worth reporting
  # cnv file
  cnvs <- fread(freec_cnv_file)
  names(cnvs) <- c('chr','start','end','copies','effect','genotype','value','type','na')
  cnvs$Gene_Tier <- 3
  cnvs$Effect_Tier <- 3
  cnvs$cosmic_genes <- ''
  cnvs$all_genes=''
  # amplifications/duplications max 5MB, homdel max 3MB
  cnvs <- cnvs[(end-start < 5e6 & effect=='gain') | (end-start < 3e6 & copies==0)]
  # add genes 
  for (i in 1:nrow(cnvs)) {
    all=allgenes[chr==cnvs$chr[i] & start<cnvs$end[i] & end>cnvs$end[i],name]
    if (length(all)>0) { # if any genes in cnv
      cnvs$all_genes[i]=paste(all,collapse = ' ') # put in table
      t2=all[(all %in% tumorgenes$`Gene Symbol`)] 
      if (length(t2)>0) { # if any in cosmic
        cnvs$cosmic_genes[i] <- paste(t2,collapse = ' ') # put in table
        cnvs$Gene_Tier[i]=2
        cnvs$Effect_Tier[i]=2
        # if the effect is TSG loss or oncogene amp, Tier 1
        if (cnvs$copies[i]==0) if (any(all %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`]))
          cnvs$Effect_Tier[i]=1
        if (cnvs$effect[i]=='gain') if (any(all %in% tumorgenes[grep('oncogene',`Role in Cancer`),`Gene Symbol`]))
          cnvs$Effect_Tier[i]=1
      }
      t1=all[(all %in% tumorgenes[Tier==1,`Gene Symbol`])] 
      if (length(t1)>0) { # if any in cosmic T1
        cnvs$Gene_Tier[i]=1
      }
    }
  }
  
}
```

### Parse Ascat
<!-- Read ASCAT copy number data -->
```{r,echo=F,include=F}
if (length(ascat_Tratio_file)>0 &
    length(ascat_Nratio_file)>0 &
    length(ascat_Tbaf_file)>0 &
    length(ascat_Nbaf_file)>0 & 
    length(ascat_segment_file)>0) {
  # segmented copy number
  ascat_cnv=fread(file = ascat_segment_file)
  # tumor log ratio
  ascat_tratio=fread(file = ascat_Tratio_file)[,-1]
  colnames(ascat_tratio)[3]='LogR'
  ascat_tratio$Chr=substr(ascat_tratio$Chr,start = 4,stop = 8)
  ascat_tratio=ascat_tratio[Chr %in% c(1:22,'X','Y')]
  ascat_tratio=ascat_tratio[order(Chr,Position)]
  # tumor BAF
  ascat_tbaf=fread(ascat_Tbaf_file)[,-1]
  colnames(ascat_tbaf)[3]='BAF'
  ascat_tbaf=ascat_tbaf[which(BAF!=0 & BAF!=1)]
  ascat_tbaf$Chr=substr(ascat_tbaf$Chr,start = 4,stop = 8)
  ascat_tbaf=ascat_tbaf[Chr %in% c(1:22,'X','Y')]
  # normal BAF
  ascat_nbaf=fread(ascat_Nbaf_file)[,-1]
  colnames(ascat_nbaf)[3]='NBAF'
  ascat_nbaf=ascat_nbaf[which(NBAF!=0 & NBAF!=1)]
  ascat_nbaf$Chr=substr(ascat_nbaf$Chr,start = 4,stop = 8)
  ascat_nbaf=ascat_nbaf[Chr %in% c(1:22,'X','Y')]
  
  # join (T) log ratio and (T+N) BAF
  ascat_tratio=merge(ascat_tratio,ascat_tbaf,all.x=T)
  ascat_tratio=merge(ascat_tratio,ascat_nbaf,all.x=T)
  
  # manipulate for plot:
  ascat_tratio$cumstart=ascat_tratio$Position
  ascat_tbaf$cumstart=ascat_tbaf$Position
  ascat_nbaf$cumstart=ascat_nbaf$Position
  ascat_cnv$cumstart=ascat_cnv$startpos
  ascat_cnv$cumend=ascat_cnv$endpos
  ascat_cnv$LOH=''; ascat_cnv$LOH[ascat_cnv$nMinor==0]='LOH'
  for (i in 2:nrow(chrsz)) {
    ix=ascat_tratio$Chr==chrsz$chr[i]
    ascat_tratio$cumstart[ix]=ascat_tratio$Position[ix]+chrsz$starts[i]
    #no n...
    ix=ascat_tbaf$Chr==chrsz$chr[i]
    ascat_tbaf$cumstart[ix]=ascat_tbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_nbaf$Chr==chrsz$chr[i]
    ascat_nbaf$cumstart[ix]=ascat_nbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_cnv$chr==chrsz$chr[i]
    ascat_cnv$cumstart[ix]=ascat_cnv$startpos[ix]+chrsz$starts[i]
    ascat_cnv$cumend[ix]=ascat_cnv$endpos[ix]+chrsz$starts[i]
  }
  # copy number hack
  ascat_tratio$Ratio=2^ascat_tratio$LogR
  ascat_tratio$smoothed=runmed(x = ascat_tratio$Ratio,k = 99)
  ascat_tratio$CopyNumber=2
  ascat_tratio$MinorCopy=1
  for (i in 1:nrow(ascat_cnv)) {
    ix=ascat_tratio$cumstart > ascat_cnv$cumstart[i] & ascat_tratio$cumstart < ascat_cnv$cumend[i]
    ascat_tratio$CopyNumber[ix]=ascat_cnv$nMajor[i]+ascat_cnv$nMinor[i]
    ascat_tratio$MinorCopy[ix]=ascat_cnv$nMinor[i]
  }
  ascat_tratio$CopyNumber[ascat_tratio$CopyNumber>4]=4
  ascat_tratio$CopyNumber[1:2]=c(0,4)
  ascat_tratio$LOH='no'
  ascat_tratio$LOH[ascat_tratio$MinorCopy==0]='yes'
  ascat_tratio$Copies=as.character(ascat_tratio$CopyNumber)
  ascat_tratio$Copies[ascat_tratio$LOH=='yes' & ascat_tratio$CopyNumber>1]='LOH'
  ascat_tratio$Copies[ascat_tratio$CopyNumber==4]='≥4'
  ascat_tratio$Copies[1:6]=c('0','1','LOH','2','3','≥4')

  # smooth data for view
  ascat_binned=NULL
  for (i in 1:nrow(chrsz)) {
    temp=data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      tmaf=NA)
    
    cix=ascat_tratio$Chr==chrsz$chr[i]
    tpos=ascat_tratio$Position[cix]
    tLogR=ascat_tratio$LogR[cix]
    tBAF=0.5+abs(ascat_tratio$BAF[cix]-0.5)
    for (j in 1:nrow(temp)) {
      ix = tpos>temp$pos[j]-5e6 & tpos<temp$pos[j]+5e6
      temp$tratio[j]=2^median(tLogR[ix],na.rm = T)
      if (sum(!is.na(tBAF[ix]))>20) {
        d=density(tBAF[ix],na.rm=T)
        temp$tmaf[j]=d$x[which.max(d$y)]
      }
    }
    ascat_binned=rbind(ascat_binned,temp)
  }
  # make CNA burden estimate from binned
  sampleData[['Ascat_CNA_burden']] <- round(ascat_binned[chr %in% 1:22,sum(abs(tratio-median(tratio))>0.1,na.rm = T)/.N]*100)
}
```



### View ControlFreec {.active}

#### Genome view
<!-- Plot Control Freec tumor copy number log ratio, tBAF and nBAF -->
```{r,fig.width=15,fig.height=10}
if (exists('tratio')) {
  cat('Control Freec: ',freec_Tratio_file)
  par(mai=c(0,4,0,2))
  temp <- tratio
  temp$Ratio[temp$Ratio>3]=3

  g=ggplot()+ylab('B allele ratio and Coverage Ratio')+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    ylim(-1,3)+
    expand_limits(x=c(0,3200e6))
  # normal logR below
  g=g+geom_point(aes(x=cumstart,y=Ratio-0.3),col='darkgrey',data=nratio,alpha=1/5,shape='.')
  # tumor logR
  g=g+geom_point(aes(x=cumstart,y=Ratio,col=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(data = tbaf,aes(x=cumstart,y=-0.5+0.5*BAF),
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(data = nbaf,aes(x=cumstart,y=-1+0.5*BAF),
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)

  g
}
```

&nbsp;

#### Scatter plot
<!-- Scatter plot -->
```{r, echo=FALSE,,fig.width=8,fig.height=4,warning=F}
if (exists('binned')) {
  cat('Control freec:')
  binned$tmaf[binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+#xlim(0.2,1.8)+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=binned[binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,1.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```



### View Ascat

#### Genome view
<!-- Plot ASCAT tumor copy number log ratio, tBAF and nBAF -->
```{r,fig.width=15,fig.height=10}
if (exists('ascat_tratio')) {
  cat('ASCAT: ',ascat_Tratio_file,ascat_segment_file,ascat_Tbaf_file,ascat_Nbaf_file)
  par(mai=c(0,4,0,2))
  temp <- ascat_tratio
  temp$smoothed[temp$smoothed>3]=3
  g=ggplot()+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    ylim(-1,3)+
    expand_limits(x=c(0,3200e6))
  # using smoothed logR
  g=g+geom_point(aes(x=cumstart,y=smoothed,color=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=ascat_binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-0.5+0.5*BAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-1+0.5*NBAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)

  g
}
```

&nbsp;

#### Scatter plot
<!-- Scatter plot -->
```{r, echo=FALSE,,fig.width=8,fig.height=4,warning=F}
if (exists('ascat_binned')) {
  cat('Ascat:')
  #ascat_binned$tmaf[ascat_binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+#xlim(0.2,1.8)+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=ascat_binned[ascat_binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,1.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```

### ControlFreec table

### Ascat table

## Structural variants {.tabset}

<!-- Read the tumor structural variant data -->
### Parse Manta (tumor)
```{r, echo=FALSE}
if (length(vep_Tstruct_file)>0) {
  tstruc_vcf=readVcf(file = vep_Tstruct_file,genome = 'GRCh38')
  g=geno(tstruc_vcf)
  inf=info(tstruc_vcf)
  rr=rowRanges(tstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  tstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))[FILTER=='PASS' & IMPRECISE==F,]
  tstruc$SVLEN=listfix(tstruc$SVLEN)
  setkey(x = tstruc,THISID)
  tstruc$chr <- substr(x = tstruc$seqnames,start = 4,stop = 6)
  tstruc$cumstart=tstruc$start
  tstruc$cumend=tstruc$end
  tstruc$altchr=tstruc$chr
  tstruc$altpos=tstruc$END
  tstruc$plot=F
  tstruc$arc= -1
  
  
  # Filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=tstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=tstruc$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  ## which keys are in reference?
  ix=key %in% btb_hg38_manta_pass$name # 11357. (why not all? this sample is in the reference.........)
  ix=btb_hg38_manta_pass$name %in% key # all but 800 ≥2 obs
  ix=key %in% btb_hg38_manta_all$name # 12559,
  ix=btb_hg38_manta_all$name %in% key #  all but 799 ≥2 obs
  # thus 1,5k reamins using "all" and 2,5k remains using "pass".
  # now using 2 observations as cutoff 
  ix <- key %in% btb_hg38_manta_all[value>1]$name
  ## do the filter
  tstruc <- tstruc[!ix]
  
  # loop through all and extract endpoint chr and pos
  for (i in 1:nrow(tstruc)) {
    t=strsplit(x = tstruc$ALT[[i]],split = ':')[[1]]
    if (length(t)>1 & t[1]!="<DUP") {
      tstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
      tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
      tstruc$altpos[i]=as.numeric(tt)
    }
    # also decide how it is to be plotted (not represented elsewhere, up or down arc)
    if (tstruc$chr[i]==tstruc$altchr[i]) {
      tstruc$plot[i]=T
      tstruc$arc[i]=1
    } else {
      tstruc$plot[i]=T
      try({
        if (tstruc[tstruc$MATEID[i][[1]],'plot']==T) tstruc$plot[i]=F # if mate not being plotted, plot this one
      }, silent=T)
    }
  }
  # # for each chr get cumulative pos
  tstruc$altcumpos=tstruc$altpos
  for (i in 1:nrow(chrsz)) {
    ix=tstruc$chr==chrsz$chr[i]
    tstruc$cumstart[ix]=tstruc$start[ix]+chrsz$starts[i]
    tstruc$cumend[ix]=tstruc$end[ix]+chrsz$starts[i]
    ix=tstruc$altchr==chrsz$chr[i]
    tstruc$altcumpos[ix]=tstruc$altpos[ix]+chrsz$starts[i]
  }
  

  

  # annotation is in the CSQ column.
  h=strsplit(info(header(tstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(h,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = length(unlist(tstruc$CSQ)),ncol = length(vep_header)+1)
  colnames(annotation_table)=c('N',vep_header)
  row=1
  for (i in 1:nrow(tstruc)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tstruc$CSQ[[i]])) { 
      line=strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[row,1]=i
      annotation_table[row,1+(1:length(line))]=line
      row=row+1
    }
  }
  annotation_table=as.data.table(annotation_table)
  
  # Add to data
  tstruc$N=as.character(1:nrow(tstruc))
  manta_tumor_table <- merge(tstruc,annotation_table,by='N')[order(cumstart),-1]
  #manta_tumor_table <- manta_tumor_table[,-c('Feature','EXON','Protein_position','BIOTYPE')]
  
  # Deletion: consider only deletions <3MB for "transcript ablation" consequence, report census gene(s)
  del_ok <- which(manta_tumor_table$SVTYPE=='DEL' & abs(manta_tumor_table$SVLEN)<3e6)
  # Amplification: consider <5M for amplification consequence, report census genes
  amp_ok <- which(manta_tumor_table$SVTYPE=='DUP' & abs(manta_tumor_table$SVLEN)<5e6)
  # Fusions/disrupting: Do the SVs start/end at gene? 
  manta_tumor_table$Starts_Near <- ''
  manta_tumor_table$Ends_Near <- ''
  manta_tumor_table$Junction <- ''
  cumstart=allgenes$cumstart
  cumend=allgenes$cumend
  for (i in 1:nrow(manta_tumor_table)) {
      i=as.integer(i)
      #cat(i,'\n')
      g=manta_tumor_table$SYMBOL[i]
      # genes overlapping SV start
      t=manta_tumor_table$cumstart[i]
      ix=cumstart-10e3 < t & cumend +10e3 > t
      if (length(ix)>0) {
        genes=allgenes[ix,unique(name)]
        set(manta_tumor_table,i,'Starts_Near',value=paste(genes,collapse = ' '))
        if (g %in% genes) set(manta_tumor_table,i,'Junction',value='Y')
      }
      # genes overlapping SV end
      t=manta_tumor_table$altcumpos[i]
      ix=cumstart-10e3 < t & cumend +10e3 > t
      if (length(ix)>0) {
        genes=allgenes[ix,unique(name)]
        set(manta_tumor_table,i,'Ends_Near',value=paste(genes,collapse = ' '))
        if (g %in% genes) set(manta_tumor_table,i,'Junction',value='Y')
      }
  }
  # Inside genes: Junctions may be putative fusion or knock-out if either end is "this" gene.
  afflicted_ok <- which(manta_tumor_table$Junction=='Y')
  
  ## Make new table with predicted effects
  selection <- unique(
    manta_tumor_table[unique(sort(c(del_ok, amp_ok, afflicted_ok))),
                                 .(SVTYPE,ID=THISID,chr,start,end,SVLEN=unlist(SVLEN),
                                   REF,altpos,ALT=unlist(ALT),SYMBOL,Consequence,Starts_Near,Ends_Near)]
  )
  selection$Gene_Tier <- 3
  selection$Effect_Tier <- 3
  selection$Comment=''
  selection$LOH <- ''
  
  # Set gene tier to 1,2 based on Cosmic
  selection[SYMBOL %in% tumorgenes$`Gene Symbol` | SYMBOL %in% local_tumorgenes$V1]$Gene_Tier <- 2
  selection[SYMBOL %in% tumorgenes[Tier==1]$`Gene Symbol`]$Gene_Tier <- 1
  selection <- selection[order(Gene_Tier)]
  
  # Set Effect tiers: 
  for (i in selection[Gene_Tier<3,.I]) {
    symbol=selection$SYMBOL[i]
    starts_at=selection$Starts_Near[i]
    ends_at=selection$Ends_Near[i]
    length=abs(selection$SVLEN[i])
    gene=tumorgenes[`Gene Symbol`==symbol,] # this gene from Cosmic table
    # first fusions/translocations
    g=grep('fusion',gene$`Role in Cancer`)
    if (length(g)>0) { # fusion gene yes
      if (symbol %in% c(starts_at,ends_at)) { # and the start or end is at this (SYMBOL) gene
        # now 1 if perfect match with Cosmic, 2 if partial match
        mates=strsplit(gene$`Translocation Partner`,', ')[[1]] # this gene's common tx mates
        if (starts_at != ends_at) { # if this symbol is either start or end, not both
          if (starts_at %in% mates | ends_at %in% mates) { # and either is a member of known fusion mates
            selection$Effect_Tier[i]=1 # common translocation/fusion
            selection$Comment[i]=paste(selection$Comment[i],'known_fusion')
          } else {
            selection$Effect_Tier[i]=2 # then at least tier 2
            selection$Comment[i]=paste0(selection$Comment[i],'partial_known_fusion')
          }
        }
      }
    }
    # then deletions
    if (selection$SVTYPE[i]=='DEL' & length<3e6) {
      g=grep('TSG',gene$`Role in Cancer`)
      if (length(g)>0) { # Tumor supressor gene yes
        selection$Effect_Tier[i]=1
        selection$Comment[i]=paste(selection$Comment[i],'TSG_del')
      } else { # else 2 unless already one
        selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
        selection$Comment[i]=paste(selection$Comment[i],'non_TSG_del')
      }
    }
    # then amplifications
    if (selection$SVTYPE[i]=='DUP' & length<5e6) {
      g=grep('oncogene',gene$`Role in Cancer`)
      if (length(g)>0) { # Oncogene yes
        selection$Effect_Tier[i]=1
        selection$Comment[i]=paste(selection$Comment[i],'oncogene_dup')
      } else {
        selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
        selection$Comment[i]=paste(selection$Comment[i],'non_oncogene_dup')
      }
    }
    # last other disruptions
    if (symbol %in% c(starts_at,ends_at)) { # if the start or end is at this (SYMBOL) gene
      selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2) # some indication of effect!
      selection$Comment[i]=paste(selection$Comment[i],'at_breakpoint')
    }
  }
  
  # Add ASCAT LOH.
  for (i in 1:nrow(selection)) {
    ix=ascat_cnv$chr==selection$chr[i] &
      ascat_cnv$startpos<selection$start[i] &
      ascat_cnv$endpos>selection$end[i]
    if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
  }
  selection <- selection[order(Effect_Tier,Gene_Tier)]

  manta_tumor_selected <- selection
  
}
```

<!-- Read the normal structural variant data -->
### Parse Manta (normal)
```{r, echo=FALSE}
if (length(vep_Nstruct_file)>0) {
  nstruc_vcf=readVcf(file = vep_Nstruct_file,genome = 'GRCh38')
  g=geno(nstruc_vcf)
  inf=info(nstruc_vcf)
  rr=rowRanges(nstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  nstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))[FILTER=='PASS' & IMPRECISE==FALSE]
  nstruc$SVLEN=listfix(nstruc$SVLEN)
  setkey(x = nstruc,THISID)
  nstruc$chr <- substr(x = nstruc$seqnames,start = 4,stop = 6)
  nstruc$cumstart=nstruc$start
  nstruc$cumend=nstruc$end
  nstruc$altchr=nstruc$chr
  nstruc$altpos=nstruc$END
  nstruc$plot=F
  nstruc$arc= -1
  
  # Filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=nstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=nstruc$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  ## which keys are in reference?
  ix=key %in% btb_hg38_manta_pass$name # 11357. (why not all? this sample is in the reference.........)
  ix=btb_hg38_manta_pass$name %in% key # all but 800 ≥2 obs
  ix=key %in% btb_hg38_manta_all$name # 12559,
  ix=btb_hg38_manta_all$name %in% key #  all but 799 ≥2 obs
  # thus 1,5k reamins using "all" and 2,5k remains using "pass".
  # now using 2 observations as cutoff 
  ix <- key %in% btb_hg38_manta_all[value>1]$name
  ## do the filter
  nstruc <- nstruc[!ix]
  
  # loop through all and extract endpoint chr and pos
  for (i in 1:nrow(nstruc)) {
    t=strsplit(x = nstruc$ALT[[i]],split = ':')[[1]]
    if (length(t)>1 & t[1]!="<DUP") {
      nstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
      tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
      nstruc$altpos[i]=as.numeric(tt)
    }
    # also decide how it is to be plotted (not represented elsewhere, up or down arc)
    if (nstruc$chr[i]==nstruc$altchr[i]) {
      nstruc$plot[i]=T
      nstruc$arc[i]=1
    } else {
      nstruc$plot[i]=T
      try({
        if (nstruc[nstruc$MATEID[i][[1]],'plot']==T) nstruc$plot[i]=F # if mate not being plotted, plot this one
      }, silent=T)
    }
  }
  # # for each chr get cumulative pos
  nstruc$altcumpos=nstruc$altpos
  for (i in 1:nrow(chrsz)) {
    ix=nstruc$chr==chrsz$chr[i]
    nstruc$cumstart[ix]=nstruc$start[ix]+chrsz$starts[i]
    nstruc$cumend[ix]=nstruc$end[ix]+chrsz$starts[i]
    ix=nstruc$altchr==chrsz$chr[i]
    nstruc$altcumpos[ix]=nstruc$altpos[ix]+chrsz$starts[i]
  }
  

  
  # annotation is in the CSQ column.
  h=strsplit(info(header(nstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(h,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = length(unlist(nstruc$CSQ)),ncol = length(vep_header)+1)
  colnames(annotation_table)=c('N',vep_header)
  row=1
  for (i in 1:nrow(nstruc)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(nstruc$CSQ[[i]])) { 
      line=strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[row,1]=i
      annotation_table[row,1+(1:length(line))]=line
      row=row+1
    }
  }
  annotation_table=as.data.table(annotation_table)
  
  # Add to data
  nstruc$N=as.character(1:nrow(nstruc))
  manta_normal_table <- merge(nstruc,annotation_table,by='N')[order(cumstart),-1]

  # Deletion: consider only deletions <3MB for "transcript ablation" consequence, report census gene(s)
  del_ok <- which(manta_normal_table$SVTYPE=='DEL' & abs(manta_normal_table$SVLEN)<3e6)
  # Amplification: consider <5M for amplification consequence, report census genes
  amp_ok <- which(manta_normal_table$SVTYPE=='DUP' & abs(manta_normal_table$SVLEN)<5e6)
  # Fusions/disrupting: Do the SVs start/end at gene? 
  manta_normal_table$Starts_Near <- ''
  manta_normal_table$Ends_Near <- ''
  manta_normal_table$Junction <- ''
  cumstart=allgenes$cumstart
  cumend=allgenes$cumend
  for (i in 1:nrow(manta_normal_table)) {
      i=as.integer(i)
      #cat(i,'\n')
      g=manta_normal_table$SYMBOL[i]
      # genes overlapping SV start
      t=manta_normal_table$cumstart[i]
      ix=cumstart-10e3 < t & cumend +10e3 > t
      if (length(ix)>0) {
        genes=allgenes[ix,unique(name)]
        set(manta_normal_table,i,'Starts_Near',value=paste(genes,collapse = ' '))
        if (g %in% genes) set(manta_normal_table,i,'Junction',value='Y')
      }
      # genes overlapping SV end
      t=manta_normal_table$altcumpos[i]
      ix=cumstart-10e3 < t & cumend +10e3 > t
      if (length(ix)>0) {
        genes=allgenes[ix,unique(name)]
        set(manta_normal_table,i,'Ends_Near',value=paste(genes,collapse = ' '))
        if (g %in% genes) set(manta_normal_table,i,'Junction',value='Y')
      }
  }
  # Inside genes: Junctions may be putative fusion or knock-out if either end is "this" gene.
  afflicted_ok <- which(manta_normal_table$Junction=='Y')
  
  ## Make new table with predicted effects
  selection <- unique(
    manta_normal_table[unique(sort(c(del_ok, amp_ok, afflicted_ok))),
                                 .(SVTYPE,ID=THISID,chr,start,end,SVLEN=unlist(SVLEN),
                                   REF,altpos,ALT=unlist(ALT),SYMBOL,Consequence,Starts_Near,Ends_Near)]
  )
  selection$Gene_Tier <- 3
  selection$Effect_Tier <- 3
  selection$Comment=''
  selection$LOH <- ''
  
  # Set gene tier to 1,2 based on Cosmic
  selection[SYMBOL %in% tumorgenes$`Gene Symbol` | SYMBOL %in% local_tumorgenes$V1]$Gene_Tier <- 2
  selection[SYMBOL %in% tumorgenes[Tier==1]$`Gene Symbol`]$Gene_Tier <- 1
  selection <- selection[order(Gene_Tier)]
  
  # Set Effect tiers: 
  for (i in selection[Gene_Tier<3,.I]) {
    symbol=selection$SYMBOL[i]
    starts_at=selection$Starts_Near[i]
    ends_at=selection$Ends_Near[i]
    length=abs(selection$SVLEN[i])
    gene=tumorgenes[`Gene Symbol`==symbol,] # this gene from Cosmic table
    # first fusions/translocations
    g=grep('fusion',gene$`Role in Cancer`)
    if (length(g)>0) { # fusion gene yes
      if (symbol %in% c(starts_at,ends_at)) { # and the start or end is at this (SYMBOL) gene
        # now 1 if perfect match with Cosmic, 2 if partial match
        mates=strsplit(gene$`Translocation Partner`,', ')[[1]] # this gene's common tx mates
        if (starts_at != ends_at) { # if this symbol is either start or end, not both
          if (starts_at %in% mates | ends_at %in% mates) { # and either is a member of known fusion mates
            selection$Effect_Tier[i]=1 # common translocation/fusion
            selection$Comment[i]=paste(selection$Comment[i],'known_fusion')
          } else {
            selection$Effect_Tier[i]=2 # then at least tier 2
            selection$Comment[i]=paste0(selection$Comment[i],'partial_known_fusion')
          }
        }
      }
    }
    # then deletions
    if (selection$SVTYPE[i]=='DEL' & length<3e6) {
      g=grep('TSG',gene$`Role in Cancer`)
      if (length(g)>0) { # Tumor supressor gene yes
        selection$Effect_Tier[i]=1
        selection$Comment[i]=paste(selection$Comment[i],'TSG_del')
      } else { # else 2 unless already one
        selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
        selection$Comment[i]=paste(selection$Comment[i],'non_TSG_del')
      }
    }
    # then amplifications
    if (selection$SVTYPE[i]=='DUP' & length<5e6) {
      g=grep('oncogene',gene$`Role in Cancer`)
      if (length(g)>0) { # Oncogene yes
        selection$Effect_Tier[i]=1
        selection$Comment[i]=paste(selection$Comment[i],'oncogene_dup')
      } else {
        selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
        selection$Comment[i]=paste(selection$Comment[i],'non_oncogene_dup')
      }
    }
    # last other disruptions
    if (symbol %in% c(starts_at,ends_at)) { # if the start or end is at this (SYMBOL) gene
      selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2) # some indication of effect!
      selection$Comment[i]=paste(selection$Comment[i],'at_breakpoint')
    }
  }
  
  # Add ASCAT LOH.
  for (i in 1:nrow(selection)) {
    ix=ascat_cnv$chr==selection$chr[i] &
      ascat_cnv$startpos<selection$start[i] &
      ascat_cnv$endpos>selection$end[i]
    if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
  }
  selection <- selection[order(Effect_Tier,Gene_Tier)]
  
  manta_normal_selected <- selection
  
}


```

&nbsp;


<!-- Plot the tumor and normal (also showing copy number) structural variant data -->
### View structural variants {.active}
```{r, echo=FALSE,fig.width=15,fig.height=3}
cat(vep_Tstruct_file)
if (exists('tstruc')) {
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Somatic structural variants')+expand_limits(x=c(0,3210e6))+ylim(-2,2)
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=-2,yend=2),data=chrsz,inherit.aes = F,alpha=1/5)
  # add copy data if present
  if (exists('tratio')) {
    g=g+geom_point(aes(x=cumstart,y=log2(Ratio)),
                   data = tratio,alpha=1/5,shape='.',
                   stat="identity",colour="grey")
  }
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.7,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  #g=g+geom_point(aes(x = cumstart,y = 1,col=SVTYPE),data=tstruc[tstruc$FILTER=='PASS',])
  g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=0.999-1,yend=1-1,col=SVTYPE,linetype=IMPRECISE),
                 data = tstruc[tstruc$FILTER=='PASS' & tstruc$plot,],alpha=1/2,
                 stat = "identity",curvature=-0.2,size=1,
                 position = "identity", angle = 90, ncp = 50,
                 arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                 inherit.aes = FALSE)+
    scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
    scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
  g
}
```

### SV table (tumor)

### SV table (normal)



## Somatic point mutations {.active .tabset}

### Parse Mutect 2
<!-- Read annotated small variant data (Mutect 2) -->
```{r, echo=FALSE}
if (length(mutect2_file)>0) {
  vcf=readVcf(file = mutect2_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  }
  tad=t(as.data.frame((g$AD[,'TUMOR'])))   #<------- the fix so that T/N pos are not hard coded
  tsum=tad[,1]+tad[,2]
  nad=t(as.data.frame((g$AD[,'NORMAL'])))
  nsum=nad[,1]+nad[,2]
  tmuts=cbind(tmuts,g$AF)
  tmuts$AD=tad[,2]
  tmuts$DP=tsum
  tmuts$ADn=nad[,2]
  tmuts$DPn=nsum
  tmuts$reads=' '
  tmuts$reads[tmuts$AD<5]='<5'
  tmuts$reads[tmuts$AD>=5]='≥5'
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  
    # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
     ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  key1_refAF[is.na(value),1]=0
  tmuts$allele1_swegen_count=key1_refAF$value
  remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  tmuts <- tmuts[!remove,] # filtered.

  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  mutect2_table=cbind(tmuts,annotation_table)
  
  # Keep high/mod-relevance in final output
  impact=which(mutect2_table$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',mutect2_table$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',mutect2_table$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',mutect2_table$SIFT)

  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious)))
  selection <- mutect2_table[
    keep,
    .(VARIANT_CLASS,chr,start,end,SYMBOL,Protein_position,Amino_acids,LOH='',Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
      ref,alt,DP,AD,DPn,ADn,allele1_swegen_count,Gene_Tier=3,Effect_Tier=3)
    ]
  if (nrow(selection)>0) {
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    
    # Genes in census or local list considered T2, census Tier1 then gets T1
    selection$Gene_Tier[selection$SYMBOL %in% local_tumorgenes$V1 | selection$SYMBOL %in%  tumorgenes$`Gene Symbol`] <- 2
    selection$Gene_Tier[selection$SYMBOL %in%  tumorgenes[Tier==1,`Gene Symbol`]] <- 1
    
    # Mutation effect: T2 if HIGH impact or (gene tier 1). T1 for MSK hotspot, or HIGH-impact in TSG, or Clinvar pathogenic
    selection$Effect_Tier[selection$IMPACT=='HIGH' | selection$Gene_Tier==1] <- 2
    temp=paste(selection$SYMBOL,selection$Protein_position)
    is_hotspot_pos <- 
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    selection[is_hotspot_pos | (IMPACT=='HIGH' &
                                  selection$SYMBOL %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`])]$Effect_Tier <- 1
    selection[grep('pathogenic',CLIN_SIG)]$Effect_Tier=1
    
    # Add ASCAT LOH.
    for (i in 1:nrow(selection)) {
      ix=ascat_cnv$chr==selection$chr[i] &
        ascat_cnv$startpos<selection$start[i] &
        ascat_cnv$endpos>selection$end[i]
      if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
    }
  }
  mutect2_selected <- selection
}
```

### Parse Strelka
<!-- Read annotated small variant data (Strelka, two files) -->
```{r, echo=FALSE}
if (length(strelka_snv_file) * length(strelka_indel_file)>0) {
  vcf=readVcf(file = strelka_snv_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
    # ix=tmuts$altchr==chrsz$chr[i]
    # tmuts$altcumpos[ix]=tmuts$altpos[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
   ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  key1_refAF[is.na(value),1]=0
  tmuts$allele1_swegen_count=key1_refAF$value
  remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  tmuts <- tmuts[!remove,] # filtered.

  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_snvs=cbind(tmuts,annotation_table)

  
    # Keep high/mod-relevance in final output
  impact=which(strelka_snvs$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',strelka_snvs$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',strelka_snvs$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',strelka_snvs$SIFT)

  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious)))
  selection <- strelka_snvs[
    keep,
    .(VARIANT_CLASS,chr,start,end,SYMBOL,Protein_position,Amino_acids,LOH='',Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
      ref,alt,DP=DPt,DPn,allele1_swegen_count,Gene_Tier=3,Effect_Tier=3)
    ]
  if (nrow(selection)>0) {
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    
    # Genes in census or local list considered T2, census Tier1 then gets T1
    selection$Gene_Tier[selection$SYMBOL %in% local_tumorgenes$V1 | selection$SYMBOL %in%  tumorgenes$`Gene Symbol`] <- 2
    selection$Gene_Tier[selection$SYMBOL %in%  tumorgenes[Tier==1,`Gene Symbol`]] <- 1
    
    # Mutation effect: T2 if HIGH impact or (gene tier 1). T1 for MSK hotspot, or HIGH-impact in TSG, or Clinvar pathogenic
    selection$Effect_Tier[selection$IMPACT=='HIGH' | selection$Gene_Tier==1] <- 2
    temp=paste(selection$SYMBOL,selection$Protein_position)
    is_hotspot_pos <- 
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    selection[is_hotspot_pos | (IMPACT=='HIGH' &
                                  selection$SYMBOL %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`])]$Effect_Tier <- 1
    selection[grep('pathogenic',CLIN_SIG)]$Effect_Tier=1
    
    # Add ASCAT LOH.
    for (i in 1:nrow(selection)) {
      ix=ascat_cnv$chr==selection$chr[i] &
        ascat_cnv$startpos<selection$start[i] &
        ascat_cnv$endpos>selection$end[i]
      if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
    }
  }
  strelka_selected <- selection
  
  # indels are read and added below.
  
  
  ## Read Strelka indels
  vcf=readVcf(file = strelka_indel_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
   ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  key1_refAF[is.na(value),1]=0
  tmuts$allele1_swegen_count=key1_refAF$value
  remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  tmuts <- tmuts[!remove,] # filtered.
  
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_indels=cbind(tmuts,annotation_table)
  
   
  # Keep high/mod-relevance in final output
  impact=which(strelka_indels$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',strelka_indels$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',strelka_indels$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',strelka_indels$SIFT)

  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious)))
  selection <- strelka_snvs[
    keep,
    .(VARIANT_CLASS,chr,start,end,SYMBOL,Protein_position,Amino_acids,LOH='',Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
      ref,alt,DP=DPt,DPn,allele1_swegen_count,Gene_Tier=3,Effect_Tier=3)
    ]
  if (nrow(selection) > 0) {
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    
    # Genes in census or local list considered T2, census Tier1 then gets T1
    selection$Gene_Tier[selection$SYMBOL %in% local_tumorgenes$V1 | selection$SYMBOL %in%  tumorgenes$`Gene Symbol`] <- 2
    selection$Gene_Tier[selection$SYMBOL %in%  tumorgenes[Tier==1,`Gene Symbol`]] <- 1
    
    # Mutation effect: T2 if HIGH impact or (gene tier 1). T1 for MSK hotspot, or HIGH-impact in TSG, or Clinvar pathogenic
    selection$Effect_Tier[selection$IMPACT=='HIGH' | selection$Gene_Tier==1] <- 2
    temp=paste(selection$SYMBOL,selection$Protein_position)
    is_hotspot_pos <- 
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    selection[is_hotspot_pos | (IMPACT=='HIGH' &
                                  selection$SYMBOL %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`])]$Effect_Tier <- 1
    selection[grep('pathogenic',CLIN_SIG)]$Effect_Tier=1
    
    # Add ASCAT LOH.
    for (i in 1:nrow(selection)) {
      ix=ascat_cnv$chr==selection$chr[i] &
        ascat_cnv$startpos<selection$start[i] &
        ascat_cnv$endpos>selection$end[i]
      if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
    }
  }
  strelka_selected <- rbind(strelka_selected,selection)
}
```


<!-- Plot Mutect 2 mutations along the genome -->
### View somatic point mutations {.active}

#### Genome view
```{r, echo=FALSE,fig.width=15,fig.height=6}
cat(mutect2_file)
if (exists('mutect2_table')) {
  par(mai=c(1,4,0,2))
  g=ggplot(data = mutect2_table[mutect2_table$filter=='PASS',],
           aes(x=cumstart,y=TUMOR,fill=type,shape=type,col=reads))+
    geom_point(alpha=1)+
    ylab('Somatic mutation allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
  # add smoothed
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  # add gene labels
  temp=mutect2_table[mutect2_table$filter=='PASS' & mutect2_table$Amino_acids!='',]
  g=g+geom_label_repel(data = temp,
                       mapping = aes(x=cumstart,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),
                       inherit.aes = F,size=2,nudge_y = 0.2)


  g
  # m=mutect2_table[,c('chr','start','end','ref','alt','AD','DP','ADn','DPn','Consequence','IMPACT','SYMBOL',
  #                    'BIOTYPE','EXON','Existing_variation','Protein_position','Amino_acids','Existing_variation')]
  # knitr::kable(m)
}
```

&nbsp;

#### Coverage - allele frequency
<!-- Make a scatter plot of Mutect2 findings, showing which are also in Strelka -->
```{r, echo=FALSE,,fig.width=6,fig.height=4,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {
  
  # First make a merged table of findings
  mutect2_temp=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                     MARGIN = 1,FUN = paste,collapse=' ')
  strelka_temp=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                               strelka_indels[,c('chr','start','end','ref','alt'),]),
                     MARGIN = 1,FUN = paste,collapse=' ')
  temp <- mutect2_table
  temp$Callers <- 'Mutect 2'
  temp$Callers[mutect2_temp %in% strelka_temp]='Mutect 2 and Strelka'
  
  g <- ggplot()+geom_point(mapping = aes(x=DP,y=AD/DP,fill=type,shape=type,col=Callers),data=temp)+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('Mutect 2'='lightgrey','Mutect 2 and Strelka'='black'))+
    ylab('ALT allele ratio') + xlab('Local coverage')+
    scale_x_log10()+
    geom_label_repel(data = temp[temp$Amino_acids!='',],mapping = aes(x=DP,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),inherit.aes = F,size=2,nudge_y = 0,nudge_x = 0)
  g#ggplotly(g)  
}
```

&nbsp;

#### Venn diagram
<!-- Make a Venn diagram of Strelka and Mutect2 findings -->
```{r, echo=FALSE,,fig.width=6,fig.height=4,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {

  VENN.LIST=list(mutect2=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                               MARGIN = 1,FUN = paste,collapse=' '),
                 strelka=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                                         strelka_indels[,c('chr','start','end','ref','alt'),]),
                               MARGIN = 1,FUN = paste,collapse=' '))
  venn.plot <- venn.diagram(VENN.LIST , NULL, fill=c("darkmagenta", "darkblue"), 
                            alpha=c(0.2,0.2), cex = 2, cat.fontface=4, 
                            category.names=c("Mutect 2", "Strelka"))
  grid.draw(venn.plot)
}
```


### Mutect 2 table

### Strelka table


## Germline point mutations {.tabset}

### Parse HaplotypeCaller (normal)
<!-- Read annotated small variant data (Haplotype caller, normal) -->
```{r, echo=FALSE}
if (length(haplotypecaller_N_file)>0) {
  vcf=readVcf(file = haplotypecaller_N_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data table with relevant data
  nmuts=as.data.table(ranges(rowRanges(vcf)))
  nmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  nmuts$cumstart=nmuts$start
  nmuts$cumend=nmuts$end
  nmuts$ref=as.character(ref(vcf))
  nmuts$alt=''
  nmuts$alt2=''
  temp=as.data.table(alt(vcf))[, .(values = list(value)), by = group]
  ix=unlist(lapply(temp$values,FUN = length))==2
  nmuts$alt[!ix]=unlist(temp$values[!ix])
  t=unlist(temp$values[ix])
  nmuts$alt[ix]=t[seq(1,length(t),2)]
  nmuts$alt2[ix]=t[seq(2,length(t),2)]
 
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=nmuts$chr==chrsz$chr[i]
    nmuts$cumstart[ix]=nmuts$start[ix]+chrsz$starts[i]
    nmuts$cumend[ix]=nmuts$end[ix]+chrsz$starts[i]
  }
  nmuts$DP=g$DP[,1]
  ad=matrix(0,nrow = nrow(nmuts),ncol = 3,dimnames=list(NULL,c('RD','AD','AD2')))
  n=unlist(lapply(g$AD,length)) # gets number of AD per variant
  ad[n==2,1:2]=matrix(data=unlist(g$AD[n==2]),ncol=2,byrow = T)
  ad[n==3,1:3]=matrix(data=unlist(g$AD[n==3]),ncol=3,byrow = T)
  nmuts=cbind(nmuts,ad)
  
  ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  key1 <- paste(nmuts$chr,nmuts$start,nmuts$alt)
  key2 <- paste(nmuts$chr,nmuts$start,nmuts$alt2)
  has2 <- nmuts$alt2!=''
  key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  key1_refAF[is.na(value),1]=0
  key2_refAF <- swegen_hg38_allele_counts[key2,.(value)]
  key2_refAF[is.na(value),1]=0
  nmuts$allele1_swegen_count=key1_refAF$value
  nmuts$allele2_swegen_count=key2_refAF$value
  remove <- key1_refAF$value>=2 & (key2_refAF$value>=2 | !has2) #remove SNP if allele1 is ≥2 and (allele2 too or no second allele)
  nmuts <- nmuts[!remove,] # filtered.
  
  # "info"" annotations also need to be added
  nmuts=cbind(nmuts,as.data.table(info(vcf)[!remove,]))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(nmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(nmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(nmuts$CSQ[[i]])) { 
      line=strsplit(nmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  haplotypecaller_table=cbind(nmuts,annotation_table)
  
  # Keep high/mod-relevance in final output
  impact=which(haplotypecaller_table$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',haplotypecaller_table$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',haplotypecaller_table$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',haplotypecaller_table$SIFT)

  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious)))
  selection <- haplotypecaller_table[
    keep,
    .(VARIANT_CLASS,chr,start,end,SYMBOL,Protein_position,Amino_acids,LOH='',Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
      ref,alt,alt2,DP,RD,AD,AD2,allele1_swegen_count,allele2_swegen_count)
    ]
  for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
  
  selection$Gene_Tier <- 3
  selection$Effect_Tier <- 3
  
  # Genes in census or local list considered T2, census Tier1 then gets T1
  selection$Gene_Tier[selection$SYMBOL %in% local_tumorgenes$V1 | selection$SYMBOL %in%  tumorgenes$`Gene Symbol`] <- 2
  selection$Gene_Tier[selection$SYMBOL %in%  tumorgenes[Tier==1,`Gene Symbol`]] <- 1
    
  # Mutation effect: T2 if HIGH impact or (gene tier 1). T1 for MSK hotspot, or HIGH-impact in TSG, or Clinvar pathogenic
  selection$Effect_Tier[selection$IMPACT=='HIGH' | selection$Gene_Tier==1] <- 2
  temp=paste(selection$SYMBOL,selection$Protein_position)
  is_hotspot_pos <- 
    temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
    temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
  selection[is_hotspot_pos | (IMPACT=='HIGH' &
                                selection$SYMBOL %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`])]$Effect_Tier <- 1
  selection[grep('pathogenic',CLIN_SIG)]$Effect_Tier=1
  selection <- selection[order(Effect_Tier,Gene_Tier)]
  
  # Add ASCAT LOH.
  for (i in 1:nrow(selection)) {
    ix=ascat_cnv$chr==selection$chr[i] &
      ascat_cnv$startpos<selection$start[i] &
      ascat_cnv$endpos>selection$end[i]
    if (sum(ix) >0) if (ascat_cnv[ix,LOH]=='LOH') selection$LOH[i]='Y'
  }
  
  haplotypecaller_selected <- selection
  
}
```

### View germline point mutations {.active}

### HaplotypeCaller table

<!-- ## Combined view? like chromwise plot but all genome. -->


## View by chromosome {.tabset}
<!-- Plot composite data, chromosome wise -->
```{r}
chromplot=function(chr) {
  par(mai=c(0,4,0,2))
  g=ggplot()+
    scale_y_continuous(limits=c(-2.5,3),
                       breaks = seq(1.5,-2.5,-0.5),
                       minor_breaks = seq(1.25,-2.25,-0.5),
                       labels = 
                         c('Tumor DNA=1.5','Tumor Copy Ratio\n(median centered)','Tumor DNA=0.5\nMutation AF=1','Mutation AF=0.5','Mutation AF=0\nTumor SNP AF=1',
                           'Tumor\nSNPs','Tumor SNP AF=0\nNormal SNP AF=1','Normal\nSNPs','Normal SNP AF=0'))+
    scale_x_continuous(
      breaks = 10e6*seq(0,ceiling(chrsz$length[chrsz$chr==chr]/10e6)),
      labels= c('0',paste0(10*seq(1,ceiling(chrsz$length[chrsz$chr==chr]/10e6)),'M')),
      minor_breaks = 1e6*seq(0,ceiling(chrsz$length[chrsz$chr==chr]/1e6)))+
    theme(panel.grid.minor.x = element_line(colour="lightgrey", size=0.05),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="black", size=0.2),
          panel.grid.minor.y = element_line(colour="lightgrey", size=0.05),
          axis.title.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.title.y=element_blank(),
          axis.ticks.y=element_blank())#+

  
  ### TUMOR LOGRATIO  
  if (exists('tratio')) {
    temp <- tratio
    temp$Ratio[temp$Ratio>3]=3
    g=g+geom_line(aes(x=Start,y=Ratio),
                   data = temp[temp$Chromosome==chr,],alpha=1/3,
                   stat="identity",colour="black")
  }
  else if (exists('ascat_tratio')) {
    temp <- ascat_tratio
    temp$smoothed[temp$smoothed>3]=3
    g=g+geom_line(aes(x=Position,y=smoothed),
                   data = temp[temp$Chr==chr,],alpha=1/3,
                   stat="identity",colour="black")
  }
  
  ### TUMOR STRVAR  
  if (exists('tstruc')) {
    temp=tstruc[tstruc$chr==chr & tstruc$FILTER=='PASS' & tstruc$altchr==tstruc$chr,]
    thischrstart=chrsz$starts[chrsz$chr==chr]
    if (nrow(temp)>0) g=g+geom_curve(aes(x = start,xend = altpos,y=0.999,yend=1,col=SVTYPE,linetype=IMPRECISE),
                                     data = temp,alpha=1/2,
                                     curvature=-0.2,size=1,
                                     position = "identity", angle = 90, ncp = 50,
                                     arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                     inherit.aes = FALSE)+
      scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
      scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
  }
  
  ### TUMOR MUTATIONS  
  if (exists('mutect2_table')) {
    temp=mutect2_table[mutect2_table$chr==chr & mutect2_table$filter=='PASS',]
    if (nrow(temp)>0) { 
      g=g+geom_point(aes(x=start,y=TUMOR -0.5,fill=type,shape=type),data = temp,alpha=1)+
        scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
        scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))#+
      #scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
      
      g=g+geom_label_repel(data = temp[temp$Amino_acids!='',],
                           mapping = aes(x=start,y=TUMOR -0.5,label=paste(SYMBOL,Protein_position,Amino_acids)),
                           inherit.aes = F,size=3,min.segment.length = 0,nudge_y = 0.3)
    }
  }
  
  ### SNPS
  if (exists('tbaf')) { 
    g=g+geom_point(aes(x=Position,y=BAF -1.5),data = tbaf[tbaf$Chromosome==chr,],alpha=1/8,shape='.') 
  } else if (exists('ascat_tbaf'))  {
    g=g+geom_point(aes(x=Position,y=BAF -1.5),data = ascat_tbaf[ascat_tbaf$Chr==chr,],alpha=1/8,shape='.')
  }
  if (exists('nbaf')) {
    g=g+geom_point(aes(x=Position,y=BAF -2.5),data = nbaf[nbaf$Chromosome==chr,],alpha=1/8,shape='.')
  } else if (exists('ascat_nbaf')) {
    g=g+geom_point(aes(x=Position,y=NBAF -1.5),data = ascat_nbaf[ascat_nbaf$Chr==chr,],alpha=1/8,shape='.')
  }
  return(g)
}
```

&nbsp;

### 1
```{r,fig.width=15,fig.height=10}
chr=1
cat('Chromosome',chr)
chromplot(chr)
```
### 2
```{r,fig.width=15,fig.height=10}
chr=2
cat('Chromosome',chr)
chromplot(chr)
```
### 3
```{r,fig.width=15,fig.height=10}
chr=3
cat('Chromosome',chr)
chromplot(chr)
```
### 4
```{r,fig.width=15,fig.height=10}
chr=4
cat('Chromosome',chr)
chromplot(chr)
```
### 5
```{r,fig.width=15,fig.height=10}
chr=5
cat('Chromosome',chr)
chromplot(chr)
```
### 6
```{r,fig.width=15,fig.height=10}
chr=6
cat('Chromosome',chr)
chromplot(chr)
```
### 7
```{r,fig.width=15,fig.height=10}
chr=7
cat('Chromosome',chr)
chromplot(chr)
```
### 8
```{r,fig.width=15,fig.height=10}
chr=8
cat('Chromosome',chr)
chromplot(chr)
```
### 9
```{r,fig.width=15,fig.height=10}
chr=9
cat('Chromosome',chr)
chromplot(chr)
```
### 10
```{r,fig.width=15,fig.height=10}
chr=10
cat('Chromosome',chr)
chromplot(chr)
```
### 11
```{r,fig.width=15,fig.height=10}
chr=11
cat('Chromosome',chr)
chromplot(chr)
```
### 12
```{r,fig.width=15,fig.height=10}
chr=12
cat('Chromosome',chr)
chromplot(chr)
```
### 13
```{r,fig.width=15,fig.height=10}
chr=13
cat('Chromosome',chr)
chromplot(chr)
```
### 14
```{r,fig.width=15,fig.height=10}
chr=14
cat('Chromosome',chr)
chromplot(chr)
```
### 15
```{r,fig.width=15,fig.height=10}
chr=15
cat('Chromosome',chr)
chromplot(chr)
```
### 16
```{r,fig.width=15,fig.height=10}
chr=16
cat('Chromosome',chr)
chromplot(chr)
```
### 17
```{r,fig.width=15,fig.height=10}
chr=17
cat('Chromosome',chr)
chromplot(chr)
```
### 18
```{r,fig.width=15,fig.height=10}
chr=18
cat('Chromosome',chr)
chromplot(chr)
```
### 19
```{r,fig.width=15,fig.height=10}
chr=19
cat('Chromosome',chr)
chromplot(chr)
```
### 20
```{r,fig.width=15,fig.height=10}
chr=20
cat('Chromosome',chr)
chromplot(chr)
```
### 21
```{r,fig.width=15,fig.height=10}
chr=21
cat('Chromosome',chr)
chromplot(chr)
```
### 22
```{r,fig.width=15,fig.height=10}
chr=22
cat('Chromosome',chr)
chromplot(chr)
```
### X
```{r,fig.width=15,fig.height=10}
chr='X'
cat('Chromosome',chr)
chromplot(chr)
```
### Y
```{r,fig.width=15,fig.height=10}
chr='Y'
cat('Chromosome',chr)
chromplot(chr)
```

&nbsp;

