---
output: 
  html_document:
      toc: true
      toc_float: true
      toc_depth: 4
editor_options: 
  chunk_output_type: console
---

<!-- Logo -->
<!-- ![](~/reports/CAW_logo.png) -->
 <!-- This script reads from a folder containing the CAW output, plus copy number data generated by controlfreec.-->

<!-- Markdown setup -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,comment = '',warning=F,error = F)
```

<!-- Dependencies and reference data -->
```{r,echo=F, include=F}
library(data.table)
library(readr)
library(dplyr)
library(tidyr)
library(tibble)
library(purrr)
library(ggplot2)
theme_set(theme_light())
library(ggrepel)
library(VariantAnnotation)
library(gridExtra)
library(VennDiagram)

load('~/reports/BTB_hg38_manta_counts.Rdata') # gets btb_hg38_manta_all/_pass
load('~/reports/swegen_hg38_allele_counts.Rdata') # gets swegen_hg38_allele_counts

chrsz <- data.table(
  chr = c("1", "2", "3", "4", "5", "6", "7", "8", 
          "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", 
          "20", "21", "22", "X", "Y"), 
  starts = c(0, 253902921, 500669562, 
             703689723, 898025604, 1084280925, 1259870526, 1424144247, 1574191848, 
             1717288929, 1855896810, 1995944331, 2134165212, 2253485013, 2363996694, 
             2470839615, 2565902256, 2654091777, 2739309138, 2802869979, 2871941700, 
             2923598421, 2979326382, 3140008743), 
  length = c(248902921, 241766641, 
             198020161, 189335881, 181255321, 170589601, 159273721, 145047601, 
             138097081, 133607881, 135047521, 133220881, 114319801, 105511681, 
             101842921, 90062641, 83189521, 80217361, 58560841, 64071721, 
             46656721, 50727961, 155682361, 56827081)
  )


```

#### Directory/Sample
<!-- First, make a list of files to read and visualize -->
```{r}
cat(date(),'\n')
cat('Current directory:', getwd(),'\n')

files <- dir(recursive = T,full.names = T)
files <- files[-grep(pattern = '.png',x = files)]
# control freec files
freec_Tratio_file <- files[grep(pattern = 'T.hg38.pileup.gz_ratio.txt',files)]
freec_Nratio_file <- files[grep(pattern = 'T.hg38.pileup.gz_normal_ratio.txt',files)]
freec_Tbaf_file <- files[grep(pattern = 'T.hg38.pileup.gz_BAF.txt',files)]
freec_Nbaf_file <- files[grep(pattern = 'B.hg38.pileup.gz_BAF.txt',files)]
freec_info_file <- files[grep(pattern = 'hg38.pileup.gz_info.txt',files)]
# Ascat files
ascat_Tratio_file <- files[grep(pattern = glob2rx("*Ascat*T.LogR"),files)]
ascat_Nratio_file <- files[grep(pattern = glob2rx("*Ascat*N.LogR"),files)]
ascat_Tbaf_file <- files[grep(pattern = glob2rx("*Ascat*T.BAF"),files)]
ascat_Nbaf_file <- files[grep(pattern = glob2rx("*Ascat*N.BAF"),files)]
ascat_segment_file <- files[grep(pattern = glob2rx("*Ascat*T.cnvs.txt"),files)]
# Manta structural variant files (VEP here)
vep_Tstruct_file <- grep(pattern = glob2rx("*VEP/Manta*somaticSV.*ann.vcf"),files,value = T)
vep_Nstruct_file <- grep(pattern = glob2rx("*VEP/Manta*diploidSV.*ann.vcf"),files,value = T)
# Haplotype caller variant files
haplotypecaller_T_file <- grep(pattern = glob2rx("*VEP/haplotypecaller*T.vcf.vep.ann.vcf"),files,value = T) # local
if (length(haplotypecaller_T_file)==0)
  haplotypecaller_T_file <- grep(pattern = glob2rx("*/Annotation/VEP/haplotypecaller_*T.vcf.vep.ann.vcf"),files,value = T) # bianca
haplotypecaller_N_file <- grep(pattern = glob2rx("*VEP/haplotypecaller*N.vcf.vep.ann.vcf"),files,value = T) # local
if (length(haplotypecaller_N_file)==0)
  haplotypecaller_N_file <- grep(pattern = glob2rx("*/Annotation/VEP/haplotypecaller_*N.vcf.vep.ann.vcf"),files,value = T) # bianca
# snpEff+VEP Mutect2 file
mutect2_file <- grep(pattern = glob2rx("*/Annotation/mutect2_*canon.vep.ann.pick.PASS.recode.vcf"),files,value = T) # <- local
if (length(mutect2_file)==0)
  mutect2_file <- grep(pattern = glob2rx("*Annotation/VEP/mutect2_*.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca
# snpEff + VEP Strelka snv file
strelka_snv_file <- grep(pattern = glob2rx("*/Annotation/Strelka_*somatic_snvs.snpeff.ann.canon.vep.ann.pick.vcf"),files,value = T) # <- local
if (length(strelka_snv_file)==0)
  strelka_snv_file <- grep(pattern = glob2rx("*Annotation/VEP/Strelka_*_somatic_snvs.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca
# snpEff + VEP Strelka indel file
strelka_indel_file <- grep(pattern = glob2rx("*/Annotation/Strelka_*somatic_indels.snpeff.ann.canon.vep.ann.pick.vcf"),files,value = T) # <- local
if (length(strelka_indel_file)==0)
  strelka_indel_file <- grep(pattern = glob2rx("*Annotation/VEP/Strelka_*_somatic_indels.vcf.vep.ann.PASS.recode.vcf"),files,value = T) # <- bianca
```






##### Read Control Freec
<!-- Read Control Freec copy number data -->
```{r}
if (length(freec_Tratio_file)>0 &
    length(freec_Nratio_file)>0 &
    length(freec_Tbaf_file)>0 &
    length(freec_Nbaf_file)>0) {
  # tumor log ratio
  tratio <- fread(file = freec_Tratio_file)
  tratio <- tratio[order(Chromosome,Start)]
  # normal log ratio
  nratio <- fread(file = freec_Nratio_file)
  nratio <- nratio[order(Chromosome,Start)]
  # tumor BAF
  tbaf <- fread(freec_Tbaf_file)[,1:3]
  # normal BAF
  nbaf <- fread(freec_Nbaf_file)[,1:3]
  # merged
  mbaf <- merge(tbaf,nbaf,by = c('Chromosome','Position'))

  # manipulate for plot:
  tratio$cumstart <- tratio$Start
  nratio$cumstart <- nratio$Start
  tbaf$cumstart <- tbaf$Position
  nbaf$cumstart <- nbaf$Position
  for (i in 2:nrow(chrsz)) {
    ix <- tratio$Chromosome==chrsz$chr[i]
    tratio$cumstart[ix] <- tratio$Start[ix]+chrsz$starts[i]
    ix <- nratio$Chromosome==chrsz$chr[i]
    nratio$cumstart[ix] <- nratio$Start[ix]+chrsz$starts[i]
    ix <- tbaf$Chromosome==chrsz$chr[i]
    tbaf$cumstart[ix] <- tbaf$Position[ix]+chrsz$starts[i]
    ix <- nbaf$Chromosome==chrsz$chr[i]
    nbaf$cumstart[ix] <- nbaf$Position[ix]+chrsz$starts[i]
  }
  
  # modify for plot
  tratio$CopyNumber[tratio$CopyNumber>4] <- 4
  nratio$CopyNumber[nratio$CopyNumber>4] <- 4
  tratio$CopyNumber[1:2] <- c(0,4)
  nratio$CopyNumber[1:2] <- c(0,4)
  tratio$BAF[tratio$BAF<0.5 | tratio$BAF>1] <- NA
  nratio$BAF[nratio$BAF<0.5 | nratio$BAF>1] <- NA
  
  # smooth data for plot
  binned <- NULL
  for (i in 1:nrow(chrsz)) {
    temp <- data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      nratio=NA,
      tmaf=NA,
      nmaf=NA)
    ctratio <- tratio[Chromosome==chrsz$chr[i]]
    cnratio <- nratio[Chromosome==chrsz$chr[i]]
    for (j in 1:nrow(temp)) {
      ix <- ctratio$Start>temp$pos[j]-5e6 & ctratio$Start<temp$pos[j]+5e6
      temp$tratio[j] <- median(ctratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) {
        d <- density(ctratio$BAF[ix],na.rm=T)
        temp$tmaf[j] <- d$x[which.max(d$y)]
      }
      ix <- cnratio$Start>temp$pos[j]-5e6 & cnratio$Start<temp$pos[j]+5e6
      temp$nratio[j] <- median(cnratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) try({
        d <- density(cnratio$BAF[ix],na.rm=T)
        temp$nmaf[j] <- d$x[which.max(d$y)]
      },silent=T)
    }
    binned <- rbind(binned,temp)
  }
}
```




#### Copy number alteration (Control Freec)
<!-- Plot Control Freec tumor copy number log ratio, tBAF and nBAF -->
```{r,fig.width=15,fig.height=10}
if (exists('tratio')) {
  cat('Control Freec: ',freec_Tratio_file)
  par(mai=c(0,4,0,2))
  temp <- tratio
  temp$Ratio[temp$Ratio>3]=3

  g=ggplot()+ylab('B allele ratio and Coverage Ratio')
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    ylim(-1,3)+
    #expand_limits(y=c(-1,1.6))+
    expand_limits(x=c(0,3200e6))
  # normal logR below
  g=g+geom_point(aes(x=cumstart,y=Ratio-0.3),col='darkgrey',data=nratio,alpha=1/5,shape='.')
  # tumor logR
  g=g+geom_point(aes(x=cumstart,y=Ratio,col=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(data = tbaf,aes(x=cumstart,y=-0.5+0.5*BAF),
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(data = nbaf,aes(x=cumstart,y=-1+0.5*BAF),
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)

  g
}
```


&nbsp;

##### Read ASCAT
<!-- Read ASCAT copy number data -->
```{r,echo=F,include=F}
if (length(ascat_Tratio_file)>0 &
    length(ascat_Nratio_file)>0 &
    length(ascat_Tbaf_file)>0 &
    length(ascat_Nbaf_file)>0 & 
    length(ascat_segment_file)>0) {
  # segmented copy number
  ascat_cnv=fread(file = ascat_segment_file)
  # tumor log ratio
  ascat_tratio=fread(file = ascat_Tratio_file)[,-1]
  colnames(ascat_tratio)[3]='LogR'
  ascat_tratio$Chr=substr(ascat_tratio$Chr,start = 4,stop = 8)
  ascat_tratio=ascat_tratio[Chr %in% c(1:22,'X','Y')]
  ascat_tratio=ascat_tratio[order(Chr,Position)]
  # tumor BAF
  ascat_tbaf=fread(ascat_Tbaf_file)[,-1]
  colnames(ascat_tbaf)[3]='BAF'
  ascat_tbaf=ascat_tbaf[which(BAF!=0 & BAF!=1)]
  ascat_tbaf$Chr=substr(ascat_tbaf$Chr,start = 4,stop = 8)
  ascat_tbaf=ascat_tbaf[Chr %in% c(1:22,'X','Y')]
  # normal BAF
  ascat_nbaf=fread(ascat_Nbaf_file)[,-1]
  colnames(ascat_nbaf)[3]='NBAF'
  ascat_nbaf=ascat_nbaf[which(NBAF!=0 & NBAF!=1)]
  ascat_nbaf$Chr=substr(ascat_nbaf$Chr,start = 4,stop = 8)
  ascat_nbaf=ascat_nbaf[Chr %in% c(1:22,'X','Y')]
  
  # join (T) log ratio and (T+N) BAF
  ascat_tratio=merge(ascat_tratio,ascat_tbaf,all.x=T)
  ascat_tratio=merge(ascat_tratio,ascat_nbaf,all.x=T)
  
  # manipulate for plot:
  ascat_tratio$cumstart=ascat_tratio$Position
  #ascat_nratio$cumstart=ascat_nratio$Position
  ascat_tbaf$cumstart=ascat_tbaf$Position
  ascat_nbaf$cumstart=ascat_nbaf$Position
  ascat_cnv$cumstart=ascat_cnv$startpos
  ascat_cnv$cumend=ascat_cnv$endpos
  ascat_cnv$LOH=''; ascat_cnv$LOH[ascat_cnv$nMinor==0]='LOH'
  for (i in 2:nrow(chrsz)) {
    ix=ascat_tratio$Chr==chrsz$chr[i]
    ascat_tratio$cumstart[ix]=ascat_tratio$Position[ix]+chrsz$starts[i]
    #no n...
    ix=ascat_tbaf$Chr==chrsz$chr[i]
    ascat_tbaf$cumstart[ix]=ascat_tbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_nbaf$Chr==chrsz$chr[i]
    ascat_nbaf$cumstart[ix]=ascat_nbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_cnv$chr==chrsz$chr[i]
    ascat_cnv$cumstart[ix]=ascat_cnv$startpos[ix]+chrsz$starts[i]
    ascat_cnv$cumend[ix]=ascat_cnv$endpos[ix]+chrsz$starts[i]
  }
  # copy number hack
  ascat_tratio$Ratio=2^ascat_tratio$LogR
  #ascat_tratio$Ratio[ascat_tratio$Ratio > 2]=2
  ascat_tratio$smoothed=runmed(x = ascat_tratio$Ratio,k = 99)
  ascat_tratio$CopyNumber=2
  ascat_tratio$MinorCopy=1
  for (i in 1:nrow(ascat_cnv)) {
    ix=ascat_tratio$cumstart > ascat_cnv$cumstart[i] & ascat_tratio$cumstart < ascat_cnv$cumend[i]
    ascat_tratio$CopyNumber[ix]=ascat_cnv$nMajor[i]+ascat_cnv$nMinor[i]
    ascat_tratio$MinorCopy[ix]=ascat_cnv$nMinor[i]
  }
  ascat_tratio$CopyNumber[ascat_tratio$CopyNumber>4]=4
  ascat_tratio$CopyNumber[1:2]=c(0,4)
  ascat_tratio$LOH='no'
  ascat_tratio$LOH[ascat_tratio$MinorCopy==0]='yes'
  ascat_tratio$Copies=as.character(ascat_tratio$CopyNumber)
  ascat_tratio$Copies[ascat_tratio$LOH=='yes' & ascat_tratio$CopyNumber>1]='LOH'
  ascat_tratio$Copies[ascat_tratio$CopyNumber==4]='≥4'
  ascat_tratio$Copies[1:6]=c('0','1','LOH','2','3','≥4')

  # smooth data for view
  ascat_binned=NULL
  for (i in 1:nrow(chrsz)) {
    temp=data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      tmaf=NA)
    
    cix=ascat_tratio$Chr==chrsz$chr[i]
    tpos=ascat_tratio$Position[cix]
    tLogR=ascat_tratio$LogR[cix]
    tBAF=0.5+abs(ascat_tratio$BAF[cix]-0.5)
    for (j in 1:nrow(temp)) {
      ix = tpos>temp$pos[j]-5e6 & tpos<temp$pos[j]+5e6
      temp$tratio[j]=2^median(tLogR[ix],na.rm = T)
      if (sum(!is.na(tBAF[ix]))>20) {
        d=density(tBAF[ix],na.rm=T)
        temp$tmaf[j]=d$x[which.max(d$y)]
      }
    }
    ascat_binned=rbind(ascat_binned,temp)
  }
}
```


#### Copy number alteration (ASCAT)
<!-- Plot ASCAT tumor copy number log ratio, tBAF and nBAF -->
```{r,fig.width=15,fig.height=10}
if (exists('ascat_tratio')) {
  cat('ASCAT: ',ascat_Tratio_file,ascat_segment_file,ascat_Tbaf_file,ascat_Nbaf_file)
  par(mai=c(0,4,0,2))
  temp <- ascat_tratio
  temp$smoothed[temp$smoothed>3]=3
  g=ggplot()
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    ylim(-1,3)+
    expand_limits(x=c(0,3200e6))
  # using smoothed logR
  g=g+geom_point(aes(x=cumstart,y=smoothed,color=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=ascat_binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-0.5+0.5*BAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-1+0.5*NBAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)

  g
}
```


&nbsp;

#### Copy number scatter plot (Control Freec)
```{r, echo=FALSE,,fig.width=8,fig.height=4,warning=F}
if (exists('binned')) {
  cat('Control freec:')
  binned$tmaf[binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+#xlim(0.2,1.8)+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=binned[binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,1.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```

#### Copy number scatter plot (ASCAT)
```{r, echo=FALSE,,fig.width=8,fig.height=4,warning=F}
if (exists('ascat_binned')) {
  cat('Ascat:')
  #ascat_binned$tmaf[ascat_binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+#xlim(0.2,1.8)+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=ascat_binned[ascat_binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,1.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```





<!-- Read the tumor structural variant data -->
##### Read MANTA (tumor)
```{r, echo=FALSE}
if (length(vep_Tstruct_file)>0) {
  tstruc_vcf=readVcf(file = vep_Tstruct_file,genome = 'GRCh38')
  g=geno(tstruc_vcf)
  inf=info(tstruc_vcf)
  rr=rowRanges(tstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  tstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))
  setkey(x = tstruc,THISID)
  tstruc$chr <- substr(x = tstruc$seqnames,start = 4,stop = 6)
  tstruc$cumstart=tstruc$start
  tstruc$cumend=tstruc$end
  tstruc$altchr=tstruc$chr
  tstruc$altpos=tstruc$END
  tstruc$plot=F
  tstruc$arc= -1
  # loop through all and extract endpoint chr and pos
  for (i in 1:nrow(tstruc)) {
    t=strsplit(x = tstruc$ALT[[i]],split = ':')[[1]]
    if (length(t)>1 & t[1]!="<DUP") {
      tstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
      tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
      tstruc$altpos[i]=as.numeric(tt)
    }
    # also decide how it is to be plotted (not represented elsewhere, up or down arc)
    if (tstruc$chr[i]==tstruc$altchr[i]) {
      tstruc$plot[i]=T
      tstruc$arc[i]=1
    } else {
      if (tstruc[tstruc$MATEID[i][[1]],'plot']==F) tstruc$plot[i]=T # if mate not being plotted, plot this one
    }
  }
  # # for each chr get cumulative pos
  tstruc$altcumpos=tstruc$altpos
  for (i in 1:nrow(chrsz)) {
    ix=tstruc$chr==chrsz$chr[i]
    tstruc$cumstart[ix]=tstruc$start[ix]+chrsz$starts[i]
    tstruc$cumend[ix]=tstruc$end[ix]+chrsz$starts[i]
    ix=tstruc$altchr==chrsz$chr[i]
    tstruc$altcumpos[ix]=tstruc$altpos[ix]+chrsz$starts[i]
  }
  # Before looking at consequences, filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=tstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=inf$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  ## which keys are in reference?
  ix=key %in% btb_hg38_manta_pass$name # 6 only, 
  ix=btb_hg38_manta_pass$name %in% key # all once only
  ix=key %in% btb_hg38_manta_all$name # 8 only,
  ix=btb_hg38_manta_all$name %in% key #  4 with ≥2 obs
  # need for filter is marginal!
  # now using 2 observations as cutoff 
  ix <- key %in% btb_hg38_manta_all[value>1]$name
  ## do the filter
  tstruc <- tstruc[!ix]
  
  # much annotation is in the CSQ column. get Symbol and Consequence for now.
  tstruc$symbol=''
  tstruc$consequence=''
  h=strsplit(info(header(tstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  h=strsplit(h,'\\|')[[1]]
  for (i in 1:length(tstruc$CSQ)) { # for each variant
    symbol=NULL
    consequence=NULL
    for (j in 1:length(tstruc$CSQ[[i]])) { # for each annotated transcript
      symbol=c(symbol,strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]][h=='SYMBOL'])
      consequence=c(consequence,strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]][h=='Consequence'])
    }
    tstruc$symbol[i]=paste(unique(symbol),collapse=',')
    tstruc$consequence[i]=paste(unique(consequence),collapse=',')
  }
}
```

<!-- Read the normal structural variant data -->
##### Read MANTA (normal)
```{r, echo=FALSE}
if (length(vep_Nstruct_file)>0) {
  nstruc_vcf=readVcf(file = vep_Nstruct_file,genome = 'GRCh38')
  g=geno(nstruc_vcf)
  inf=info(nstruc_vcf)
  rr=rowRanges(nstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  nstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))
  setkey(x = nstruc,THISID)
  nstruc$chr <- substr(x = nstruc$seqnames,start = 4,stop = 6)
  nstruc$cumstart=nstruc$start
  nstruc$cumend=nstruc$end
  nstruc$altchr=nstruc$chr
  nstruc$altpos=nstruc$END
  nstruc$plot=F
  nstruc$arc= -1
  # loop through all and extract endpoint chr and pos
  for (i in 1:nrow(nstruc)) {
    t=strsplit(x = nstruc$ALT[[i]],split = ':')[[1]]
    if (length(t)>1 & t[1]!="<DUP") {
      nstruc$altchr[i]=strsplit(x = t[1],split = 'chr')[[1]][2]
      tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
      nstruc$altpos[i]=as.numeric(tt)
    }
    # also decide how it is to be plotted (not represented elsewhere, up or down arc)
    if (nstruc$chr[i]==nstruc$altchr[i]) {
      nstruc$plot[i]=T
      nstruc$arc[i]=1
    } else {
      if (nstruc[nstruc$MATEID[i][[1]],'plot']==F) nstruc$plot[i]=T # if mate not being plotted, plot this one
    }
  }
  # # for each chr get cumulative pos
  nstruc$altcumpos=nstruc$altpos
  for (i in 1:nrow(chrsz)) {
    ix=nstruc$chr==chrsz$chr[i]
    nstruc$cumstart[ix]=nstruc$start[ix]+chrsz$starts[i]
    nstruc$cumend[ix]=nstruc$end[ix]+chrsz$starts[i]
    ix=nstruc$altchr==chrsz$chr[i]
    nstruc$altcumpos[ix]=nstruc$altpos[ix]+chrsz$starts[i]
  }
  # Before looking at consequences, filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=nstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=inf$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  ## which keys are in reference?
  ix=key %in% btb_hg38_manta_pass$name # 11357. (why not all? this sample is in the reference.........)
  ix=btb_hg38_manta_pass$name %in% key # all but 800 ≥2 obs
  ix=key %in% btb_hg38_manta_all$name # 12559,
  ix=btb_hg38_manta_all$name %in% key #  all but 799 ≥2 obs
  # thus 1,5k reamins using "all" and 2,5k remains using "pass".
  # now using 2 observations as cutoff 
  ix <- key %in% btb_hg38_manta_all[value>1]$name
  ## do the filter
  nstruc <- nstruc[!ix]

  # much annotation is in the CSQ column. get Symbol and Consequence for now.
  nstruc$symbol=''
  nstruc$consequence=''
  h=strsplit(info(header(nstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  h=strsplit(h,'\\|')[[1]]
  for (i in 1:length(nstruc$CSQ)) { # for each variant
    symbol=NULL
    consequence=NULL
    for (j in 1:length(nstruc$CSQ[[i]])) { # for each annotated transcript
      symbol=c(symbol,strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]][h=='SYMBOL'])
      consequence=c(consequence,strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]][h=='Consequence'])
    }
    nstruc$symbol[i]=paste(unique(symbol),collapse=',')
    nstruc$consequence[i]=paste(unique(consequence),collapse=',')
  }
}
```

&nbsp;


<!-- Plot the tumor structural variant data -->
#### Structural variants (MANTA, tumor)
```{r, echo=FALSE,fig.width=15,fig.height=3}
cat(vep_Tstruct_file)
if (exists('tstruc')) {
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Somatic structural variants')+expand_limits(x=c(0,3210e6))+ylim(-2,2)
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=-2,yend=2),data=chrsz,inherit.aes = F,alpha=1/5)
  # add copy data if present
  if (exists('tratio')) {
    g=g+geom_point(aes(x=cumstart,y=log2(Ratio)),
                   data = tratio,alpha=1/5,shape='.',
                   stat="identity",colour="grey")
  }
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.7,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  #g=g+geom_point(aes(x = cumstart,y = 1,col=SVTYPE),data=tstruc[tstruc$FILTER=='PASS',])
  g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=0.999-1,yend=1-1,col=SVTYPE,linetype=IMPRECISE),
                 data = tstruc[tstruc$FILTER=='PASS' & tstruc$plot,],alpha=1/2,
                 stat = "identity",curvature=-0.2,size=1,
                 position = "identity", angle = 90, ncp = 50,
                 arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                 inherit.aes = FALSE)+
    scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
    scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
  g
}
```




<!-- Plot the normal structural variant data... Not done, too many. To be filtered. -->
#### Structural variants (MANTA, normal)
```{r, echo=FALSE,fig.width=15,fig.height=3}
cat(vep_Nstruct_file)
if (F) if (exists('nstruc')) {
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Diploid structural variants')+expand_limits(x=c(0,3210e6))
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=2),data=chrsz,inherit.aes = F,alpha=1/5)
  # add copy data if present
  if (exists('nratio')) {
    g=g+geom_point(aes(x=cumstart,y=Ratio),
                   data = nratio,alpha=1/5,shape='.',
                   stat="identity",colour="grey")
  }
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=1,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=1,label=chr),data = chrsz,inherit.aes = F)
  # add the structural variants
  g=g+geom_point(aes(x = cumstart,y = 1,col=SVTYPE),data=nstruc[nstruc$FILTER=='PASS',])
  g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=1,yend=1.001,col=SVTYPE,linetype=),
               data = nstruc[nstruc$FILTER=='PASS',],alpha=1/2,
                 stat = "identity", curvature=1,
                 position = "identity", curvature = 0.2, angle = 90, ncp = 50,
                 arrow = NULL, lineend = "butt", na.rm = FALSE, show.legend = NA,
                 inherit.aes = FALSE)
  g
}
```

##### Read Mutect2
<!-- Read annotated small variant data (Mutect 2) -->
```{r, echo=FALSE}
if (length(mutect2_file)>0) {
  vcf=readVcf(file = mutect2_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  }
  tad=t(as.data.frame((g$AD[,'TUMOR'])))   #<------- the fix so that T/N pos are not hard coded
  tsum=tad[,1]+tad[,2]
  nad=t(as.data.frame((g$AD[,'NORMAL'])))
  nsum=nad[,1]+nad[,2]
  tmuts=cbind(tmuts,g$AF)
  tmuts$AD=tad[,2]
  tmuts$DP=tsum
  tmuts$ADn=nad[,2]
  tmuts$DPn=nsum
  tmuts$reads=' '
  tmuts$reads[tmuts$AD<5]='<5'
  tmuts$reads[tmuts$AD>=5]='≥5'
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  mutect2_table=cbind(tmuts,annotation_table)
}
```
&nbsp;

<!-- Plot Mutect 2 mutations along the genome -->
#### Somatic SNVs and small indels (Mutect2)
```{r, echo=FALSE,fig.width=15,fig.height=6}
cat(mutect2_file)
if (exists('mutect2_table')) {
  par(mai=c(1,4,0,2))
  g=ggplot(data = mutect2_table[mutect2_table$filter=='PASS',],
           aes(x=cumstart,y=TUMOR,fill=type,shape=type,col=reads))+
    geom_point(alpha=1)+
    ylab('Somatic mutation allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
  # add smoothed
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                          axis.title.x=element_blank(),
                          axis.text.x=element_blank(),
                          axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  # add gene labels
  temp=mutect2_table[mutect2_table$filter=='PASS' & mutect2_table$Amino_acids!='',]
  g=g+geom_label_repel(data = temp,
                       mapping = aes(x=cumstart,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),
                       inherit.aes = F,size=2,nudge_y = 0.2)


  g
  # m=mutect2_table[,c('chr','start','end','ref','alt','AD','DP','ADn','DPn','Consequence','IMPACT','SYMBOL',
  #                    'BIOTYPE','EXON','Existing_variation','Protein_position','Amino_acids','Existing_variation')]
  # knitr::kable(m)
}
```

##### Read Strelka
<!-- Read annotated small variant data (Strelka, two files) -->
```{r, echo=FALSE}
if (length(strelka_snv_file) * length(strelka_indel_file)>0) {
  vcf=readVcf(file = strelka_snv_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
    ix=tmuts$altchr==chrsz$chr[i]
    tmuts$altcumpos[ix]=tmuts$altpos[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_snvs=cbind(tmuts,annotation_table)

  ## Read Strelka indels
  vcf=readVcf(file = strelka_indel_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
    ix=tmuts$altchr==chrsz$chr[i]
    tmuts$altcumpos[ix]=tmuts$altpos[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_indels=cbind(tmuts,annotation_table)
}
```

&nbsp;


#### Coverage and allele ratio (Mutect2, Strelka)
<!-- Make a scatter plot of Mutect2 findings -->
```{r, echo=FALSE,,fig.width=6,fig.height=4,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {
  
  # First make a merged table of findings
  mutect2_temp=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                     MARGIN = 1,FUN = paste,collapse=' ')
  strelka_temp=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                               strelka_indels[,c('chr','start','end','ref','alt'),]),
                     MARGIN = 1,FUN = paste,collapse=' ')
  temp <- mutect2_table
  temp$Callers <- 'Mutect 2'
  temp$Callers[mutect2_temp %in% strelka_temp]='Mutect 2 and Strelka'
  
  g <- ggplot()+geom_point(mapping = aes(x=DP,y=AD/DP,fill=type,shape=type,col=Callers),data=temp)+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('Mutect 2'='lightgrey','Mutect 2 and Strelka'='black'))+
    ylab('ALT allele ratio') + xlab('Local coverage')+
    scale_x_log10()+
    geom_label_repel(data = temp[temp$Amino_acids!='',],mapping = aes(x=DP,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),inherit.aes = F,size=2,nudge_y = 0,nudge_x = 0)
  g#ggplotly(g)  
}
```

&nbsp;

#### Somatic snv/indel overlap (Mutect2, Strelka)
<!-- Make a Venn diagram of Strelka and Mutect2 findings -->
```{r, echo=FALSE,,fig.width=6,fig.height=4,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {
  
  
  
  VENN.LIST=list(mutect2=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                               MARGIN = 1,FUN = paste,collapse=' '),
                 strelka=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                                         strelka_indels[,c('chr','start','end','ref','alt'),]),
                               MARGIN = 1,FUN = paste,collapse=' '))
venn.plot <- venn.diagram(VENN.LIST , NULL, fill=c("darkmagenta", "darkblue"), 
                          alpha=c(0.2,0.2), cex = 2, cat.fontface=4, 
                          category.names=c("Mutect 2", "Strelka"))
grid.draw(venn.plot)
  
  
}




```
&nbsp;




##### Read HaplotypeCaller (Normal)
<!-- Read annotated small variant data (Haplotype caller, normal) -->
```{r, echo=FALSE}
if (length(haplotypecaller_N_file)>0) {
  vcf=readVcf(file = haplotypecaller_N_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data table with relevant data
  nmuts=as.data.table(ranges(rowRanges(vcf)))
  nmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  nmuts$cumstart=nmuts$start
  nmuts$cumend=nmuts$end
  nmuts$ref=as.character(ref(vcf))
  nmuts$alt=''
  nmuts$alt2=''
  temp=as.data.table(alt(vcf))[, .(values = list(value)), by = group]
  ix=unlist(lapply(temp$values,FUN = length))==2
  nmuts$alt[!ix]=unlist(temp$values[!ix])
  t=unlist(temp$values[ix])
  nmuts$alt[ix]=t[seq(1,length(t),2)]
  nmuts$alt2[ix]=t[seq(2,length(t),2)]
 
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=nmuts$chr==chrsz$chr[i]
    nmuts$cumstart[ix]=nmuts$start[ix]+chrsz$starts[i]
    nmuts$cumend[ix]=nmuts$end[ix]+chrsz$starts[i]
  }
  nmuts$DP=g$DP[,1]
  ad=matrix(0,nrow = nrow(nmuts),ncol = 3,dimnames=list(NULL,c('RD','AD','AD2')))
  n=unlist(lapply(g$AD,length)) # gets number of AD per variant
  ad[n==2,1:2]=matrix(data=unlist(g$AD[n==2]),ncol=2,byrow = T)
  ad[n==3,1:3]=matrix(data=unlist(g$AD[n==3]),ncol=3,byrow = T)
  nmuts=cbind(nmuts,ad)
  
  ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  key1 <- paste(nmuts$chr,nmuts$start,nmuts$alt)
  key2 <- paste(nmuts$chr,nmuts$start,nmuts$alt2)
  has2 <- nmuts$alt2!=''
  key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  key2_refAF <- swegen_hg38_allele_counts[key2,.(value)]
  
  # "info"" annotations also need to be added
  nmuts=cbind(tmuts,data.table(info(vcf)[keep,]))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  mutect2_table=cbind(tmuts,annotation_table)
  rownames(mutect2_table)=NULL
}
```
&nbsp;





#### All data by chromosome
<!-- Plot composite data, chromosome wise -->
```{r}
chromplot=function(chr) {
  par(mai=c(0,4,0,2))
  g=ggplot()+
    scale_y_continuous(limits=c(-2.5,3),
                       breaks = seq(1.5,-2.5,-0.5),
                       minor_breaks = seq(1.25,-2.25,-0.5),
                       labels = 
                         c('Tumor DNA=1.5','Tumor Copy Ratio\n(median centered)','Tumor DNA=0.5\nMutation AF=1','Mutation AF=0.5','Mutation AF=0\nTumor SNP AF=1',
                           'Tumor\nSNPs','Tumor SNP AF=0\nNormal SNP AF=1','Normal\nSNPs','Normal SNP AF=0'))+
    scale_x_continuous(
      breaks = 10e6*seq(0,ceiling(chrsz$length[chrsz$chr==chr]/10e6)),
      labels= c('0',paste0(10*seq(1,ceiling(chrsz$length[chrsz$chr==chr]/10e6)),'M')),
      minor_breaks = 1e6*seq(0,ceiling(chrsz$length[chrsz$chr==chr]/1e6)))+
    theme(panel.grid.minor.x = element_line(colour="lightgrey", size=0.05),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="black", size=0.2),
          panel.grid.minor.y = element_line(colour="lightgrey", size=0.05),
          axis.title.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.title.y=element_blank(),
          axis.ticks.y=element_blank())#+

  
  ### TUMOR LOGRATIO  
  if (exists('tratio')) {
    temp <- tratio
    temp$Ratio[temp$Ratio>3]=3
    g=g+geom_line(aes(x=Start,y=Ratio),
                   data = temp[temp$Chromosome==chr,],alpha=1/3,
                   stat="identity",colour="black")
  }
  else if (exists('ascat_tratio')) {
    temp <- ascat_tratio
    temp$smoothed[temp$smoothed>3]=3
    g=g+geom_line(aes(x=Position,y=smoothed),
                   data = temp[temp$Chr==chr,],alpha=1/3,
                   stat="identity",colour="black")
  }
  
  ### TUMOR STRVAR  
  if (exists('tstruc')) {
    temp=tstruc[tstruc$chr==chr & tstruc$FILTER=='PASS' & tstruc$altchr==tstruc$chr,]
    thischrstart=chrsz$starts[chrsz$chr==chr]
    if (nrow(temp)>0) g=g+geom_curve(aes(x = start,xend = altpos,y=0.999,yend=1,col=SVTYPE,linetype=IMPRECISE),
                                     data = temp,alpha=1/2,
                                     curvature=-0.2,size=1,
                                     position = "identity", angle = 90, ncp = 50,
                                     arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                     inherit.aes = FALSE)+
      scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
      scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
  }
  
  ### TUMOR MUTATIONS  
  if (exists('mutect2_table')) {
    temp=mutect2_table[mutect2_table$chr==chr & mutect2_table$filter=='PASS',]
    if (nrow(temp)>0) { 
      g=g+geom_point(aes(x=start,y=TUMOR -0.5,fill=type,shape=type),data = temp,alpha=1)+
        scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
        scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))#+
      #scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
      
      g=g+geom_label_repel(data = temp[temp$Amino_acids!='',],
                           mapping = aes(x=start,y=TUMOR -0.5,label=paste(SYMBOL,Protein_position,Amino_acids)),
                           inherit.aes = F,size=3,min.segment.length = 0,nudge_y = 0.3)
    }
  }
  
  ### SNPS
  if (exists('tbaf')) { 
    g=g+geom_point(aes(x=Position,y=BAF -1.5),data = tbaf[tbaf$Chromosome==chr,],alpha=1/8,shape='.') 
  } else if (exists('ascat_tbaf'))  {
    g=g+geom_point(aes(x=Position,y=BAF -1.5),data = ascat_tbaf[ascat_tbaf$Chr==chr,],alpha=1/8,shape='.')
  }
  if (exists('nbaf')) {
    g=g+geom_point(aes(x=Position,y=BAF -2.5),data = nbaf[nbaf$Chromosome==chr,],alpha=1/8,shape='.')
  } else if (exists('ascat_nbaf')) {
    g=g+geom_point(aes(x=Position,y=NBAF -1.5),data = ascat_nbaf[ascat_nbaf$Chr==chr,],alpha=1/8,shape='.')
  }
  return(g)
}
```

&nbsp;

#### Chromosome 1
```{r,fig.width=15,fig.height=10}
chr=1
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 2
```{r,fig.width=15,fig.height=10}
chr=2
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 3
```{r,fig.width=15,fig.height=10}
chr=3
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 4
```{r,fig.width=15,fig.height=10}
chr=4
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 5
```{r,fig.width=15,fig.height=10}
chr=5
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 6
```{r,fig.width=15,fig.height=10}
chr=6
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 7
```{r,fig.width=15,fig.height=10}
chr=7
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 8
```{r,fig.width=15,fig.height=10}
chr=8
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 9
```{r,fig.width=15,fig.height=10}
chr=9
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 10
```{r,fig.width=15,fig.height=10}
chr=10
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 11
```{r,fig.width=15,fig.height=10}
chr=11
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 12
```{r,fig.width=15,fig.height=10}
chr=12
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 13
```{r,fig.width=15,fig.height=10}
chr=13
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 14
```{r,fig.width=15,fig.height=10}
chr=14
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 15
```{r,fig.width=15,fig.height=10}
chr=15
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 16
```{r,fig.width=15,fig.height=10}
chr=16
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 17
```{r,fig.width=15,fig.height=10}
chr=17
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 18
```{r,fig.width=15,fig.height=10}
chr=18
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 19
```{r,fig.width=15,fig.height=10}
chr=19
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 20
```{r,fig.width=15,fig.height=10}
chr=20
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 21
```{r,fig.width=15,fig.height=10}
chr=21
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome 22
```{r,fig.width=15,fig.height=10}
chr=22
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome X
```{r,fig.width=15,fig.height=10}
chr='X'
cat('Chromosome',chr)
chromplot(chr)
```
#### Chromosome Y
```{r,fig.width=15,fig.height=10}
chr='Y'
cat('Chromosome',chr)
chromplot(chr)
```

&nbsp;

#### Directory content
```{r}
{
  cat(date(),'\n')
  cat('Current directory:', getwd(),'\n')
}
{
  print(files)
}
```
