---
output: 
  html_document:
      toc: false
      toc_float: false
      toc_depth: 4
      code_folding: NULL
      df_print: paged
editor_options: 
  chunk_output_type: console
---

<!-- This script reads from a folder containing (VEP+snpEff) annotated CAW (aka Sarek) output, including copy number from ControlFreec and Ascat, structural variation from Manta, somatic point mutations from Mutact 2 and Strelka, and germline mutations from HaplotypeCaller.-->

<!-- CSS style for left alignment and for tables to look ok: -->
<style>
body {
position: absolute;
left: 42px;}
.main-container {max-width: none !important;}
table, td, th {
border: none;
padding-left: 1em;
padding-right: 1em;
min-width: 100%;
word-wrap: break-word;
max-width: 200px;
margin-left: auto;
margin-right: auto;
margin-top: 1em;
margin-bottom: 1em;
}

</style>

<!-- Markdown setup -->
```{r setup, include=F}
knitr::opts_chunk$set(echo = FALSE,comment = '',warning=F,error = F)
options(width = 600)
```

# {.tabset .tabset-pills}

```{r sample_name}
cat('Sample:',basename(getwd()),'\tDirectory:',getwd(),'\nDate:',date())
```

## Setup {.tabset}

### List files {.active}
<!-- First, make a list of files to read and visualize -->
```{r list_files}
files <- dir(recursive = T,full.names = T)
files <- files[-grep(pattern = '.png',x = files)]
# control freec files
freec_Tratio_file <- files[grep(pattern = '[TR].hg38.pileup.gz_ratio.txt',files,perl = T)][1]
freec_Nratio_file <- files[grep(pattern = '[TR].hg38.pileup.gz_normal_ratio.txt',files,perl = T)][1]
freec_Tbaf_file <- files[grep(pattern = '[TR].hg38.pileup.gz_BAF.txt',files,perl = T)][1]
freec_Nbaf_file <- files[grep(pattern = '[BN].hg38.pileup.gz_BAF.txt',files,perl = T)][1]
freec_info_file <- files[grep(pattern = 'hg38.pileup.gz_info.txt',files,perl = T)][1]
freec_cnv_file <- files[grep(pattern = "^.*[TR]\\.hg38\\.pileup\\.gz_CNVs$",files,perl = T)][1]
# Ascat files
ascat_Tratio_file <- files[grep(pattern = "^.*Ascat.*[TR]\\.LogR$",files)]
ascat_Nratio_file <- files[grep(pattern = "^.*Ascat.*[BN]\\.LogR$",files)]
ascat_Tbaf_file <- files[grep(pattern = "^.*Ascat.*[TR]\\.BAF$",files)]
ascat_Nbaf_file <- files[grep(pattern = "^.*Ascat.*[BN]\\.BAF$",files)]
ascat_segment_file <- files[grep(pattern = "^.*Ascat.*[TR]\\.cnvs\\.txt$",files)]
# Manta structural variant files
vep_Tstruct_file <- grep(pattern = "^.*Annotation/VEP/Manta.*vs.*somaticSV.vcf\\..*ann.*vcf$",files,value = T)[1]
vep_Nstruct_file <- grep(pattern = "^.*Annotation/VEP/Manta.*vs.*diploidSV.vcf\\..*ann.*vcf$",files,value = T)[1]
# Haplotype caller variant files
haplotypecaller_T_file <- grep(pattern = "^.*Annotation/VEP/haplotypecaller.*[TR]\\.vcf.*vcf$",files,value = T)[1] # local
haplotypecaller_N_file <- grep(pattern = "^.*Annotation/VEP/haplotypecaller.*[BN]\\.vcf.*vcf$",files,value = T)[1] # local
# snpEff+VEP Mutect2 file
mutect2_file <- grep(pattern = "^.*Annotation/VEP/mutect2_.*[TR]_vs_.*[BN]\\.vcf.*vcf$",files,value = T)[1]
# snpEff + VEP Strelka snv file
strelka_snv_file <- grep(pattern = "^.*Annotation/VEP/Strelka_.*_somatic_snvs.*vcf$",files,value = T)[1]
# snpEff + VEP Strelka indel file
strelka_indel_file <- grep(pattern = "^.*Annotation/VEP/Strelka_.*_somatic_indels.*vcf$",files,value = T)[1]


rm(files)
{
  cat('Selected files in',getwd(),'\n\n')
  for ( obj in ls() ) { 
    cat('---',obj,'---\n') 
    print(get(obj)) 
  }
}  


```


```{r list_all_files}
{
  cat('All files in',getwd(),'\n\n')
  print(dir(recursive = T,full.names = T))
}
```

### Load dependencies
<!-- Dependencies -->
```{r load_dependencies}
library(data.table)
library(ggplot2); theme_set(theme_light())
library(grid)
library(dplyr)
library(ggrepel)
library(VariantAnnotation)
library(VennDiagram)
library(knitr)
library(htmlTable)
#library(kableExtra)

listfix <- function(list) {
  l=lapply(list,length)
  for (i in which(l!=1)) list[[i]]=0
  unlist(list)
}

sampleData <- data.table(
  date=date(),
  directory=getwd(),
  name=basename(getwd())
)
chrsz <- data.table(
  chr = c("1", "2", "3", "4", "5", "6", "7", "8", 
          "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", 
          "20", "21", "22", "X", "Y"), 
  starts = c(0, 253902921, 500669562, 
             703689723, 898025604, 1084280925, 1259870526, 1424144247, 1574191848, 
             1717288929, 1855896810, 1995944331, 2134165212, 2253485013, 2363996694, 
             2470839615, 2565902256, 2654091777, 2739309138, 2802869979, 2871941700, 
             2923598421, 2979326382, 3140008743), 
  length = c(248902921, 241766641, 
             198020161, 189335881, 181255321, 170589601, 159273721, 145047601, 
             138097081, 133607881, 135047521, 133220881, 114319801, 105511681, 
             101842921, 90062641, 83189521, 80217361, 58560841, 64071721, 
             46656721, 50727961, 155682361, 56827081)
)
```

<!-- Reference data files -->
```{r load_reference}
snptable=fread('~/reports/swegen_snp_counts.csv',key='name')
swegen_hg38_manta_all=fread('~/reports/swegen_sv_counts.csv',key='name')
allgenes=fread('~/reports/genes_ensemble_canonical.csv',key='name')
tumorgenes=fread('~/reports/cancer_gene_census.csv')
local_tumorgenes=fread('~/reports/Gene_list_2018.csv')[Gene!='']
local_tumorgenes$`Gene Symbol`=local_tumorgenes$Gene
tumorgenes=merge(tumorgenes,local_tumorgenes,all=T)
setkey(tumorgenes,`Gene Symbol`)
hotspots_inframe=fread('~/reports/hotspots_v2_inframe.csv')
hotspots_snv=fread('~/reports/hotspots_v2_snv.csv')
#cosmic_coding=fread('~/reports/cosmic_coding.csv',key = 'name')
#cosmic_noncoding=fread('~/reports/cosmic_noncoding.csv',key = 'name')
#cosmic_fusions=fread('~/reports/cosmic_fusions.csv',key = 'name')
```


## Copy number {.tabset}

### ControlFreec
<!-- Read Control Freec copy number data -->
```{r load_control_freec}
cnvs=NULL
if (length(freec_Tratio_file)>0 &
    length(freec_Nratio_file)>0 &
    length(freec_Tbaf_file)>0 &
    length(freec_Nbaf_file)>0) {
  # tumor log ratio
  tratio <- fread(file = freec_Tratio_file)
  tratio <- tratio[order(Chromosome,Start)]
  # normal log ratio
  nratio <- fread(file = freec_Nratio_file)
  nratio <- nratio[order(Chromosome,Start)]
  # tumor BAF
  tbaf <- fread(freec_Tbaf_file)[,1:3]
  # normal BAF
  nbaf <- fread(freec_Nbaf_file)[,1:3]
  # merged
  mbaf <- merge(tbaf,nbaf,by = c('Chromosome','Position'))
  
  # manipulate for plot:
  tratio$cumstart <- tratio$Start
  nratio$cumstart <- nratio$Start
  tbaf$cumstart <- tbaf$Position
  nbaf$cumstart <- nbaf$Position
  for (i in 2:nrow(chrsz)) {
    ix <- tratio$Chromosome==chrsz$chr[i]
    tratio$cumstart[ix] <- tratio$Start[ix]+chrsz$starts[i]
    ix <- nratio$Chromosome==chrsz$chr[i]
    nratio$cumstart[ix] <- nratio$Start[ix]+chrsz$starts[i]
    ix <- tbaf$Chromosome==chrsz$chr[i]
    tbaf$cumstart[ix] <- tbaf$Position[ix]+chrsz$starts[i]
    ix <- nbaf$Chromosome==chrsz$chr[i]
    nbaf$cumstart[ix] <- nbaf$Position[ix]+chrsz$starts[i]
  }
  
  # modify for plot
  tratio$CopyNumber[tratio$CopyNumber>4] <- 4
  nratio$CopyNumber[nratio$CopyNumber>4] <- 4
  tratio$CopyNumber[1:2] <- c(0,4)
  nratio$CopyNumber[1:2] <- c(0,4)
  tratio$BAF[tratio$BAF<0.5 | tratio$BAF>1] <- NA
  nratio$BAF[nratio$BAF<0.5 | nratio$BAF>1] <- NA
  
  # smooth data for plot
  binned <- NULL
  for (i in 1:nrow(chrsz)) {
    temp <- data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      nratio=NA,
      tmaf=NA,
      nmaf=NA)
    ctratio <- tratio[Chromosome==chrsz$chr[i]]
    cnratio <- nratio[Chromosome==chrsz$chr[i]]
    for (j in 1:nrow(temp)) {
      ix <- ctratio$Start>temp$pos[j]-5e5 & ctratio$Start<temp$pos[j]+5e5
      #temp$tratio[j] <- median(ctratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) {
        d <- density(ctratio$Ratio[ix],na.rm=T)
        temp$tratio[j] <- d$x[which.max(d$y)]
        t=ctratio$BAF[ix]
        t=t[!is.na(t)]
        if (length(t)>10) {
          d <- density(ctratio$BAF[ix],na.rm=T)
          temp$tmaf[j] <- d$x[which.max(d$y)]
        }
      }
      ix <- cnratio$Start>temp$pos[j]-5e6 & cnratio$Start<temp$pos[j]+5e6
      temp$nratio[j] <- median(cnratio$Ratio[ix],na.rm = T)
      if (sum(ix)>20) try({
        d <- density(cnratio$BAF[ix],na.rm=T)
        temp$nmaf[j] <- d$x[which.max(d$y)]
      },silent=T)
    }
    binned <- rbind(binned,temp)
  }
  
  # make CNA burden estimate from binned:
  sampleData[['Control_Freec_CNA_burden']] <- round(binned[chr %in% 1:22,sum(abs(tratio-median(tratio))>0.1,na.rm = T)/.N]*100)
  
  # check which regions are worth reporting
  # cnv file
  cnvs <- as.data.table(read.csv(freec_cnv_file,sep='\t',header=F))
  names(cnvs) <- c('chr','start','end','copies','effect','genotype','value','type','na')
  freec_loh=cnvs[grep('^A*$',genotype)][type=='somatic'][copies<5]
  cnvs=cnvs[effect!='neutral']
  cnvs$rank_score <- 0
  cnvs$rank_terms <- ''
  cnvs$cancer_genes <- ''
  
  for (i in cnvs[end-start < 10e6, .I]) {
    all=allgenes[chr==cnvs$chr[i] & start<cnvs$end[i] & end>cnvs$start[i],name] # genes in segment
    if (length(all)>0) { # if any genes
      genes=tumorgenes[`Gene Symbol` %in% all] 
      if (nrow(genes)>0) { # if any of our cancer genes
        cnvs$cancer_genes[i] <- paste(genes$`Gene Symbol`,collapse = ' ') # put in table
        if (any(genes$Tier==1,na.rm = T)) {
          cnvs$rank_score[i]=3
          cnvs$rank_terms[i]='Tier1'
        } else if (any(genes$Tier==2,na.rm = T)) {
          cnvs$rank_score[i]=2
          cnvs$rank_terms[i]='Tier2'
        } else {
          cnvs$rank_score[i]=2
          cnvs$rank_terms[i]='published_gene'
        }
      }
    }
    
    # if the effect is focal
    if (cnvs[i,end-start]<3e6) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'focal')
    }
    
    # if the effect is loss/gain/amp etc
    if (cnvs$copies[i]==0) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+2
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'hz_loss')
    } else if (cnvs$copies[i] > 5) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+2
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'high_amp')
    } else if (cnvs$copies[i] > 2) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'dup')
    } else if (cnvs$effect[i] =='loss') {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'del')
    }
  }
  
  freec_cnv=cnvs[order(rank_score,decreasing = T)]
  
  htmlTable(freec_cnv, col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #scroll_box(width = "1400px", height = "600px")# %>%
  # row_spec(row = which(cnvs$Gene_Tier==2 | cnvs$Effect_Tier==2), background = "#FF500040") %>%
  # row_spec(row = which(cnvs$Gene_Tier==1 | cnvs$Effect_Tier==1), background = "#FFFF0040") %>%
  # row_spec(row = which(cnvs$Gene_Tier==1 & cnvs$Effect_Tier==1), background = "#00FF0040")
  # 
  
}
```

### Ascat
<!-- Read ASCAT copy number data -->
```{r load_ascat}
ascat_cnv=NULL
if (length(ascat_Tratio_file)>0 &
    length(ascat_Nratio_file)>0 &
    length(ascat_Tbaf_file)>0 &
    length(ascat_Nbaf_file)>0 & 
    length(ascat_segment_file)>0) {
  # segmented copy number
  ascat_cnv=fread(file = ascat_segment_file)
  # tumor log ratio
  ascat_tratio=fread(file = ascat_Tratio_file)[,-1]
  colnames(ascat_tratio)[3]='LogR'
  ascat_tratio$Chr=substr(ascat_tratio$Chr,start = 4,stop = 8)
  ascat_tratio=ascat_tratio[Chr %in% c(1:22,'X','Y')]
  ascat_tratio=ascat_tratio[order(Chr,Position)]
  # tumor BAF
  ascat_tbaf=fread(ascat_Tbaf_file)[,-1]
  colnames(ascat_tbaf)[3]='BAF'
  ascat_tbaf=ascat_tbaf[which(BAF!=0 & BAF!=1)]
  ascat_tbaf$Chr=substr(ascat_tbaf$Chr,start = 4,stop = 8)
  ascat_tbaf=ascat_tbaf[Chr %in% c(1:22,'X','Y')]
  # normal BAF
  ascat_nbaf=fread(ascat_Nbaf_file)[,-1]
  colnames(ascat_nbaf)[3]='NBAF'
  ascat_nbaf=ascat_nbaf[which(NBAF!=0 & NBAF!=1)]
  ascat_nbaf$Chr=substr(ascat_nbaf$Chr,start = 4,stop = 8)
  ascat_nbaf=ascat_nbaf[Chr %in% c(1:22,'X','Y')]
  
  # join (T) log ratio and (T+N) BAF
  ascat_tratio=merge(ascat_tratio,ascat_tbaf,all.x=T)
  ascat_tratio=merge(ascat_tratio,ascat_nbaf,all.x=T)
  
  # manipulate for plot:
  ascat_tratio$cumstart=ascat_tratio$Position
  ascat_tbaf$cumstart=ascat_tbaf$Position
  ascat_nbaf$cumstart=ascat_nbaf$Position
  ascat_cnv$cumstart=ascat_cnv$startpos
  ascat_cnv$cumend=ascat_cnv$endpos
  ascat_cnv$LOH=''; ascat_cnv$LOH[ascat_cnv$nMinor==0]='LOH'
  for (i in 2:nrow(chrsz)) {
    ix=ascat_tratio$Chr==chrsz$chr[i]
    ascat_tratio$cumstart[ix]=ascat_tratio$Position[ix]+chrsz$starts[i]
    #no n...
    ix=ascat_tbaf$Chr==chrsz$chr[i]
    ascat_tbaf$cumstart[ix]=ascat_tbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_nbaf$Chr==chrsz$chr[i]
    ascat_nbaf$cumstart[ix]=ascat_nbaf$Position[ix]+chrsz$starts[i]
    ix=ascat_cnv$chr==chrsz$chr[i]
    if (sum(ix)>0) {
      ascat_cnv$cumstart[ix]=ascat_cnv$startpos[ix]+chrsz$starts[i]
      ascat_cnv$cumend[ix]=ascat_cnv$endpos[ix]+chrsz$starts[i]
    }
  }
  # copy number hack
  ascat_tratio$Ratio=2^ascat_tratio$LogR
  ascat_tratio$smoothed=runmed(x = ascat_tratio$Ratio,k = 99)
  ascat_tratio$CopyNumber=2
  ascat_tratio$MinorCopy=1
  for (i in 1:nrow(ascat_cnv)) {
    ix=ascat_tratio$cumstart > ascat_cnv$cumstart[i] & ascat_tratio$cumstart < ascat_cnv$cumend[i]
    ascat_tratio$CopyNumber[ix]=ascat_cnv$nMajor[i]+ascat_cnv$nMinor[i]
    ascat_tratio$MinorCopy[ix]=ascat_cnv$nMinor[i]
  }
  ascat_tratio$CopyNumber[ascat_tratio$CopyNumber>4]=4
  ascat_tratio$CopyNumber[1:2]=c(0,4)
  ascat_tratio$LOH='no'
  ascat_tratio$LOH[ascat_tratio$MinorCopy==0]='yes'
  ascat_tratio$Copies=as.character(ascat_tratio$CopyNumber)
  ascat_tratio$Copies[ascat_tratio$LOH=='yes' & ascat_tratio$CopyNumber>1]='LOH'
  ascat_tratio$Copies[ascat_tratio$CopyNumber==4]='≥4'
  ascat_tratio$Copies[1:6]=c('0','1','LOH','2','3','≥4')
  
  # smooth data for view
  ascat_binned=NULL
  for (i in 1:nrow(chrsz)) {
    temp=data.table(
      chr=chrsz$chr[i],
      pos=seq(5e5,chrsz$length[i],5e5),
      cumpos=seq(5e5,chrsz$length[i],5e5)+chrsz$starts[i],
      tratio=NA,
      tmaf=NA)
    
    cix=ascat_tratio$Chr==chrsz$chr[i]
    tpos=ascat_tratio$Position[cix]
    tLogR=ascat_tratio$LogR[cix]
    tBAF=0.5+abs(ascat_tratio$BAF[cix]-0.5)
    for (j in 1:nrow(temp)) {
      ix = tpos>temp$pos[j]-5e5 & tpos<temp$pos[j]+5e5
      if (sum(!is.na(tLogR[ix]))>10) {
        d <- density(2^tLogR[ix],na.rm=T)
        temp$tratio[j] <- d$x[which.max(d$y)]
      }
      #temp$tratio[j]=2^median(tLogR[ix],na.rm = T)
      if (sum(!is.na(tBAF[ix]))>20) {
        d=density(tBAF[ix],na.rm=T)
        temp$tmaf[j]=d$x[which.max(d$y)]
      }
    }
    ascat_binned=rbind(ascat_binned,temp)
  }
  # make CNA burden estimate from binned
  sampleData[['Ascat_CNA_burden']] <- round(ascat_binned[chr %in% 1:22,sum(abs(tratio-median(tratio))>0.1,na.rm = T)/.N]*100)
  
  
  ascat_loh=ascat_cnv[nMinor==0 & nMajor<5]
  cnvs=ascat_cnv[nMajor+nMinor != 2]
  cnvs$rank_score <- 0
  cnvs$rank_terms <- ''
  cnvs$cancer_genes <- ''
  
  for (i in which(cnvs$endpos-cnvs$startpos < 5e6)) {
    all=allgenes[chr==cnvs$chr[i] & start<cnvs$end[i] & end>cnvs$start[i],name] # genes in segment
    if (length(all)>0) { # if any genes
      genes=tumorgenes[`Gene Symbol` %in% all] 
      if (nrow(genes)>0) { # if any of our cancer genes
        cnvs$cancer_genes[i] <- paste(genes$`Gene Symbol`,collapse = ' ') # put in table
        if (any(genes$Tier==1,na.rm = T)) {
          cnvs$rank_score[i]=3
          cnvs$rank_terms[i]='Tier1'
        } else if (any(genes$Tier==2,na.rm = T)) {
          cnvs$rank_score[i]=2
          cnvs$rank_terms[i]='Tier2'
        } else {
          cnvs$rank_score[i]=2
          cnvs$rank_terms[i]='published_gene'
        }
      }
    }
    
    # if the effect is focal
    if (cnvs[i,endpos-startpos]<3e6) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'focal')
    }
    
    # if the effect is loss/gain/amp etc
    if (cnvs[i,nMajor+nMinor]==0) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+2
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'hz_loss')
    } else if (cnvs[i,nMajor+nMinor] > 5) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+2
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'amp')
    } else if (cnvs[i,nMajor+nMinor] > 2) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'dup')
    } else if (cnvs[i,nMajor+nMinor] < 2) {
      cnvs$rank_score[i]=cnvs$rank_score[i]+1
      cnvs$rank_terms[i]=paste(cnvs$rank_terms[i],'del')
    }
  }
  
  ascat_cnv=cnvs[order(rank_score,decreasing = T)]
  
  # # table of findings from ascat_cnv
  # ascat_cnv$Gene_Tier <- 3
  # ascat_cnv$Effect_Tier <- 3
  # ascat_cnv$cosmic_genes <- ''
  # #ascat_cnv$all_genes=''
  # # amplifications/duplications max 5MB, homdel max 3MB
  # ascat_cnv <- ascat_cnv#[(endpos-startpos < 5e6 & nMajor+nMinor > 4) | (endpos-startpos < 3e6 & nMajor==0)]
  # # add genes 
  # for (i in 1:nrow(ascat_cnv)) {
  #   all=allgenes[chr==ascat_cnv$chr[i] & start<ascat_cnv$endpos[i] & end>ascat_cnv$startpos[i],name]
  #   if (length(all)>0) { # if any genes in cnv
  #     #ascat_cnv$all_genes[i]=paste(all,collapse = ' ') # put in table
  #     t2=all[(all %in% tumorgenes$`Gene Symbol`)] 
  #     if (length(t2)>0) { # if any in cosmic
  #       ascat_cnv$cosmic_genes[i] <- paste(t2,collapse = ' ') # put in table
  #       ascat_cnv$Gene_Tier[i]=2
  #       ascat_cnv$Effect_Tier[i]=2
  #       # if the effect is TSG loss or oncogene amp, Tier 1
  #       if (ascat_cnv$nMajor[i]==0) if (any(all %in% tumorgenes[grep('TSG',`Role in Cancer`),`Gene Symbol`]))
  #         ascat_cnv$Effect_Tier[i]=1
  #       if (ascat_cnv$nMajor[i] > 4) if (any(all %in% tumorgenes[grep('oncogene',`Role in Cancer`),`Gene Symbol`]))
  #         ascat_cnv$Effect_Tier[i]=1
  #     }
  #     t1=all[(all %in% tumorgenes[Tier==1,`Gene Symbol`])] 
  #     if (length(t1)>0) { # if any in cosmic T1
  #       ascat_cnv$Gene_Tier[i]=1
  #     }
  #   }
  # }
  htmlTable(ascat_cnv, col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #scroll_box(width = "1400px", height = "600px")
}
```



### View ControlFreec {.active}

#### Genome view
<!-- Plot Control Freec tumor copy number log ratio, tBAF and nBAF -->
```{r plot_control_freec,fig.width=15,fig.height=10}
if (exists('tratio')) {
  cat('Control Freec: ',freec_Tratio_file)
  par(mai=c(0,4,0,2))
  temp <- tratio
  temp$Ratio[temp$Ratio>3]=3
  
  g=ggplot()+ylab('B allele ratio and Coverage Ratio')+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    expand_limits(x=c(0,3200e6))+
    scale_y_continuous(limits=c(-1,3),
                       breaks = 0:2,
                       minor_breaks = seq(0,2.5,0.5),
                       labels = 0:2
    )
  # normal logR below
  g=g+geom_point(aes(x=cumstart,y=Ratio-0.3),col='darkgrey',data=nratio,alpha=1/5,shape='.')
  # tumor logR
  g=g+geom_point(aes(x=cumstart,y=Ratio,col=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.title.x=element_blank(),
            axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(data = tbaf,aes(x=cumstart,y=-0.5+0.5*BAF),
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(data = nbaf,aes(x=cumstart,y=-1+0.5*BAF),
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)
  
  g
}
```

&nbsp;

#### Scatter plot
<!-- Scatter plot -->
```{r scatter_plot_control_freec, echo=FALSE,fig.width=15,fig.height=8,warning=F}
if (exists('binned')) {
  binned$tmaf[binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+expand_limits(x=c(.3,3))+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=binned[binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,2.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```



### View Ascat

#### Genome view
<!-- Plot ASCAT tumor copy number log ratio, tBAF and nBAF -->
```{r plot_ascat,fig.width=15,fig.height=10}
if (exists('ascat_tratio')) {
  cat('ASCAT: ',ascat_Tratio_file,ascat_segment_file,ascat_Tbaf_file,ascat_Nbaf_file)
  par(mai=c(0,4,0,2))
  temp <- ascat_tratio
  temp$smoothed[temp$smoothed>3]=3
  g=ggplot()+ylab('B allele ratio and Coverage Ratio')+
    scale_color_gradientn(colours = c('violet','blue','black','red','orange'))+
    scale_y_continuous(limits=c(-1,3),
                       breaks = 0:2,
                       minor_breaks = seq(0,2.5,0.5),
                       labels = 0:2
    )
  expand_limits(x=c(0,3200e6))
  # using smoothed logR
  g=g+geom_point(aes(x=cumstart,y=smoothed,color=CopyNumber),data=temp,alpha=1/5,shape='.')
  # add smoothed
  g=g+geom_line(aes(x=cumpos,y=tratio),stat="identity",colour="green",data=ascat_binned)
  
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.title.x=element_blank(),
            axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
  # add tBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-0.5+0.5*BAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add nBAF
  g=g+geom_point(mapping = aes(x=cumstart,y=-1+0.5*NBAF),data = ascat_tratio,
                 alpha=1/50,shape='.')
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y= -1,yend=3),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0.1,label=chr),data = chrsz,inherit.aes = F)
  
  g
}
```

&nbsp;

#### Scatter plot
<!-- Scatter plot -->
```{r scatter_plot_ascat, echo=FALSE,fig.width=15,fig.height=8,warning=F}
if (exists('ascat_binned')) {
  #ascat_binned$tmaf[ascat_binned$tmaf<0]=1
  g=ggplot()+expand_limits(y=c(.5,1))+expand_limits(x=c(.3,3))+
    geom_point(aes(x = tratio,y = tmaf,col=chr),data=ascat_binned[ascat_binned$tratio<1.8,])+
    xlab('Tumor (1Mb mean) DNA ratio')+ylab('Tumor (1Mb mean) major allele ratio')+
    scale_y_continuous(breaks = seq(0.5,1,0.1)) +
    scale_x_continuous(breaks = seq(0.5,2.5,0.1))+
    theme(panel.grid.minor.x = element_blank(),
          panel.grid.major.x = element_line(colour="grey", size=0.2),
          panel.grid.major.y = element_line(colour="grey", size=0.2),
          panel.grid.minor.y = element_blank()
    )
  g
}
```


## Structural variants {.tabset}

<!-- Read the tumor structural variant data -->
### Manta (tumor)
```{r load_manta_tumor, echo=FALSE}
if (length(vep_Tstruct_file)>0) {
  tstruc_vcf=readVcf(file = vep_Tstruct_file,genome = 'GRCh38')
  g=geno(tstruc_vcf)
  inf=info(tstruc_vcf)
  rr=rowRanges(tstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  tstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))[FILTER=='PASS']
  tstruc$SVLEN=listfix(tstruc$SVLEN)
  setkey(x = tstruc,THISID)
  tstruc$chr <- substr(x = tstruc$seqnames,start = 4,stop = 6)
  tstruc$cumstart=tstruc$start
  tstruc$cumend=tstruc$end
  tstruc$altchr=tstruc$chr
  tstruc$altpos=tstruc$END
  tstruc$plot=F
  tstruc$arc= -1
  
  # Filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=tstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=tstruc$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  
  # now using 5 observations as cutoff
  tstruc$Swegen_count=swegen_hg38_manta_all[key,value]
  tstruc$Swegen_count[is.na(tstruc$Swegen_count)]=0
  ## do the filter
  tstruc <- tstruc[Swegen_count<5]
  if (nrow(tstruc)>0) {
    
    # loop through all and extract endpoint chr and pos
    for (i in 1:nrow(tstruc)) try( {                    # <----  sometimes error here, so try..
      t=strsplit(x = tstruc$ALT[[i]],split = ':')[[1]]
      if (length(t)>1 & t[1]!="<DUP") {
        tchr=strsplit(x = t[1],split = 'chr')[[1]][2]
        if (is.na(tchr)) tchr=strsplit(x = t[1],split = 'CHR')[[1]][2]
        tstruc$altchr[i]=tchr
        tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
        tstruc$altpos[i]=as.numeric(tt)
      }
      
      # also decide how it is to be plotted (not represented elsewhere, up or down arc)
      if (tstruc$chr[i]==tstruc$altchr[i]) {
        tstruc$plot[i]=T
        tstruc$arc[i]=1
      } else {
        tstruc$plot[i]=T
        try({
          if (tstruc[tstruc$MATEID[i][[1]],'plot']==T) tstruc$plot[i]=F # if mate not being plotted, plot this one
        }, silent=T)
      }
    }, silent=T)
    # # for each chr get cumulative pos
    tstruc$altcumpos=tstruc$altpos
    for (i in 1:nrow(chrsz)) {
      ix=tstruc$chr==chrsz$chr[i]
      if (sum(ix)>0) {
        tstruc$cumstart[ix]=tstruc$start[ix]+chrsz$starts[i]
        tstruc$cumend[ix]=tstruc$end[ix]+chrsz$starts[i]
      }
      ix=tstruc$altchr==chrsz$chr[i]
      if (sum(ix)>0) {
        tstruc$altcumpos[ix]=tstruc$altpos[ix]+chrsz$starts[i]
      }
    }
    
    
    # snpEff annotation is in the ANN column.
    h=strsplit(info(header(tstruc_vcf))['ANN',][,3],'annotations: \'')[[1]][2]
    snpEff_header=trimws(strsplit(h,'\\|')[[1]])
    
    # snpEff annotation is put in snpEff_table
    snpEff_table=matrix(data = NA,nrow = length(unlist(tstruc$ANN)),ncol = length(snpEff_header)+1)
    colnames(snpEff_table)=c('N',snpEff_header)
    row=1
    for (i in 1:nrow(tstruc)) { # for each variant
      # for each VEP annotation:
      for (j in 1:length(tstruc$ANN[[i]]))  if (length(tstruc$ANN[[i]])>0) {
        line=strsplit(tstruc$ANN[[i]][j],'\\|')[[1]]
        snpEff_table[row,1]=i
        snpEff_table[row,1+(1:length(line))]=line
        row=row+1
      }
    }
    snpEff_table=unique(as.data.table(snpEff_table)[,1:5])
    
    # Add to data
    tstruc$N=as.character(1:nrow(tstruc))
    manta_tumor_table <- merge(tstruc,snpEff_table,by='N')[order(cumstart),-1][nchar(Gene_Name)<40]
    
    ## Make new table with predicted effects
    selection <- unique(
      manta_tumor_table[,
                        .(SVTYPE,ID=THISID,chr,start,end,Swegen_count,
                          altchr,altpos,SVLEN=unlist(SVLEN),
                          Gene_Name,rank_score=0,rank_terms='',LOH='',Comment='',
                          Annotation,Annotation_Impact,Allele,
                          REF,ALT=unlist(ALT))]
    )
    
    if (nrow(selection)>0) {  
      
      for (gene in unique(selection$Gene_Name)) if (gene!='') {
        ix=selection$Gene_Name==gene
        gene=strsplit(gene,'&')[[1]]
        # Add gene scores
        if (any(gene %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==1)])) {
          selection$rank_score[ix]=3
          selection$rank_terms[ix]='Tier1'
        } else if (any(gene %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==2)])) {
          selection$rank_score[ix]=2
          selection$rank_terms[ix]='Tier2'
        }
      }
      
      # Add high impact
      ix=selection$Annotation_Impact=='HIGH'
      if (any(ix)) {
        selection$rank_score[ix]=selection$rank_score[ix]+2
        selection$rank_terms[ix]=paste(selection$rank_terms[ix],'high_impact')
      }
      
      # +1 for focal
      ix=selection$chr==selection$altchr & selection$end-selection$start < 3e6 & selection$altpos-selection$start < 3e6
      if (any(ix)) {
        selection$rank_score[ix]=selection$rank_score[ix]+1
        selection$rank_terms[ix]=paste(selection$rank_terms[ix],'focal')
      }
      
      # +1 for fusion
      ix=grep('fusion',selection$Annotation)
      if (length(ix)>0) {
        selection$rank_score[ix]=selection$rank_score[ix]+1
        selection$rank_terms[ix]=paste(selection$rank_terms[ix],'fusion')
      }
      
      # -1 for ablation if long del
      ix=intersect(
        which(selection$SVTYPE=='DEL' & selection$chr==selection$altchr & abs(selection$altpos-selection$start) > 3e6),
        grep('ablation',selection$Annotation)
      )
      if (length(ix)>0) {
        selection$rank_score[ix]=selection$rank_score[ix]-1
        selection$rank_terms[ix]=paste(selection$rank_terms[ix],'long_del')
      }
      
      # -1 for duplication if long dup
      ix=intersect(
        which(selection$SVTYPE=='DUP' & selection$chr==selection$altchr & abs(selection$altpos-selection$start) > 3e6),
        grep('duplication',selection$Annotation)
      )
      if (length(ix)>0) {
        selection$rank_score[ix]=selection$rank_score[ix]-1
        selection$rank_terms[ix]=paste(selection$rank_terms[ix],'long_dup')
      }
    }
    
    
  }
  
  # Add Control Freec LOH.
  if (nrow(selection)>0) for (i in 1:nrow(selection)) {
    ix=freec_loh$chr==selection$chr[i] &
      freec_loh$start<selection$start[i] &
      freec_loh$end>selection$end[i]
    ix=ix[!is.na(ix)]
    if (sum(ix) >0) selection$LOH[i]='Y'
  }
  
  
  # # Set Gene and Effect tiers: 
  # for (i in selection[,.I]) {
  #   symbol=strsplit(selection$Gene_Name[i],'&')[[1]] # fusions are reported as "gene1&gene2"
  #   
  #   if (any(symbol %in% tumorgenes$`Gene Symbol`)) selection$Gene_Tier[i]=2
  #   if (any(symbol %in% tumorgenes[Tier==1,`Gene Symbol`]) & length(symbol)<3) selection$Gene_Tier[i]=1
  #   
  #   length=abs(selection$SVLEN[i])
  #   gene=tumorgenes[`Gene Symbol` %in% symbol,] # this gene(s) from Cosmic table
  #   if (nrow(gene)>0) {
  #     # first fusions/translocations
  #     if (selection$Annotation[i]=='gene_fusion') {
  #       g=grep('fusion',gene$`Role in Cancer`)
  #       if (length(g)>0) { # this is a fusion and either is fusion gene yes
  #         selection$Effect_Tier[i]=1 # then at least tier 1/2?
  #         selection$Comment[i]='partial_known_fusion'
  #       }    
  #       # now 1 if perfect match with Cosmic
  #       if (nrow(gene)>1) if (as.logical(grep(gene$`Gene Symbol`[2],gene$`Translocation Partner`[1]))) {
  #         selection$Effect_Tier[i]=1 # common translocation/fusion
  #         selection$Comment[i]='known_fusion'
  #       } 
  #     }
  #     # then deletions
  #     if (selection$SVTYPE[i]=='DEL' & length<3e6) if (length(grep('ablation',selection$Annotation[i]))>0) { # small DEL, ablation effect
  #       g=grep('TSG',gene$`Role in Cancer`)
  #       if (length(g)>0) { # Tumor supressor gene yes
  #         selection$Effect_Tier[i]=1
  #         selection$Comment[i]=paste(selection$Comment[i],'TSG_del')
  #       } else { # else 2 unless already one
  #         selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #         selection$Comment[i]=paste(selection$Comment[i],'non_TSG_del')
  #       }
  #     }
  #     # then amplifications
  #     if (selection$SVTYPE[i]=='DUP' & length<5e6) if (length(grep('duplication',selection$Annotation[i]))>0) { # small amp, dupplication effect
  #       g=grep('oncogene',gene$`Role in Cancer`)
  #       if (length(g)>0) { # Oncogene yes
  #         selection$Effect_Tier[i]=1
  #         selection$Comment[i]=paste(selection$Comment[i],'oncogene_dup')
  #       } else {
  #         selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #         selection$Comment[i]=paste(selection$Comment[i],'non_oncogene_dup')
  #       }
  #     }
  #     # last other disruptions
  #     if (length(grep('duplication',selection$Annotation[i])==0) &
  #         length(grep('ablation',selection$Annotation[i])==0) &
  #         length(grep('fusion',selection$Annotation[i]))==0) # if not dup/del/fusion
  #       if (any(symbol %in% tumorgenes$`Gene Symbol`)) # and if in cosmic
  #         if (selection$Annotation_Impact[i]=='HIGH') {
  #           selection$Effect_Tier[i]=2 # some indication of effect!
  #           selection$Comment[i]=paste(selection$Comment[i],'high_impact')
  #         }
  #   }
  # }
  # selection <- selection[order(Effect_Tier,Gene_Tier)]
  
  
  # # VEP annotation is in the CSQ column.
  # h=strsplit(info(header(tstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  # vep_header=strsplit(h,'\\|')[[1]]
  # 
  # # VEP annotation is put in vep_table
  # vep_table=matrix(data = NA,nrow = length(unlist(tstruc$CSQ)),ncol = length(vep_header)+1)
  # colnames(vep_table)=c('N',vep_header)
  # row=1
  # for (i in 1:nrow(tstruc)) { # for each variant
  #   # for each VEP annotation:
  #   for (j in 1:length(tstruc$CSQ[[i]])) {
  #     line=strsplit(tstruc$CSQ[[i]][j],'\\|')[[1]]
  #     vep_table[row,1]=i
  #     vep_table[row,1+(1:length(line))]=line
  #     row=row+1
  #   }
  # }
  # vep_table=as.data.table(vep_table)
  
  # # Add to data
  # tstruc$N=as.character(1:nrow(tstruc))
  # manta_tumor_table <- merge(tstruc,vep_table,by='N')[order(cumstart),-1]
  # 
  # # Deletion: consider only deletions <3MB for "transcript ablation" consequence, report census gene(s)
  # del_ok <- which(manta_tumor_table$SVTYPE=='DEL' & abs(manta_tumor_table$SVLEN)<3e6)
  # # Amplification: consider <5M for amplification consequence, report census genes
  # amp_ok <- which(manta_tumor_table$SVTYPE=='DUP' & abs(manta_tumor_table$SVLEN)<5e6)
  # # Fusions/disrupting: Do the SVs start/end at gene? 
  # manta_tumor_table$Starts_Near <- ''
  # manta_tumor_table$Ends_Near <- ''
  # manta_tumor_table$Junction <- ''
  # cumstart=allgenes$cumstart
  # cumend=allgenes$cumend
  # for (i in 1:nrow(manta_tumor_table)) {
  #   i=as.integer(i)
  #   #cat(i,'\n')
  #   g=manta_tumor_table$SYMBOL[i]
  #   # genes overlapping SV start
  #   t=manta_tumor_table$cumstart[i]
  #   ix=cumstart-10e3 < t & cumend +10e3 > t
  #   if (length(ix)>0) {
  #     genes=allgenes[ix,unique(name)]
  #     set(manta_tumor_table,i,'Starts_Near',value=paste(genes,collapse = ' '))
  #     if (g %in% genes) set(manta_tumor_table,i,'Junction',value='Y')
  #   }
  #   # genes overlapping SV end
  #   t=manta_tumor_table$altcumpos[i]
  #   ix=cumstart-10e3 < t & cumend +10e3 > t
  #   if (length(ix)>0) {
  #     genes=allgenes[ix,unique(name)]
  #     set(manta_tumor_table,i,'Ends_Near',value=paste(genes,collapse = ' '))
  #     if (g %in% genes) set(manta_tumor_table,i,'Junction',value='Y')
  #   }
  # }
  # # Inside genes: Junctions may be putative fusion or knock-out if either end is "this" gene.
  # afflicted_ok <- which(manta_tumor_table$Junction=='Y')
  # 
  # ## Make new table with predicted effects
  # selection <- unique(
  #   manta_tumor_table[unique(sort(c(del_ok, amp_ok, afflicted_ok))),
  #                     .(SVTYPE,ID=THISID,chr,start,end,SVLEN=unlist(SVLEN),
  #                       SYMBOL,Starts_Near,Ends_Near,
  #                       Gene_Tier=3,Effect_Tier=3,LOH='',Comment='',
  #                       Consequence,REF,altpos,ALT=unlist(ALT))]
  # )
  # 
  # 
  # # Set gene tier to 1,2 based on Cosmic
  # selection[SYMBOL %in% tumorgenes$`Gene Symbol` | SYMBOL %in% local_tumorgenes$Gene]$Gene_Tier <- 2
  # selection[SYMBOL %in% tumorgenes[Tier==1]$`Gene Symbol`]$Gene_Tier <- 1
  # selection <- selection[order(Gene_Tier)]
  # 
  # # Set Effect tiers: 
  # for (i in selection[Gene_Tier<3,.I]) {
  #   symbol=selection$SYMBOL[i]
  #   starts_at=selection$Starts_Near[i]
  #   ends_at=selection$Ends_Near[i]
  #   length=abs(selection$SVLEN[i])
  #   gene=tumorgenes[`Gene Symbol`==symbol,] # this gene from Cosmic table
  #   # first fusions/translocations
  #   g=grep('fusion',gene$`Role in Cancer`)
  #   if (length(g)>0) { # fusion gene yes
  #     if (symbol %in% c(starts_at,ends_at)) { # and the start or end is at this (SYMBOL) gene
  #       # now 1 if perfect match with Cosmic, 2 if partial match
  #       mates=strsplit(gene$`Translocation Partner`,', ')[[1]] # this gene's common tx mates
  #       if (starts_at != ends_at) { # if this symbol is either start or end, not both
  #         if (starts_at %in% mates | ends_at %in% mates) { # and either is a member of known fusion mates
  #           selection$Effect_Tier[i]=1 # common translocation/fusion
  #           selection$Comment[i]=paste(selection$Comment[i],'known_fusion')
  #         } else {
  #           selection$Effect_Tier[i]=2 # then at least tier 2
  #           selection$Comment[i]=paste0(selection$Comment[i],'partial_known_fusion')
  #         }
  #       }
  #     }
  #   }
  #   # then deletions
  #   if (selection$SVTYPE[i]=='DEL' & length<3e6) {
  #     g=grep('TSG',gene$`Role in Cancer`)
  #     if (length(g)>0) { # Tumor supressor gene yes
  #       selection$Effect_Tier[i]=1
  #       selection$Comment[i]=paste(selection$Comment[i],'TSG_del')
  #     } else { # else 2 unless already one
  #       selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #       selection$Comment[i]=paste(selection$Comment[i],'non_TSG_del')
  #     }
  #   }
  #   # then amplifications
  #   if (selection$SVTYPE[i]=='DUP' & length<5e6) {
  #     g=grep('oncogene',gene$`Role in Cancer`)
  #     if (length(g)>0) { # Oncogene yes
  #       selection$Effect_Tier[i]=1
  #       selection$Comment[i]=paste(selection$Comment[i],'oncogene_dup')
  #     } else {
  #       selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #       selection$Comment[i]=paste(selection$Comment[i],'non_oncogene_dup')
  #     }
  #   }
  #   # last other disruptions
  #   if (symbol %in% c(starts_at,ends_at)) { # if the start or end is at this (SYMBOL) gene
  #     selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2) # some indication of effect!
  #     selection$Comment[i]=paste(selection$Comment[i],'at_breakpoint')
  #   }
  # }
  # 
  # 
  # # Add ASCAT LOH.
  # if (nrow(selection)>0) for (i in 1:nrow(selection)) {
  #   ix=ascat_cnv$chr==selection$chr[i] &
  #     ascat_cnv$startpos<selection$start[i] &
  #     ascat_cnv$endpos>selection$end[i]
  #   ix=ix[!is.na(ix)]
  #   if (sum(ix) >0) if (ascat_cnv[ix,LOH][1]=='LOH') selection$LOH[i]='Y'
  # }
  
  
  

manta_tumor_selected <- selection[order(rank_score,decreasing = T)]
htmlTable(manta_tumor_selected[rank_score>2], col.rgroup = c("none", "#F7F7F7")) #%>%
#kable_styling(bootstrap_options = "striped",position = "left") %>%
#column_spec(7, width = "20em") %>%
#column_spec(9, width = "20em") %>%
#scroll_box(width = "1400px", height = "600px")

}
```

<!-- Read the normal structural variant data -->
### Manta (normal)
```{r load_manta_normal, echo=FALSE}
if (length(vep_Nstruct_file)>0) {
  nstruc_vcf=readVcf(file = vep_Nstruct_file,genome = 'GRCh38')
  g=geno(nstruc_vcf)
  inf=info(nstruc_vcf)
  rr=rowRanges(nstruc_vcf)
  
  # manipulate into a data frame with relevant data, "info"" annotations also need to be added
  nstruc=cbind(data.table(THISID=names(rr)),as.data.table(rr),as.data.table(inf))[FILTER=='PASS' & IMPRECISE=='FALSE']
  nstruc$SVLEN=listfix(nstruc$SVLEN)
  setkey(x = nstruc,THISID)
  nstruc$chr <- substr(x = nstruc$seqnames,start = 4,stop = 6)
  nstruc$cumstart=nstruc$start
  nstruc$cumend=nstruc$end
  nstruc$altchr=nstruc$chr
  nstruc$altpos=nstruc$END
  nstruc$plot=F
  nstruc$arc= -1
  
  # Filter out variants seen 2+ (?) times in reference data
  ## Key has only chr,start,end
  key=nstruc[,c('chr','start','end')]
  key$imprecise='(pr)'
  ## If imprecise, round the pos to 10
  ix=nstruc$IMPRECISE==T
  key$imprecise[ix]='(impr)'
  key$start[ix]=round(key$start[ix]/10)*10
  key$end[ix]=round(key$end[ix]/10)*10
  key=paste(key$chr,key$start,key$end,key$imprecise)
  
  # now using 5 observations as cutoff
  nstruc$Swegen_count=swegen_hg38_manta_all[key,value]
  nstruc$Swegen_count[is.na(nstruc$Swegen_count)]=0
  ## do the filter
  nstruc <- nstruc[Swegen_count<5]
  
  # loop through all and extract endpoint chr and pos
  for (i in 1:nrow(nstruc)) {
    t=strsplit(x = nstruc$ALT[[i]],split = ':')[[1]]
    if (length(t)>1 & t[1]!="<DUP") {
      tchr=strsplit(x = t[1],split = 'chr')[[1]][2]
      if (is.na(tchr)) tchr=strsplit(x = t[1],split = 'CHR')[[1]][2]
      nstruc$altchr[i]=tchr
      tt=strsplit(t[2],'\\[')[[1]][1]; tt=strsplit(tt[1],'\\]')[[1]][1]
      nstruc$altpos[i]=as.numeric(tt)
    }
    # also decide how it is to be plotted (not represented elsewhere, up or down arc)
    if (nstruc$chr[i]==nstruc$altchr[i]) {
      nstruc$plot[i]=T
      nstruc$arc[i]=1
    } else {
      nstruc$plot[i]=T
      try({
        if (nstruc[nstruc$MATEID[i][[1]],'plot']==T) nstruc$plot[i]=F # if mate not being plotted, plot this one
      }, silent=T)
    }
  }
  # # for each chr get cumulative pos
  nstruc$altcumpos=nstruc$altpos
  for (i in 1:nrow(chrsz)) {
    ix=nstruc$chr==chrsz$chr[i]
    if (sum(ix)>0) {
      nstruc$cumstart[ix]=nstruc$start[ix]+chrsz$starts[i]
      nstruc$cumend[ix]=nstruc$end[ix]+chrsz$starts[i]
    }
    ix=nstruc$altchr==chrsz$chr[i]
    if (sum(ix)>0) {
      nstruc$altcumpos[ix]=nstruc$altpos[ix]+chrsz$starts[i]
    }
  }
  
  
  # snpEff annotation is in the ANN column.
  h=strsplit(info(header(nstruc_vcf))['ANN',][,3],'annotations: \'')[[1]][2]
  snpEff_header=trimws(strsplit(h,'\\|')[[1]])
  
  # snpEff annotation is put in snpEff_table
  snpEff_table=matrix(data = NA,nrow = length(unlist(nstruc$ANN)),ncol = length(snpEff_header)+1)
  colnames(snpEff_table)=c('N',snpEff_header)
  row=1
  for (i in 1:nrow(nstruc)) if (length(nstruc$ANN[[i]])>0) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(nstruc$ANN[[i]])) {
      line=strsplit(nstruc$ANN[[i]][j],'\\|')[[1]]
      snpEff_table[row,1]=i
      snpEff_table[row,1+(1:length(line))]=line
      row=row+1
    }
  }
  snpEff_table=unique(as.data.table(snpEff_table)[,1:5])
  
  # Add to data
  nstruc$N=as.character(1:nrow(nstruc))
  manta_normal_table <- merge(nstruc,snpEff_table,by='N')[order(cumstart),-1][nchar(Gene_Name)<40]
  
  ## Make new table with predicted effects
  selection <- unique(
    manta_normal_table[,
                       .(SVTYPE,ID=THISID,chr,start,end,Swegen_count,
                         altchr,altpos,SVLEN=unlist(SVLEN),
                         Gene_Name,rank_score=0,rank_terms='',LOH='',Comment='',
                         Annotation,Annotation_Impact,Allele,
                         REF,ALT=unlist(ALT))]
  )
  
  if (nrow(selection)>0) {  
    
    for (gene in unique(selection$Gene_Name)) if (gene!='') {
      ix=selection$Gene_Name==gene
      gene=strsplit(gene,'&')[[1]]
      # Add gene scores
      if (any(gene %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==1)])) {
        selection$rank_score[ix]=3
        selection$rank_terms[ix]='Tier1'
      } else if (any(gene %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==2)])) {
        selection$rank_score[ix]=2
        selection$rank_terms[ix]='Tier2'
      }
    }
    
    # Add high impact
    ix=selection$Annotation_Impact=='HIGH'
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'high_impact')
    }
    
    # +1 for focal
    ix=selection$chr==selection$altchr & selection$end-selection$start < 3e6 & selection$altpos-selection$start < 3e6
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'focal')
    }
    
    # +1 for fusion
    ix=grep('fusion',selection$Annotation)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'fusion')
    }
    
    # -1 for ablation if long del
    ix=intersect(
      which(selection$SVTYPE=='DEL' & selection$chr==selection$altchr & abs(selection$altpos-selection$start) > 3e6),
      grep('ablation',selection$Annotation)
    )
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]-1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'long_del')
    }
    
    # -1 for duplication if long dup
    ix=intersect(
      which(selection$SVTYPE=='DUP' & selection$chr==selection$altchr & abs(selection$altpos-selection$start) > 3e6),
      grep('duplication',selection$Annotation)
    )
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]-1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'long_dup')
    }
  }
  
  # Add Control Freec LOH.
  if (nrow(selection)>0) for (i in 1:nrow(selection)) {
    ix=freec_loh$chr==selection$chr[i] &
      freec_loh$start<selection$start[i] &
      freec_loh$end>selection$end[i]
    ix=ix[!is.na(ix)]
    if (sum(ix) >0) selection$LOH[i]='Y'
  }
  
  
  # # annotation is in the CSQ column.
  # h=strsplit(info(header(nstruc_vcf))['CSQ',][,3],'Format: ')[[1]][2]
  # vep_header=strsplit(h,'\\|')[[1]]
  # 
  # # VEP annotation is put in annotation_table
  # annotation_table=matrix(data = NA,nrow = length(unlist(nstruc$CSQ)),ncol = length(vep_header)+1)
  # colnames(annotation_table)=c('N',vep_header)
  # row=1
  # for (i in 1:nrow(nstruc)) { # for each variant
  #   # for each VEP annotation:
  #   for (j in 1:length(nstruc$CSQ[[i]])) { 
  #     line=strsplit(nstruc$CSQ[[i]][j],'\\|')[[1]]
  #     annotation_table[row,1]=i
  #     annotation_table[row,1+(1:length(line))]=line
  #     row=row+1
  #   }
  # }
  # annotation_table=as.data.table(annotation_table)
  # 
  # # Add to data
  # nstruc$N=as.character(1:nrow(nstruc))
  # manta_normal_table <- merge(nstruc,annotation_table,by='N')[order(cumstart),-1]
  # 
  # # Deletion: consider only deletions <3MB for "transcript ablation" consequence, report census gene(s)
  # del_ok <- which(manta_normal_table$SVTYPE=='DEL' & abs(manta_normal_table$SVLEN)<3e6)
  # # Amplification: consider <5M for amplification consequence, report census genes
  # amp_ok <- which(manta_normal_table$SVTYPE=='DUP' & abs(manta_normal_table$SVLEN)<5e6)
  # # Fusions/disrupting: Do the SVs start/end at gene? 
  # manta_normal_table$Starts_Near <- ''
  # manta_normal_table$Ends_Near <- ''
  # manta_normal_table$Junction <- ''
  # cumstart=allgenes$cumstart
  # cumend=allgenes$cumend
  # for (i in 1:nrow(manta_normal_table)) {
  #   i=as.integer(i)
  #   #cat(i,'\n')
  #   g=manta_normal_table$SYMBOL[i]
  #   # genes overlapping SV start
  #   t=manta_normal_table$cumstart[i]
  #   ix=cumstart-10e3 < t & cumend +10e3 > t
  #   if (length(ix)>0) {
  #     genes=allgenes[ix,unique(name)]
  #     set(manta_normal_table,i,'Starts_Near',value=paste(genes,collapse = ' '))
  #     if (g %in% genes) set(manta_normal_table,i,'Junction',value='Y')
  #   }
  #   # genes overlapping SV end
  #   t=manta_normal_table$altcumpos[i]
  #   ix=cumstart-10e3 < t & cumend +10e3 > t
  #   if (length(ix)>0) {
  #     genes=allgenes[ix,unique(name)]
  #     set(manta_normal_table,i,'Ends_Near',value=paste(genes,collapse = ' '))
  #     if (g %in% genes) set(manta_normal_table,i,'Junction',value='Y')
  #   }
  # }
  # # Inside genes: Junctions may be putative fusion or knock-out if either end is "this" gene.
  # afflicted_ok <- which(manta_normal_table$Junction=='Y')
  # 
  # ## Make new table with predicted effects
  # selection <- unique(
  #   manta_normal_table[unique(sort(c(del_ok, amp_ok, afflicted_ok))),
  #                      .(SVTYPE,ID=THISID,chr,start,end,SVLEN=unlist(SVLEN),
  #                        SYMBOL,Starts_Near,Ends_Near,
  #                        Gene_Tier=3,Effect_Tier=3,LOH='',Comment='',
  #                        Consequence,REF,altpos,ALT=unlist(ALT))]
  # )
  # 
  # # Set gene tier to 1,2 based on Cosmic
  # selection[SYMBOL %in% tumorgenes$`Gene Symbol` | SYMBOL %in% local_tumorgenes$Gene]$Gene_Tier <- 2
  # selection[SYMBOL %in% tumorgenes[Tier==1]$`Gene Symbol`]$Gene_Tier <- 1
  # selection <- selection[order(Gene_Tier)]
  # 
  # # Set Effect tiers: 
  # for (i in selection[Gene_Tier<3,.I]) {
  #   symbol=selection$SYMBOL[i]
  #   starts_at=selection$Starts_Near[i]
  #   ends_at=selection$Ends_Near[i]
  #   length=abs(selection$SVLEN[i])
  #   gene=tumorgenes[`Gene Symbol`==symbol,] # this gene from Cosmic table
  #   # first fusions/translocations
  #   g=grep('fusion',gene$`Role in Cancer`)
  #   if (length(g)>0) { # fusion gene yes
  #     if (symbol %in% c(starts_at,ends_at)) { # and the start or end is at this (SYMBOL) gene
  #       # now 1 if perfect match with Cosmic, 2 if partial match
  #       mates=strsplit(gene$`Translocation Partner`,', ')[[1]] # this gene's common tx mates
  #       if (starts_at != ends_at) { # if this symbol is either start or end, not both
  #         if (starts_at %in% mates | ends_at %in% mates) { # and either is a member of known fusion mates
  #           selection$Effect_Tier[i]=1 # common translocation/fusion
  #           selection$Comment[i]=paste(selection$Comment[i],'known_fusion')
  #         } else {
  #           selection$Effect_Tier[i]=2 # then at least tier 2
  #           selection$Comment[i]=paste0(selection$Comment[i],'partial_known_fusion')
  #         }
  #       }
  #     }
  #   }
  #   # then deletions
  #   if (selection$SVTYPE[i]=='DEL' & length<3e6) {
  #     g=grep('TSG',gene$`Role in Cancer`)
  #     if (length(g)>0) { # Tumor supressor gene yes
  #       selection$Effect_Tier[i]=1
  #       selection$Comment[i]=paste(selection$Comment[i],'TSG_del')
  #     } else { # else 2 unless already one
  #       selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #       selection$Comment[i]=paste(selection$Comment[i],'non_TSG_del')
  #     }
  #   }
  #   # then amplifications
  #   if (selection$SVTYPE[i]=='DUP' & length<5e6) {
  #     g=grep('oncogene',gene$`Role in Cancer`)
  #     if (length(g)>0) { # Oncogene yes
  #       selection$Effect_Tier[i]=1
  #       selection$Comment[i]=paste(selection$Comment[i],'oncogene_dup')
  #     } else {
  #       selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2)
  #       selection$Comment[i]=paste(selection$Comment[i],'non_oncogene_dup')
  #     }
  #   }
  #   # last other disruptions
  #   if (symbol %in% c(starts_at,ends_at)) { # if the start or end is at this (SYMBOL) gene
  #     selection$Effect_Tier[i]=min(selection$Effect_Tier[i],2) # some indication of effect!
  #     selection$Comment[i]=paste(selection$Comment[i],'at_breakpoint')
  #   }
  # }
  # # Add ASCAT LOH.
  # if (nrow(selection)>0) for (i in 1:nrow(selection)) {
  #   ix=ascat_cnv$chr==selection$chr[i] &
  #     ascat_cnv$startpos<selection$start[i] &
  #     ascat_cnv$endpos>selection$end[i]
  #   ix=ix[!is.na(ix)]
  #   if (sum(ix) >0) if (ascat_cnv[ix,LOH][1]=='LOH') selection$LOH[i]='Y'
  # }
  
  manta_normal_selected <- selection[order(rank_score,decreasing = T)]
  htmlTable(manta_normal_selected[rank_score>2], col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #column_spec(7, width = "20em") %>%
  #column_spec(9, width = "20em") %>%
  #scroll_box(width = "1400px", height = "600px")
  
}


```

&nbsp;


<!-- Plot the tumor and normal (also showing copy number) structural variant data -->
### View structural variants {.active}
```{r plot_manta, echo=FALSE,fig.width=15,fig.height=8}
cat(vep_Tstruct_file,'\n',vep_Nstruct_file)
if (exists('tstruc')) {
  par(mai=c(0,4,0,2))
  g=ggplot()+ylab('Structural variants')+expand_limits(x=c(0,3210e6))+ylim(-4,4)+
    scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
    scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.title.x=element_blank(),
            axis.text.x=element_blank(),
            axis.ticks.x=element_blank(),
            axis.title.y=element_blank(),
            axis.text.y=element_blank(),
            axis.ticks.y=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=-4,yend=4),data=chrsz,inherit.aes = F,alpha=1)
  # add copy data if present
  ### TUMOR LOGRATIO  
  if (exists('tratio')) {
    g=g+geom_line(aes(x=cumstart,y=log2(Ratio)+2),
                  data = tratio,alpha=1/3,
                  stat="identity",colour="black")
  } else if (exists('ascat_tratio')) {
    g=g+geom_line(aes(x=cumstart,y=log2(smoothed)),
                  data = ascat_tratio,alpha=1/3,
                  stat="identity",colour="black")
  }
  ### NORMAL LOGRATIO  
  if (exists('nratio')) {
    g=g+geom_line(aes(x=cumstart,y=log2(Ratio)-1),
                  data = nratio,alpha=1/3,
                  stat="identity",colour="grey")
  }
  
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  # add the TUMOR structural variants
  if (exists('manta_tumor_table')) {
    temp=unique(manta_tumor_table[,.(cumstart,altcumpos,SVTYPE,IMPRECISE)])
    if (nrow(temp)>0) g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=1.499,yend=1.5,col=SVTYPE,linetype=IMPRECISE),
                                     data = temp,alpha=1/2,
                                     stat = "identity",curvature=-0.2,size=0.5,
                                     position = "identity", angle = 90, ncp = 10,
                                     arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                     inherit.aes = FALSE)
  }
  
  # add the NORMAL structural variants
  temp=unique(manta_normal_table[THISID %in% manta_normal_selected[rank_score>3,ID],.(cumstart,altcumpos,SVTYPE,IMPRECISE)])
  if (nrow(temp)>0) g=g+geom_curve(aes(x = cumstart,xend = altcumpos,y=-1.499,yend=-1.5,col=SVTYPE,linetype=IMPRECISE),
                                   data = temp,alpha=1/2,
                                   stat = "identity",curvature=-0.2,size=0.5,
                                   position = "identity", angle = 90, ncp = 10,
                                   arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                   inherit.aes = FALSE)
  # add TUMOR data afflicted cosmic genes
  # temp=allgenes[name %in% manta_tumor_selected$Gene_Name[manta_tumor_selected$Effect_Tier<3]]
  # if (nrow(temp)>0) g=g+geom_label_repel(mapping = aes(x = cumstart,y = 1.45,label=name),
  #                                        data = temp,
  #                                        size=3)
  # # # add NORMAL data afflicted cosmic genes
  # temp=allgenes[name %in% manta_normal_selected$Gene_Name[manta_normal_selected$Effect_Tier<3]]
  # if (nrow(temp)>0) g=g+geom_label_repel(mapping = aes(x = cumstart,y = -1.55,label=name),
  #                                        data = temp,
  #                                        size=3)
  
  g
}
```



## Somatic point mutations {.tabset}

### Mutect 2
<!-- Read annotated small variant data (Mutect 2) -->
```{r load_mutect2, echo=FALSE}
if (length(mutect2_file)>0) {
  vcf=readVcf(file = mutect2_file,genome ='GRCh38')
  
  # first filter on Swegen SNPs/artefacts
  in_ref=snptable[names(vcf)]
  in_ref$value[is.na(in_ref$value)]=0
  vcf=vcf[in_ref$value<5]
  
  
  m2vcf <- vcf # for later export
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  tmuts$Swegen_count=snptable[names(vcf),value]
  tmuts$Swegen_count[is.na(tmuts$Swegen_count)]=0
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  }
  # Headers are sometimes sample names instead of TUMOR-NORMAL
  headers=colnames(g$AD)
  if (!'TUMOR' %in% headers) {
    ix=grep('[TR]',headers)
    headers[ix]='TUMOR'
    headers[-ix]='NORMAL'
    colnames(g$AD)=headers
    colnames(g$AF)=headers # <-- assumes the same header order in AF
  }
  tad=t(as.data.frame((g$AD[,'TUMOR'])))   #<------- the fix so that T/N pos are not hard coded
  tsum=tad[,1]+tad[,2]
  nad=t(as.data.frame((g$AD[,'NORMAL'])))
  nsum=nad[,1]+nad[,2]
  tmuts=cbind(tmuts,g$AF)
  tmuts$AD=tad[,2]
  tmuts$DP=tsum
  tmuts$ADn=nad[,2]
  tmuts$DPn=nsum
  tmuts$reads=' '
  tmuts$reads[tmuts$AD<5]='<5'
  tmuts$reads[tmuts$AD>=5]='≥5'
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  
  ## get snpEff headers
  snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
  # ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  # key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  # key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  # key1_refAF[is.na(value),1]=0
  # tmuts$allele1_swegen_count=key1_refAF$value
  # remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  # tmuts <- tmuts[!remove,] # filtered.
  
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  
  
  # Annotation table then concatenated 
  mutect2_table=cbind(tmuts,annotation_table)
  
  # Keep high/mod-relevance in final output
  impact=which(mutect2_table$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',mutect2_table$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',mutect2_table$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',mutect2_table$SIFT)
  # And certain effects
  regulatory=c(grep('TF',mutect2_table$Consequence),grep('regulatory',mutect2_table$Consequence)) # <-  used below?
  
  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious,regulatory)))
  selection <- mutect2_table[
    keep,
    .(VARIANT_CLASS,Existing_variation,chr,start,end,
      Swegen_count,SYMBOL,Protein_position,Amino_acids,rank_score=0,rank_terms='',LOH='',
      Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
      MOTIF_NAME, MOTIF_POS, HIGH_INF_POS, MOTIF_SCORE_CHANGE,
      ref,alt,DP,AD,DPn,ADn)
    ]
  
  
  
  if (nrow(selection)>0) {  
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    selection$Protein_position[is.na(selection$Protein_position)]=''
    
    # if cancer genes within 100k of variant with no symbol (e.g. regulatory), assign (one, by tier) as symbol
    for (i in which(selection$SYMBOL=='')) {
      near=allgenes[chr==selection$chr[i] &
                      start<selection$end[i]+1e5 &
                      end>selection$start[i]-1e5]
      if (nrow(near)>0) {
        near_cancer=tumorgenes[`Gene Symbol` %in% near$name]
        if (nrow(near_cancer)>0) selection$SYMBOL[i]=near_cancer[order(Tier,na.last = T),`Gene Symbol`][1]
      }
    }  
    
    # Add gene scores
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==1)]
    selection$rank_score[ix]=3
    selection$rank_terms[ix]='Tier1'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==2)]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='Tier2'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(is.na(tumorgenes$Tier) & tumorgenes$`Gene Symbol`!='')]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='published_gene'
    
    # Add high impact
    ix=selection$IMPACT=='HIGH'
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'high_impact')
    }
    
    # Add clinvar pathogenic
    ix=grep('pathogenic',selection$CLIN_SIG)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'pathogenic')
    }
    
    # Add hotspots
    temp=paste(selection$SYMBOL,selection$Protein_position)
    ix <-
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+3
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'hotspot')
    }
    
    # Add trancription factors
    ix=grep('TF',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'TFBS')
    }
    
    # Add regulatory
    ix=grep('regulatory',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+0
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'regulatory')
    }
  }
  
  # Add Control Freec LOH.
  if (nrow(selection)>0) for (i in 1:nrow(selection)) {
    ix=freec_loh$chr==selection$chr[i] &
      freec_loh$start<selection$start[i] &
      freec_loh$end>selection$end[i]
    ix=ix[!is.na(ix)]
    if (sum(ix) >0) selection$LOH[i]='Y'
  }
  
  mutect2_selected <- selection[order(rank_score,decreasing = T)]
  
  htmlTable(mutect2_selected, col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #scroll_box(width = "1400px", height = "600px")
  
}
```

### Strelka
<!-- Read annotated small variant data (Strelka, two files) -->
```{r load_strelka, echo=FALSE}
if (length(strelka_snv_file) * length(strelka_indel_file)>0) {
  vcf=readVcf(file = strelka_snv_file,genome ='GRCh38')
  stvcf=vcf # for later use
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]='other'
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  tmuts$Swegen_count=NA
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
    # ix=tmuts$altchr==chrsz$chr[i]
    # tmuts$altcumpos[ix]=tmuts$altpos[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
  # ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  # key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  # key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  # key1_refAF[is.na(value),1]=0
  # tmuts$allele1_swegen_count=key1_refAF$value
  # remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  # tmuts <- tmuts[!remove,] # filtered.
  # 
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_snvs=cbind(tmuts,annotation_table)
  
  # Filter takes place here, as the variant rownames are not compatible with Swegen/Mutect/Haplotypecaller
  # filter by RSID
  rs_table=snptable[strelka_snvs$Existing_variation]
  rs_table$value[is.na(rs_table$value)]=0
  pos_table=snptable[strelka_snvs$names]
  pos_table$value[is.na(pos_table$value)]=0
  t=cbind(rs_table,pos_table)
  t$maxvalue=apply(t[,c(2,4)],MARGIN = 1,FUN = max)
  strelka_snvs$Swegen_count=t$maxvalue
  strelka_snvs=strelka_snvs[Swegen_count<5]
  
  # Keep high/mod-relevance in final output
  impact=which(strelka_snvs$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',strelka_snvs$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',strelka_snvs$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',strelka_snvs$SIFT)
  # And certain effects
  regulatory=c(grep('TF',strelka_snvs$Consequence),grep('regulatory',strelka_snvs$Consequence)) # <-  used below?
  
  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious,regulatory)))
  
  selection <- strelka_snvs[keep,
                            .(VARIANT_CLASS,Existing_variation,chr,start,end,Swegen_count,
                              SYMBOL,Protein_position,Amino_acids,rank_score=0,rank_terms='',LOH='',
                              Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
                              MOTIF_NAME, MOTIF_POS, HIGH_INF_POS, MOTIF_SCORE_CHANGE,
                              ref,alt,DP=DPt,DPn)]
  
  # indels are read and added below.
  
  ## Read Strelka indels
  vcf=readVcf(file = strelka_indel_file,genome ='GRCh38')
  g=geno(vcf)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data frame with relevant data
  tmuts=as.data.table(ranges(rowRanges(vcf)))
  tmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  tmuts$cumstart=tmuts$start
  tmuts$cumend=tmuts$end
  tmuts$ref=as.character(ref(vcf))
  tmuts$alt=as.character(unlist(alt(vcf)))
  tmuts$type=paste0(tmuts$ref,'>',tmuts$alt)
  tmuts$type[nchar(tmuts$ref)>nchar(tmuts$alt)]='del'
  tmuts$type[nchar(tmuts$ref)<nchar(tmuts$alt)]='ins'
  tmuts$type[nchar(tmuts$type)>3]=NA
  tmuts$type[tmuts$type=='T>G']='A>C'
  tmuts$type[tmuts$type=='T>C']='A>G'
  tmuts$type[tmuts$type=='T>A']='A>T'
  tmuts$type[tmuts$type=='G>T']='C>A'
  tmuts$type[tmuts$type=='G>C']='C>G'
  tmuts$type[tmuts$type=='G>A']='C>T'
  tmuts$filter=rr$FILTER
  tmuts$Swegen_count=NA
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=tmuts$chr==chrsz$chr[i]
    tmuts$cumstart[ix]=tmuts$start[ix]+chrsz$starts[i]
    tmuts$cumend[ix]=tmuts$end[ix]+chrsz$starts[i]
  }
  tmuts$DPt=g$DP[,2]
  tmuts$DPn=g$DP[,1]
  # "info"" annotations also need to be added
  tmuts=cbind(tmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # keep only PASS
  tmuts=tmuts[filter=='PASS',]
  
  # ## Before adding the (LARGE) info and annotation fields, filter with Swegen 
  # # Make 2 key sets: key1 and key2. If variant has 2 alt alleles, remove if both alleles satisfy criteria.
  # key1 <- paste(tmuts$chr,tmuts$start,tmuts$alt)
  # key1_refAF <- swegen_hg38_allele_counts[key1,.(value)]
  # key1_refAF[is.na(value),1]=0
  # tmuts$allele1_swegen_count=key1_refAF$value
  # remove <- key1_refAF$value>=2 #remove SNP if allele1 is ≥2
  # tmuts <- tmuts[!remove,] # filtered.
  
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(tmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(tmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(tmuts$CSQ[[i]])) { 
      line=strsplit(tmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  strelka_indels=cbind(tmuts,annotation_table)
  
  # Filter takes place here, as the variant rownames are not compatible with Swegen/Mutect/Haplotypecaller
  # filter by RSID
  rs_table=snptable[strelka_indels$Existing_variation]
  rs_table$value[is.na(rs_table$value)]=0
  pos_table=snptable[strelka_indels$names]
  pos_table$value[is.na(pos_table$value)]=0
  t=cbind(rs_table,pos_table)
  t$maxvalue=apply(t[,c(2,4)],MARGIN = 1,FUN = max)
  strelka_indels$Swegen_count=t$maxvalue
  strelka_indels=strelka_indels[Swegen_count<5]
  
  # Keep high/mod-relevance in final output
  impact=which(strelka_indels$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',strelka_indels$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',strelka_indels$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',strelka_indels$SIFT)
  # And certain effects
  regulatory=c(grep('TF',strelka_indels$Consequence),grep('regulatory',strelka_indels$Consequence)) # <-  used below?
  
  
  keep <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious,regulatory)))
  selection <- rbind(selection, # this already exists for SNVs
                     strelka_indels[keep,
                                    .(VARIANT_CLASS,Existing_variation,chr,start,end,Swegen_count,
                                      SYMBOL,Protein_position,Amino_acids,rank_score=0,rank_terms='',LOH='',
                                      Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
                                      MOTIF_NAME, MOTIF_POS, HIGH_INF_POS, MOTIF_SCORE_CHANGE,
                                      ref,alt,DP=DPt,DPn)]
  )
  if (nrow(selection)>0) {  
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    selection$Protein_position[is.na(selection$Protein_position)]=''
    
    # if cancer genes within 100k of variant with no symbol (e.g. regulatory), assign (one, by tier) as symbol
    for (i in which(selection$SYMBOL=='')) {
      near=allgenes[chr==selection$chr[i] &
                      start<selection$end[i]+1e5 &
                      end>selection$start[i]-1e5]
      if (nrow(near)>0) {
        near_cancer=tumorgenes[`Gene Symbol` %in% near$name]
        if (nrow(near_cancer)>0) selection$SYMBOL[i]=near_cancer[order(Tier,na.last = T),`Gene Symbol`][1]
      }
    }  
    
    # Add gene scores
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==1)]
    selection$rank_score[ix]=3
    selection$rank_terms[ix]='Tier1'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==2)]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='Tier2'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(is.na(tumorgenes$Tier) & tumorgenes$`Gene Symbol`!='')]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='published_gene'
    
    # Add high impact
    ix=selection$IMPACT=='HIGH'
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'high_impact')
    }
    
    # Add clinvar pathogenic
    ix=grep('pathogenic',selection$CLIN_SIG)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'pathogenic')
    }
    
    # Add hotspots
    temp=paste(selection$SYMBOL,selection$Protein_position)
    ix <-
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+3
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'hotspot')
    }
    
    # Add trancription factors
    ix=grep('TF',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'TFBS')
    }
    
    # Add regulatory
    ix=grep('regulatory',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+0
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'regulatory')
    }
  }
  
  # Add Control Freec LOH.
  if (nrow(selection)>0) for (i in 1:nrow(selection)) {
    ix=freec_loh$chr==selection$chr[i] &
      freec_loh$start<selection$start[i] &
      freec_loh$end>selection$end[i]
    ix=ix[!is.na(ix)]
    if (sum(ix) >0) selection$LOH[i]='Y'
  }
  
  strelka_selected <- selection[order(rank_score,decreasing = T)]
  
  htmlTable(strelka_selected, col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #scroll_box(width = "1000px", height = "600px")
  
  # compare content in callers
  mutect2_temp=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                     MARGIN = 1,FUN = paste,collapse=' ')
  strelka_temp=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                               strelka_indels[,c('chr','start','end','ref','alt'),]),
                     MARGIN = 1,FUN = paste,collapse=' ')
  mutect2_table$Callers <- 'Mutect 2'
  mutect2_table$Callers[mutect2_temp %in% strelka_temp]='Mutect 2 and Strelka'
  
}
```


<!-- Plot Mutect 2 mutations along the genome -->
### View somatic point mutations {.active}

#### Genome view
```{r plot_mutect2, echo=FALSE,fig.width=15,fig.height=6}
cat(mutect2_file,strelka_indel_file,strelka_snv_file,sep = '\n')
if (exists('mutect2_table')) {
  par(mai=c(1,4,0,2))
  
  g=ggplot(data = mutect2_table[mutect2_table$filter=='PASS' & mutect2_table$Callers=='Mutect 2 and Strelka',],
           aes(x=cumstart,y=TUMOR,fill=type,shape=type,col=reads))+
    geom_point(alpha=1)+
    ylab('Somatic mutation allele ratio')+
    expand_limits(y=c(0,1))+
    expand_limits(x=c(0,3210e6))+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
  # add smoothed
  # fix theme
  g=g+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.title.x=element_blank(),
            axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
  # add chromosome lines
  g=g+geom_segment(mapping = aes(x = starts,xend=starts,y=0,yend=1),data=chrsz,inherit.aes = F,alpha=1/5)
  # add chromosome labels
  g=g+geom_text(aes(x=starts+0.5*length,y=0,label=chr),data = chrsz,inherit.aes = F)
  
  # add gene labels
  temp=mutect2_table[mutect2_table$filter=='PASS' & mutect2_table$Amino_acids!='',]
  g=g+geom_label_repel(data = temp,
                       mapping = aes(x=cumstart,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),
                       inherit.aes = F,size=3,nudge_y = 0.2)
  
  
  g
  # m=mutect2_table[,c('chr','start','end','ref','alt','AD','DP','ADn','DPn','Consequence','IMPACT','SYMBOL',
  #                    'BIOTYPE','EXON','Existing_variation','Protein_position','Amino_acids','Existing_variation')]
  # knitr::kable(m)
}
```

&nbsp;

#### Coverage - allele frequency
<!-- Make a scatter plot of Mutect2 findings, showing which are also in Strelka -->
```{r scatter_plot_mutect2, echo=FALSE,,fig.width=15,fig.height=10,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {
  
  # compare content in callers
  mutect2_temp=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                     MARGIN = 1,FUN = paste,collapse=' ')
  strelka_temp=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                               strelka_indels[,c('chr','start','end','ref','alt'),]),
                     MARGIN = 1,FUN = paste,collapse=' ')
  temp <- mutect2_table
  temp$Callers <- 'Mutect 2'
  temp$Callers[mutect2_temp %in% strelka_temp]='Mutect 2 and Strelka'
  
  g <- ggplot()+geom_point(mapping = aes(x=DP,y=AD/DP,fill=type,shape=type,col=Callers),data=temp)+
    scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
    scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))+
    scale_color_manual(values=c('Mutect 2'='lightgrey','Mutect 2 and Strelka'='black'))+
    ylab('ALT allele ratio') + xlab('Local coverage')+
    scale_x_log10()+
    geom_label_repel(data = temp[temp$Amino_acids!='',],mapping = aes(x=DP,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),inherit.aes = F,size=2,nudge_y = 0,nudge_x = 0)
  g#ggplotly(g)  
}
```

&nbsp;

#### Venn diagram
<!-- Make a Venn diagram of Strelka and Mutect2 findings -->
```{r mutect2_strelka_venn_diagram, echo=FALSE,,fig.width=6,fig.height=4,warning=F}
if (exists('strelka_indels') &
    exists('strelka_snvs') &
    exists('mutect2_table')) {
  
  VENN.LIST=list(mutect2=apply(X = mutect2_table[,c('chr','start','end','ref','alt'),],
                               MARGIN = 1,FUN = paste,collapse=' '),
                 strelka=apply(X = rbind(strelka_snvs[,c('chr','start','end','ref','alt'),],
                                         strelka_indels[,c('chr','start','end','ref','alt'),]),
                               MARGIN = 1,FUN = paste,collapse=' '))
  venn.plot <- venn.diagram(VENN.LIST , NULL, fill=c("darkmagenta", "darkblue"), 
                            alpha=c(0.2,0.2), cex = 2, cat.fontface=4, 
                            category.names=c("Mutect 2", "Strelka"))
  grid.draw(venn.plot)
}
```




## Germline point mutations {.tabset}

### HaplotypeCaller (normal+tumor)
<!-- Read annotated small variant data (Haplotype caller, normal) -->
```{r load_haplotype_caller, echo=FALSE}
if (length(haplotypecaller_N_file)>0) {
  vcf=readVcf(file = haplotypecaller_N_file,genome ='GRCh38')
  in_ref=snptable[names(vcf)]
  in_ref$value[is.na(in_ref$value)]=0
  vcf=vcf[in_ref$value<5]  # 32000 kvar, bättre än de 97k som fanns kvar med alt 1 (och <2)
  
  vcf_t=readVcf(file = haplotypecaller_T_file,genome ='GRCh38')
  vcf_t=vcf_t[names(vcf_t) %in% names(vcf)] # 22k återstår, alltså bara var 3dje från N
  vcf=vcf[names(vcf) %in% names(vcf_t)] # rensa även bort de som är specifika för N
  vcf_t=vcf_t[match(names(vcf),names(vcf_t))] # ordning är nu matchad
  ix_NinT=match(names(vcf),names(vcf_t)) # how to pick from vcf_t to match vcf (not really needed anymore)
  g=geno(vcf)
  g_t=geno(vcf_t)
  inf=(info(vcf))
  rr=rowRanges(vcf)
  # manipulate into a data table with relevant data
  nmuts=as.data.table(ranges(rowRanges(vcf)))
  nmuts$chr <- substr(as.character(seqnames(rowRanges(vcf))),start = 4,stop = 6)
  nmuts$cumstart=nmuts$start
  nmuts$cumend=nmuts$end
  nmuts$ref=as.character(ref(vcf))
  nmuts$alt=''
  nmuts$alt2=''
  temp=as.data.table(alt(vcf))[, .(values = list(value)), by = group]
  ix=unlist(lapply(temp$values,FUN = length))==2
  nmuts$alt[!ix]=unlist(temp$values[!ix])
  t=unlist(temp$values[ix])
  nmuts$alt[ix]=t[seq(1,length(t),2)]
  nmuts$alt2[ix]=t[seq(2,length(t),2)]
  
  # # for each chr get cumulative pos
  for (i in 1:nrow(chrsz)) {
    ix=nmuts$chr==chrsz$chr[i]
    nmuts$cumstart[ix]=nmuts$start[ix]+chrsz$starts[i]
    nmuts$cumend[ix]=nmuts$end[ix]+chrsz$starts[i]
  }
  # read depth from normal and tumor
  nmuts$DP=g$DP[,1]
  nmuts$DP_t=g_t$DP[,1][ix_NinT]
  nmuts$DP_t[is.na(nmuts$DP_t)]=0
  # get allele depth from normal
  ad=matrix(0,nrow = nrow(nmuts),ncol = 3,dimnames=list(NULL,c('RD','AD','AD2')))
  n=unlist(lapply(g$AD,length)) # gets number of AD per variant
  ad[n==2,1:2]=matrix(data=unlist(g$AD[n==2]),ncol=2,byrow = T)
  ad[n==3,1:3]=matrix(data=unlist(g$AD[n==3]),ncol=3,byrow = T)
  nmuts=cbind(nmuts,ad)
  nmuts$AF=apply(X = ad[,2:3],MARGIN = 1,FUN = max)/nmuts$DP
  # gets allele depth from tumor
  ad=matrix(0,nrow = length(vcf_t),ncol = 3,dimnames=list(NULL,c('RD_t','AD_t','AD2_t')))
  n=unlist(lapply(g_t$AD,length)) # gets number of AD per variant
  ad[n==2,1:2]=matrix(data=unlist(g_t$AD[n==2]),ncol=2,byrow = T)
  ad[n==3,1:3]=matrix(data=unlist(g_t$AD[n==3]),ncol=3,byrow = T)
  ad=ad[ix_NinT,]; ad[is.na(ad)]=0
  nmuts=cbind(nmuts,ad)
  nmuts$AF_t=apply(X = ad[,2:3],MARGIN = 1,FUN = max)/nmuts$DP_t
  
  # add info on how many in Swegen
  nmuts$Swegen_count=snptable[names(vcf),value]
  nmuts$Swegen_count[is.na(nmuts$Swegen_count)]=0
  
  # add type
  nmuts$type=paste0(nmuts$ref,'>',nmuts$alt)
  nmuts$type[nchar(nmuts$ref)>nchar(nmuts$alt)]='del'
  nmuts$type[nchar(nmuts$ref)<nchar(nmuts$alt)]='ins'
  nmuts$type[nchar(nmuts$type)>3 | nmuts$alt2!='']=NA
  nmuts$type[nmuts$type=='T>G']='A>C'
  nmuts$type[nmuts$type=='T>C']='A>G'
  nmuts$type[nmuts$type=='T>A']='A>T'
  nmuts$type[nmuts$type=='G>T']='C>A'
  nmuts$type[nmuts$type=='G>C']='C>G'
  nmuts$type[nmuts$type=='G>A']='C>T'
  
  # "info"" annotations also need to be added
  nmuts=cbind(nmuts,as.data.table(info(vcf)))
  ## get snpEff headers
  #snpeff_header=strsplit(info(header(vcf))['ANN',][,3],'ions: \'')[[1]][2]
  #snpeff_header=strsplit(snpeff_header,' \\| ')[[1]]
  ## get VEP headers
  vep_header=strsplit(info(header(vcf))['CSQ',][,3],'Format: ')[[1]][2]
  vep_header=strsplit(vep_header,'\\|')[[1]]
  
  # VEP annotation is put in annotation_table
  annotation_table=matrix(data = NA,nrow = nrow(nmuts),ncol = length(vep_header))
  colnames(annotation_table)=vep_header
  for (i in 1:nrow(nmuts)) { # for each variant
    # for each VEP annotation:
    for (j in 1:length(nmuts$CSQ[[i]])) { 
      line=strsplit(nmuts$CSQ[[i]][j],'\\|')[[1]]
      annotation_table[i,1:length(line)]=line
    }
  }
  
  # Annotation table then concatenated 
  haplotypecaller_table=cbind(nmuts,annotation_table)
  
  
  # double check Swegen in case ID was not matched (rsid in reference but not in data)
  temp=haplotypecaller_table[Swegen_count==0]
  ids=strsplit(paste(temp$Existing_variation,collapse='&'),'&')[[1]]
  ids=grep('^rs',ids,value = T)
  snps=snptable[ids][!is.na(value)]
  if (nrow(snps)>0)
    for (i in 1:nrow(snps)) 
      haplotypecaller_table$Swegen_count[grep(snps$name[i],haplotypecaller_table$Existing_variation)]=snps$value[i]
  # double check Swegen in case ID was not matched (pos in reference but not properly matched in data)
  pos=paste0('chr',haplotypecaller_table$chr,':',haplotypecaller_table$start,'_',haplotypecaller_table$ref,'/',haplotypecaller_table$alt)
  snps=snptable[c(pos)][!is.na(value)]
  if (nrow(snps)>0) haplotypecaller_table$Swegen_count[match(snps$name,pos)]=snps$value
  pos=paste0('chr',haplotypecaller_table$chr,':',haplotypecaller_table$start,'_',haplotypecaller_table$ref,'/',haplotypecaller_table$alt2)
  snps=snptable[c(pos)][!is.na(value)]
  if (nrow(snps)>0) haplotypecaller_table$Swegen_count[match(snps$name,pos)]=snps$value
  
  # Re-filter on counts
  haplotypecaller_table=haplotypecaller_table[Swegen_count<5]
  
  # Keep high/mod-relevance in final output
  impact=which(haplotypecaller_table$IMPACT %in% c('HIGH','MODERATE'))
  # Also keep Clinvar-pathogenic
  clinvar_pathogenic=grep('pathogenic',haplotypecaller_table$CLIN_SIG)
  # And PolyPhen-damaging
  polyphen_damaging=grep('damaging',haplotypecaller_table$PolyPhen)
  # And SIFT-deleterious
  sift_deleterious=grep('deleterious',haplotypecaller_table$SIFT)
  # And certain effects
  regulatory=c(grep('TF',haplotypecaller_table$Consequence),grep('regulatory',haplotypecaller_table$Consequence)) # <-  used below?
  
  keep_haplotypecaller <- sort(unique(c(impact,clinvar_pathogenic,polyphen_damaging,sift_deleterious,regulatory))) #,regulatory
  
  selection <- haplotypecaller_table[keep_haplotypecaller,
                                     .(VARIANT_CLASS,Existing_variation,chr,start,end,
                                       Swegen_count,SYMBOL,Protein_position,Amino_acids,rank_score=0,rank_terms='',LOH='',
                                       Consequence,IMPACT,CLIN_SIG,PolyPhen,SIFT,
                                       MOTIF_NAME, MOTIF_POS, HIGH_INF_POS, MOTIF_SCORE_CHANGE,
                                       ref,alt,alt2,DP,DP_t,RD,RD_t,AD,AD_t,AD2,AD2_t)]
  
  if (nrow(selection)>0) {  
    for (i in 1:nrow(selection)) selection$Protein_position[i] <- strsplit(selection$Protein_position[i],'/')[[1]][1]
    selection$Protein_position[is.na(selection$Protein_position)]=''
    
    # if cancer genes within 100k of variant with no symbol (e.g. regulatory), assign (one, by tier) as symbol
    for (i in which(selection$SYMBOL=='')) {
      near=allgenes[chr==selection$chr[i] &
                      start<selection$end[i]+1e5 &
                      end>selection$start[i]-1e5]
      if (nrow(near)>0) {
        near_cancer=tumorgenes[`Gene Symbol` %in% near$name]
        if (nrow(near_cancer)>0) selection$SYMBOL[i]=near_cancer[order(Tier,na.last = T),`Gene Symbol`][1]
      }
    }  
    
    # Add gene scores
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==1)]
    selection$rank_score[ix]=3
    selection$rank_terms[ix]='Tier1'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(tumorgenes$Tier==2)]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='Tier2'
    ix=selection$SYMBOL %in% tumorgenes$`Gene Symbol`[which(is.na(tumorgenes$Tier) & tumorgenes$`Gene Symbol`!='')]
    selection$rank_score[ix]=2
    selection$rank_terms[ix]='published_gene'
    
    # Add high impact
    ix=selection$IMPACT=='HIGH'
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'high_impact')
    }
    
    # Add clinvar pathogenic
    ix=grep('pathogenic',selection$CLIN_SIG)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+2
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'pathogenic')
    }
    
    # Add hotspots
    temp=paste(selection$SYMBOL,selection$Protein_position)
    ix <-
      temp %in% paste(hotspots_snv[,Hugo_Symbol],hotspots_snv[,Amino_Acid_Position]) |
      temp %in% paste(hotspots_inframe[,Hugo_Symbol],hotspots_inframe[,Amino_Acid_Position])
    if (any(ix)) {
      selection$rank_score[ix]=selection$rank_score[ix]+3
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'hotspot')
    }
    
    # Add trancription factors
    ix=grep('TF',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+1
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'TFBS')
    }
    
    # Add regulatory
    ix=grep('regulatory',selection$Consequence)
    if (length(ix)>0) {
      selection$rank_score[ix]=selection$rank_score[ix]+0
      selection$rank_terms[ix]=paste(selection$rank_terms[ix],'regulatory')
    }
  }
  
  # Add ASCAT LOH.
  if (nrow(selection)>0) for (i in 1:nrow(selection)) {
    ix=freec_loh$chr==selection$chr[i] &
      freec_loh$start<selection$start[i] &
      freec_loh$end>selection$end[i]
    ix=ix[!is.na(ix)]
    if (sum(ix) >0) selection$LOH[i]='Y'
  }
  
  
  haplotypecaller_selected <- selection[order(rank_terms)][order(rank_score,decreasing = T)]
  
  htmlTable(haplotypecaller_selected, col.rgroup = c("none", "#F7F7F7")) #%>%
  #kable_styling(bootstrap_options = "striped",position = "left") %>%
  #scroll_box(width = "1400px", height = "600px")
  
}
```



## All {.tabset .active}

### Combined table {.active}
```{r combined_findings}
## Combine all T1/2 data into smaller tables

# # T1 copy number only
# t=cnvs[Effect_Tier==2]
# if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                   Somatic=t$type,
#                                                   Gene=t$cosmic_genes,
#                                                   Mutation=t$effect,
#                                                   LOH='',
#                                                   Source='cFreec',
#                                                   Gene_Tier=t$Gene_Tier,
#                                                   Effect_Tier=t$Effect_Tier))
# t=ascat_cnv[Effect_Tier==2]
# if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                   Somatic='somatic',
#                                                   Gene=t$cosmic_genes,
#                                                   Mutation=ifelse(t$nMajor==0,'loss','gain'),
#                                                   LOH='',
#                                                   Source='Ascat',
#                                                   Gene_Tier=t$Gene_Tier,
#                                                   Effect_Tier=t$Effect_Tier))
# # Structural variants
# if (exists('manta_tumor_selected')) if (nrow(manta_tumor_selected)>0) {
#   t=unique(manta_tumor_selected[Effect_Tier==1 | Annotation=='gene_fusion',
#                                 .(Annotation,Gene_Name,LOH,Gene_Tier,Effect_Tier)])[nchar(Gene_Name)<30]
#   if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                     Somatic='somatic',
#                                                     Gene=t$Gene_Name,
#                                                     Mutation=t$Annotation,
#                                                     LOH=t$LOH,
#                                                     Source='Manta',
#                                                     Gene_Tier=t$Gene_Tier,
#                                                     Effect_Tier=t$Effect_Tier))
# }
# t=unique(manta_normal_selected[Effect_Tier==1 | Annotation=='gene_fusion',
#                                .(SVTYPE,Gene_Name,LOH,Gene_Tier,Effect_Tier)])[nchar(Gene_Name)<30]
# if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                   Somatic='germline',
#                                                   Gene=t$Gene_Name,
#                                                   Mutation=t$SVTYPE,
#                                                   LOH=t$LOH,
#                                                   Source='Manta',
#                                                   Gene_Tier=t$Gene_Tier,
#                                                   Effect_Tier=t$Effect_Tier))
# # SNVs/indels
# if (exists('mutect2_selected')) if (nrow(mutect2_selected)>0) {
#   t=unique(mutect2_selected[Effect_Tier<=2 & Gene_Tier<=2,
#                             .(Protein_position,Amino_acids,SYMBOL,LOH,Gene_Tier,Effect_Tier)])
#   if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                     Somatic='somatic',
#                                                     Gene=t$SYMBOL,
#                                                     Mutation=paste(t$Protein_position,t$Amino_acids),
#                                                     LOH=t$LOH,
#                                                     Source='Mutect2',
#                                                     Gene_Tier=t$Gene_Tier,
#                                                     Effect_Tier=t$Effect_Tier))
# }
# if (exists('strelka_selected')) if (nrow(strelka_selected)>0) {
#   t=unique(strelka_selected[Effect_Tier<=2 & Gene_Tier<=2,
#                             .(Protein_position,Amino_acids,SYMBOL,LOH,Gene_Tier,Effect_Tier)])
#   if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                     Somatic='somatic',
#                                                     Gene=t$SYMBOL,
#                                                     Mutation=paste(t$Protein_position,t$Amino_acids),
#                                                     LOH=t$LOH,
#                                                     Source='Strelka',
#                                                     Gene_Tier=t$Gene_Tier,
#                                                     Effect_Tier=t$Effect_Tier))
# }
# if (exists('haplotypecaller_selected')) if (nrow(haplotypecaller_selected)>0) {
#   t=unique(haplotypecaller_selected[Effect_Tier<=2 & Gene_Tier<=2,
#                                     .(Protein_position,Amino_acids,SYMBOL,LOH,Gene_Tier,Effect_Tier)])
#   if (nrow(t)>0) combined=rbind(combined,data.table(Patient=sampleData$name,
#                                                     Somatic='germline',
#                                                     Gene=t$SYMBOL,
#                                                     Mutation=paste(t$Protein_position,t$Amino_acids),
#                                                     LOH=t$LOH,
#                                                     Source='HaplotypeCaller',
#                                                     Gene_Tier=t$Gene_Tier,
#                                                     Effect_Tier=t$Effect_Tier))
# }
# 
# htmlTable(combined, col.rgroup = c("none", "#F7F7F7"))



## Below: saves tables to file.

# export overlap with Mutect2 to new file:
vrA <- as(m2vcf, "VRanges")
vrB <- as(stvcf, "VRanges")
writeVcf(obj = vrA[vrA %in% vrB & rowRanges(m2vcf)$FILTER=='PASS'],filename = paste0('~/reports/',sampleData$name,'_forPatterns.vcf'))
#
# export bed file of all relevant genomic regions
write.table(
  x = unique(rbind(
    manta_tumor_table[,.(chr,start,end=start)],
    manta_tumor_table[,.(chr,start=end,end)],
    manta_tumor_table[!is.na(altpos),.(chr=altchr,start=altpos,end=altpos)],
    manta_normal_table[,.(chr,start,end=start)],
    manta_normal_table[,.(chr,start=end,end)],
    manta_normal_table[!is.na(altpos),.(chr=altchr,start=altpos,end=altpos)],
    mutect2_table[,.(chr,start,end)],
    strelka_snvs[,.(chr,start,end)],
    strelka_indels[,.(chr,start,end)],
    haplotypecaller_table[,.(chr,start,end)])
  ),
  file = paste0('~/reports/',sampleData$name,'_forBamfile.bed'
  ),
  sep = '\t',row.names = F,col.names = F)
#
# export a data structure for downstream analysis
sampleList <- list(
  nameHere=list(
    sampleData=sampleData,
    cnvs=cnvs,
    ascat_cnv=ascat_cnv,
    manta_normal_selected=manta_normal_selected,#manta_normal_table,
    manta_tumor_selected=manta_tumor_selected,#manta_tumor_table,
    mutect2_selected=mutect2_selected,#mutect2_table,
    strelka_selected=strelka_selected,#strelka_indels,strelka_snvs,
    haplotypecaller_selected=haplotypecaller_selected#,#haplotypecaller_table # >2GB with the bid tables added
    #combined_findings=combined
  )
)
names(sampleList)=sampleData$name
save(sampleList,file=paste0('~/reports/',sampleData$name,'_data.Rata'))
```


### Combined view
```{r combined_view}
## Combine all data into a pretty set of figures

## basic plot element




# grid.newpage()
# grid.draw(rbind(ggplotGrob(plot1), ggplotGrob(plot2), size = "last"))

```


## Chromosomes {.tabset}
<!-- Plot composite data, chromosome wise -->
```{r chromosomes_plot_function}
chromplot2=function(thischr) {
  # basic plot element:
  g0=ggplot()+
    scale_x_continuous(
      breaks = 10e6*seq(0,ceiling(chrsz$length[chrsz$chr==thischr]/10e6)),
      labels= c('0',paste0(10*seq(1,ceiling(chrsz$length[chrsz$chr==thischr]/10e6)),'M')),
      minor_breaks = 1e6*seq(0,ceiling(chrsz$length[chrsz$chr==thischr]/1e6))
    )+
    theme(
      panel.grid.minor.x = element_line(colour="lightgrey", size=0.05),
      panel.grid.major.x = element_line(colour="grey", size=0.2),
      panel.grid.major.y = element_line(colour="black", size=0.2),
      panel.grid.minor.y = element_line(colour="lightgrey", size=0.05),
      axis.title.x=element_blank(),
      axis.ticks.x=element_blank(),
      axis.title.y=element_blank(),
      axis.ticks.y=element_blank()
    )
  
  #copy number element
  g=g0+ylim(y=c(0,3))
  if (exists('tratio')) {
    temp <- nratio
    temp$Ratio[temp$Ratio>3]=3
    g=g+geom_line(aes(x=Start,y=Ratio),
                  data = temp[temp$Chromosome==thischr,],alpha=1/3,
                  stat="identity",colour="red")
    temp <- tratio
    temp$Ratio[temp$Ratio>3]=3
    g=g+geom_line(aes(x=Start,y=Ratio),
                  data = temp[temp$Chromosome==thischr,],alpha=2/3,
                  stat="identity",colour="black")
  } else if (exists('ascat_tratio')) {
    temp <- ascat_tratio
    temp$smoothed[temp$smoothed>3]=3
    g=g+geom_line(aes(x=Position,y=smoothed),
                  data = temp[temp$Chr==thischr,],alpha=1/3,
                  stat="identity",colour="black")
  }
  g_cnv=g
  
  # structutal variants element
  g=g0+ylim(c(-1,1))
  if (exists('tratio')) {
    temp <- tratio
    temp$Ratio[temp$Ratio>3]=3
    g=g+geom_line(aes(x=Start,y=log2(Ratio)),
                  data = temp[temp$Chromosome==thischr,],alpha=1/10,
                  stat="identity",colour="black")
  } else if (exists('ascat_tratio')) {
    temp <- ascat_tratio
    temp$smoothed[temp$smoothed>3]=3
    g=g+geom_line(aes(x=Position,y=log2(smoothed)),
                  data = temp[temp$Chr==thischr,],alpha=1/10,
                  stat="identity",colour="black")
  }
  if (exists('tstruc')) if (nrow(tstruc)>0) {
    # first same chrom SVs
    temp=tstruc[tstruc$chr==thischr & tstruc$FILTER=='PASS' & tstruc$altchr==tstruc$chr,]
    if (nrow(temp)>0) g=g+geom_curve(aes(x = start,xend = altpos,y=0,yend=0.001,col=SVTYPE,linetype=IMPRECISE),
                                     data = temp,alpha=1/2,
                                     curvature=-0.2,size=1,
                                     position = "identity", angle = 90, ncp = 50,
                                     arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                     inherit.aes = FALSE)+
        scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
        scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
    
  }
  g_sv=g
  
  # somatic small variants element
  g=g0+ylim(y=c(0,1))
  if (exists('mutect2_table')) if (nrow(mutect2_table)>0) {
    temp=mutect2_table[chr==thischr & filter=='PASS' & Callers=='Mutect 2 and Strelka',]
    if (nrow(temp)>0) { 
      g=g+geom_point(aes(x=start,y=TUMOR,fill=type,shape=type),data = temp,alpha=1)+
        scale_fill_manual(
          values=c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", 
                   "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", 
                   "del"="#A6761D", "ins"="#666666"))+
        scale_shape_manual(
          values=c("A>C"=21,"A>G"=21, "A>T"=21, 
                   "C>A"=21, "C>G"=21, "C>T"=21, 
                   "del"=24, "ins"=25))+
        scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))
      
      g=g+geom_label_repel(
        data = temp[temp$Amino_acids!='',],
        mapping = aes(x=start,y=TUMOR,label=paste(SYMBOL,Protein_position,Amino_acids)),
        inherit.aes = F,size=3,min.segment.length = 0,nudge_y = 0
      )
    }
  }
  g_mut=g
  
  # tumor SNPs
  g=g0+ylim(y=c(0,1))
  if (exists('tbaf')) { 
    g=g+geom_point(aes(x=Position,y=BAF),data = tbaf[tbaf$Chromosome==thischr,],alpha=1/8,shape='.')
  } else if (exists('ascat_tbaf'))  {
    g=g+geom_point(aes(x=Position,y=BAF),data = ascat_tbaf[ascat_tbaf$Chr==thischr,],alpha=1/8,shape='.')
  }
  if (exists('haplotypecaller_table')) { 
      temp=haplotypecaller_table[keep_haplotypecaller][chr==thischr]
      g=g+
        geom_point(aes(x=start,y=AF_t,fill=type,shape=type),
                   data = temp,alpha=1)+
        geom_label_repel(data = temp[Amino_acids!='',],
                         mapping = aes(x=start,y=AF_t,label=paste(SYMBOL,Protein_position,Amino_acids)),
                         inherit.aes = F,size=3,min.segment.length = 0)+
        scale_fill_manual(
          values=c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", 
                   "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", 
                   "del"="#A6761D", "ins"="#666666"))+
        scale_shape_manual(
          values=c("A>C"=21,"A>G"=21, "A>T"=21, 
                   "C>A"=21, "C>G"=21, "C>T"=21, 
                   "del"=24, "ins"=25))
  }
  g_tsnp=g
  
  # normal SNPs
  g=g0+ylim(y=c(0,1))
  if (exists('nbaf')) {
    g=g+geom_point(aes(x=Position,y=BAF),data = nbaf[nbaf$Chromosome==thischr,],alpha=1/8,shape='.')
  } else if (exists('ascat_nbaf')) {
    g=g+geom_point(aes(x=Position,y=NBAF),data = ascat_nbaf[ascat_nbaf$Chr==thischr,],alpha=1/8,shape='.')
  }
  if (exists('haplotypecaller_table')) { 
    g=g+
      geom_point(aes(x=start,y=AF,fill=type,shape=type),
                 data = temp,alpha=1)+
      geom_label_repel(data = temp[Amino_acids!='',],
                       mapping = aes(x=start,y=AF,label=paste(SYMBOL,Protein_position,Amino_acids)),
                       inherit.aes = F,size=3,min.segment.length = 0)+
        scale_fill_manual(
          values=c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", 
                   "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", 
                   "del"="#A6761D", "ins"="#666666"))+
        scale_shape_manual(
          values=c("A>C"=21,"A>G"=21, "A>T"=21, 
                   "C>A"=21, "C>G"=21, "C>T"=21, 
                   "del"=24, "ins"=25))
  }
  g_nsnp=g
  
  grid.newpage()
  grid.draw(rbind(ggplotGrob(g_cnv),
                  ggplotGrob(g_sv),
                  ggplotGrob(g_mut),
                  ggplotGrob(g_tsnp),
                  ggplotGrob(g_nsnp),size = "last"))
  
} # end of plot function
```

```{r old_plot_function}  
  chromplot=function(thischr) if (T) {
    par(mai=c(0,4,0,2))
    g=ggplot()+
      scale_y_continuous(limits=c(-2.7,3),
                         breaks = seq(1.5,-2.5,-0.5),
                         minor_breaks = seq(1.25,-2.25,-0.5),
                         labels =
                           c('Tumor DNA=1.5','Tumor Copy Ratio\n(median centered)','Tumor DNA=0.5\nMutation AF=1','Mutation AF=0.5','Mutation AF=0\nTumor SNP AF=1',
                             'Tumor\nSNPs','Tumor SNP AF=0\nNormal SNP AF=1','Normal\nSNPs','Normal SNP AF=0'))+
      scale_x_continuous(
        breaks = 10e6*seq(0,ceiling(chrsz$length[chrsz$chr==thischr]/10e6)),
        labels= c('0',paste0(10*seq(1,ceiling(chrsz$length[chrsz$chr==thischr]/10e6)),'M')),
        minor_breaks = 1e6*seq(0,ceiling(chrsz$length[chrsz$chr==thischr]/1e6)))+
      theme(panel.grid.minor.x = element_line(colour="lightgrey", size=0.05),
            panel.grid.major.x = element_line(colour="grey", size=0.2),
            panel.grid.major.y = element_line(colour="black", size=0.2),
            panel.grid.minor.y = element_line(colour="lightgrey", size=0.05),
            axis.title.x=element_blank(),
            axis.ticks.x=element_blank(),
            axis.title.y=element_blank(),
            axis.ticks.y=element_blank())#+

    # lines to separate parts of fig
    g=g+geom_segment(mapping = aes(x = 0,xend=chrsz$length[chrsz$chr==thischr],y=0.5,yend=0.5),alpha=1)
    g=g+geom_segment(mapping = aes(x = 0,xend=chrsz$length[chrsz$chr==thischr],y=-0.5,yend=-0.5),alpha=1)
    g=g+geom_segment(mapping = aes(x = 0,xend=chrsz$length[chrsz$chr==thischr],y=-1.5,yend=-1.5),alpha=1)
    g=g+geom_segment(mapping = aes(x = 0,xend=chrsz$length[chrsz$chr==thischr],y=-2.5,yend=-2.5),alpha=1)

    ### TUMOR (& normal) LOGRATIO
    if (exists('tratio')) {
      temp <- nratio
      temp$Ratio[temp$Ratio>3]=3
      g=g+geom_line(aes(x=Start,y=Ratio),
                    data = temp[temp$Chromosome==thischr,],alpha=1/3,
                    stat="identity",colour="red")
      temp <- tratio
      temp$Ratio[temp$Ratio>3]=3
      g=g+geom_line(aes(x=Start,y=Ratio),
                    data = temp[temp$Chromosome==thischr,],alpha=2/3,
                    stat="identity",colour="black")
    } else if (exists('ascat_tratio')) {
      temp <- ascat_tratio
      temp$smoothed[temp$smoothed>3]=3
      g=g+geom_line(aes(x=Position,y=smoothed),
                    data = temp[temp$Chr==thischr,],alpha=1/3,
                    stat="identity",colour="black")
    }

    ### TUMOR STRVAR
    if (exists('tstruc')) if (nrow(tstruc)>0) {
      # first same chrom SVs
      temp=tstruc[tstruc$chr==thischr & tstruc$FILTER=='PASS' & tstruc$altchr==tstruc$chr,]
      if (nrow(temp)>0) g=g+geom_curve(aes(x = start,xend = altpos,y=1.99,yend=2,col=SVTYPE,linetype=IMPRECISE),
                                       data = temp,alpha=1/2,
                                       curvature=-0.2,size=1,
                                       position = "identity", angle = 90, ncp = 50,
                                       arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
                                       inherit.aes = FALSE)+
          scale_color_manual(values=c("BND"="#E41A1C", "DEL"="#377EB8", "INV"="#4DAF4A", "INS"="#984EA3", "DUP"="#FF7F00"))+
          scale_linetype_manual(values=c('TRUE'=2,'FALSE'=1))
      # then diff chrom SVs
      # temp=tstruc[tstruc$chr==thischr & tstruc$FILTER=='PASS' & tstruc$altchr!=tstruc$chr,]
      # if (nrow(temp)>0) {
      #   g=g+geom_label_repel(data = temp,mapping = aes(x=start,y=rep(2,nrow(temp)),label=altchr),
      #                        inherit.aes = F,size=3)
      # }
      # labels
      # temp=allgenes[name %in% manta_tumor_selected$Gene_Name[manta_tumor_selected$Effect_Tier<3 & manta_tumor_selected$chr==thischr]]
      # if (nrow(temp)>0) g=g+geom_label_repel(mapping = aes(x = start,y = 2,label=name),
      #                                        data = temp,
      #                                        size=3)
    }
    ### NORMAL SVs
    # if (exists('nstruc')) if (nrow(nstruc)>0) {
    #   # first same chrom SVs
    #   temp=nstruc[nstruc$chr==thischr & nstruc$FILTER=='PASS' & nstruc$altchr==nstruc$chr,]
    #   if (nrow(temp)>0) g=g+geom_curve(aes(x = start,xend = altpos,y=0.149,yend=0.25,col=SVTYPE,linetype=IMPRECISE),
    #                                    data = temp,alpha=1/2,
    #                                    curvature=-0.2,size=1,
    #                                    position = "identity", angle = 90, ncp = 50,
    #                                    arrow = arrow(length=unit(0.10,"cm"), ends="both", type = "open"), lineend = "butt", na.rm = FALSE, show.legend = NA,
    #                                    inherit.aes = FALSE)
    #   # then diff chrom SVs
    #   temp=nstruc[nstruc$chr==thischr & nstruc$FILTER=='PASS' & nstruc$altchr!=nstruc$chr,]
    #   if (nrow(temp)>0) {
    #     g=g+geom_label_repel(data = temp,mapping = aes(x=start,y=0.25,label=altchr),
    #                          inherit.aes = F,size=3)
    #   }
    #   # labels
    #   temp=allgenes[name %in% manta_normal_selected$Gene_Name[manta_normal_selected$Effect_Tier<3 & manta_normal_selected$chr==thischr]]
    #   if (nrow(temp)>0) g=g+geom_label_repel(mapping = aes(x = start,y = 0.25,label=name),
    #                                          data = temp,
    #                                          size=3)
    # }
    #
    ### TUMOR MUTATIONS
    if (exists('mutect2_table')) if (nrow(mutect2_table)>0) {
      temp=mutect2_table[chr==thischr & filter=='PASS' & Callers=='Mutect 2 and Strelka',]
      if (nrow(temp)>0) {
        g=g+geom_point(aes(x=start,y=TUMOR -0.5,fill=type,shape=type),data = temp,alpha=1)+
          scale_fill_manual(values = c("A>C"="#1B9E77","A>G"="#D95F02", "A>T"="#7570B3", "C>A"="#E7298A", "C>G"="#66A61E", "C>T"="#E6AB02", "del"="#A6761D", "ins"="#666666"))+
          scale_shape_manual(values=c("A>C"=21,"A>G"=21, "A>T"=21, "C>A"=21, "C>G"=21, "C>T"=21, "del"=24, "ins"=25))#+
        #scale_color_manual(values=c('<5'='lightgrey','≥5'='black'))

        g=g+geom_label_repel(data = temp[temp$Amino_acids!='',],
                             mapping = aes(x=start,y=TUMOR -0.5,label=paste(SYMBOL,Protein_position,Amino_acids)),
                             inherit.aes = F,size=3,min.segment.length = 0,nudge_y = 0.3)
      }
    }

    ### SNPS + germline mutation (at the corresponding track)
    if (exists('tbaf')) {
      g=g+geom_point(aes(x=Position,y=BAF -1.5),data = tbaf[tbaf$Chromosome==thischr,],alpha=1/8,shape='.')
    } else if (exists('ascat_tbaf'))  {
      g=g+geom_point(aes(x=Position,y=BAF -1.5),data = ascat_tbaf[ascat_tbaf$Chr==thischr,],alpha=1/8,shape='.')
    }
    if (exists('nbaf')) {
      g=g+geom_point(aes(x=Position,y=BAF -2.5),data = nbaf[nbaf$Chromosome==thischr,],alpha=1/8,shape='.')
    } else if (exists('ascat_nbaf')) {
      g=g+geom_point(aes(x=Position,y=NBAF -2.5),data = ascat_nbaf[ascat_nbaf$Chr==thischr,],alpha=1/8,shape='.')
    }
    ## Germline muts
    if (exists('haplotypecaller_table')) {
      temp=haplotypecaller_table[keep_haplotypecaller][chr==thischr]
      g=g+
        geom_point(aes(x=start,y=AF_t -1.5,fill=type,shape=type),
                   data = temp,alpha=1)+
        geom_label_repel(data = temp[Amino_acids!='',],
                         mapping = aes(x=start,y=AF_t -1.5,label=paste(SYMBOL,Protein_position,Amino_acids)),
                         inherit.aes = F,size=3,min.segment.length = 0)
      g=g+
        geom_point(aes(x=start,y=AF -2.5,fill=type,shape=type),
                   data = temp,alpha=1)+
        geom_label_repel(data = temp[Amino_acids!='',],
                         mapping = aes(x=start,y=AF -2.5,label=paste(SYMBOL,Protein_position,Amino_acids)),
                         inherit.aes = F,size=3,min.segment.length = 0)
    }

    #genes!
    temp=allgenes[name %in% tumorgenes$`Gene Symbol` & chr==thischr]
    g=g+geom_text_repel(mapping = aes(x = (start+end)/2,y =  -2.6,label=name),
                        data=temp,
                        nudge_y= -0.1,size=2)
    return(g)
  }
```



### 1
```{r plot_chr1,fig.width=15,fig.height=10}
chr=1
cat('Chromosome',chr)
chromplot(chr)
```

### 2
```{r plot_chr2,fig.width=15,fig.height=10}
chr=2
cat('Chromosome',chr)
chromplot(chr)
```

### 3
```{r plot_chr3,fig.width=15,fig.height=10}
chr=3
cat('Chromosome',chr)
chromplot(chr)
```

### 4
```{r plot_chr4,fig.width=15,fig.height=10}
chr=4
cat('Chromosome',chr)
chromplot(chr)
```

### 5
```{r plot_chr5,fig.width=15,fig.height=10}
chr=5
cat('Chromosome',chr)
chromplot(chr)
```

### 6
```{r plot_chr6,fig.width=15,fig.height=10}
chr=6
cat('Chromosome',chr)
chromplot(chr)
```

### 7
```{r plot_chr7,fig.width=15,fig.height=10}
chr=7
cat('Chromosome',chr)
chromplot(chr)
```

### 8
```{r plot_chr8,fig.width=15,fig.height=10}
chr=8
cat('Chromosome',chr)
chromplot(chr)
```

### 9
```{r plot_chr9,fig.width=15,fig.height=10}
chr=9
cat('Chromosome',chr)
chromplot(chr)
```

### 10
```{r plot_chr10,fig.width=15,fig.height=10}
chr=10
cat('Chromosome',chr)
chromplot(chr)
```

### 11
```{r plot_chr11,fig.width=15,fig.height=10}
chr=11
cat('Chromosome',chr)
chromplot(chr)
```

### 12
```{r plot_chr12,fig.width=15,fig.height=10}
chr=12
cat('Chromosome',chr)
chromplot(chr)
```

### 13
```{r plot_chr13,fig.width=15,fig.height=10}
chr=13
cat('Chromosome',chr)
chromplot(chr)
```

### 14
```{r plot_chr14,fig.width=15,fig.height=10}
chr=14
cat('Chromosome',chr)
chromplot(chr)
```

### 15
```{r plot_chr15,fig.width=15,fig.height=10}
chr=15
cat('Chromosome',chr)
chromplot(chr)
```

### 16
```{r plot_chr16,fig.width=15,fig.height=10}
chr=16
cat('Chromosome',chr)
chromplot(chr)
```

### 17
```{r plot_chr17,fig.width=15,fig.height=10}
chr=17
cat('Chromosome',chr)
chromplot(chr)
```

### 18
```{r plot_chr18,fig.width=15,fig.height=10}
chr=18
cat('Chromosome',chr)
chromplot(chr)
```

### 19
```{r plot_chr19,fig.width=15,fig.height=10}
chr=19
cat('Chromosome',chr)
chromplot(chr)
```

### 20
```{r plot_chr20,fig.width=15,fig.height=10}
chr=20
cat('Chromosome',chr)
chromplot(chr)
```

### 21
```{r plot_chr21,fig.width=15,fig.height=10}
chr=21
cat('Chromosome',chr)
chromplot(chr)
```

### 22
```{r plot_chr22,fig.width=15,fig.height=10}
chr=22
cat('Chromosome',chr)
chromplot(chr)
```

### X
```{r plot_chrX,fig.width=15,fig.height=10}
chr='X'
cat('Chromosome',chr)
chromplot(chr)
```

### Y
```{r plot_chrY,fig.width=15,fig.height=10}
chr='Y'
cat('Chromosome',chr)
chromplot(chr)
```

&nbsp;
